// nocodb-sdk/build/main/lib/error/nc-base.error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NcBaseErrorv2 = exports.TestConnectionError = exports.OptionsNotExistsError = exports.UnprocessableEntity = exports.ExternalTimeout = exports.ExternalError = exports.MetaError = exports.SsoError = exports.NotFound = exports.Forbidden = exports.Unauthorized = exports.NotAllowed = exports.BadRequestV2 = exports.NcBaseError = void 0;
class NcBaseError extends Error {
    constructor(message) {
        super(message);
    }
}
exports.NcBaseError = NcBaseError;
class BadRequestV2 extends NcBaseError {
}
exports.BadRequestV2 = BadRequestV2;
class NotAllowed extends NcBaseError {
}
exports.NotAllowed = NotAllowed;
class Unauthorized extends NcBaseError {
}
exports.Unauthorized = Unauthorized;
class Forbidden extends NcBaseError {
}
exports.Forbidden = Forbidden;
class NotFound extends NcBaseError {
}
exports.NotFound = NotFound;
class SsoError extends NcBaseError {
}
exports.SsoError = SsoError;
class MetaError extends NcBaseError {
    constructor(param) {
        super(param.message);
        Object.assign(this, param);
    }
}
exports.MetaError = MetaError;
class ExternalError extends NcBaseError {
    constructor(error) {
        super(error.message);
        Object.assign(this, error);
    }
}
exports.ExternalError = ExternalError;
class ExternalTimeout extends ExternalError {
}
exports.ExternalTimeout = ExternalTimeout;
class UnprocessableEntity extends NcBaseError {
}
exports.UnprocessableEntity = UnprocessableEntity;
class OptionsNotExistsError extends BadRequestV2 {
    constructor({ columnTitle, options, validOptions, }) {
        super(`Invalid option(s) "${options.join(', ')}" provided for column "${columnTitle}". Valid options are "${validOptions.join(', ')}"`);
        this.columnTitle = columnTitle;
        this.options = options;
        this.validOptions = validOptions;
    }
}
exports.OptionsNotExistsError = OptionsNotExistsError;
class TestConnectionError extends NcBaseError {
    constructor(message, sql_code) {
        super(message);
        this.sql_code = sql_code;
    }
}
exports.TestConnectionError = TestConnectionError;
class NcBaseErrorv2 extends NcBaseError {
    constructor(message, code, error, args) {
        super(message);
        this.error = error;
        this.code = code;
        this.details = args === null || args === void 0 ? void 0 : args.details;
    }
}
exports.NcBaseErrorv2 = NcBaseErrorv2;
