// nocodb-sdk/build/main/lib/columnHelper/columns/Checkbox.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckboxHelper = void 0;
const __1 = require("..");
const column_interface_1 = __importDefault(require("../column.interface"));
class CheckboxHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            iconIdx: 0,
            icon: {
                checked: 'mdi-check-bold',
                unchecked: 'mdi-crop-square',
            },
            color: '#777',
        };
    }
    serializeValue(value) {
        return (0, __1.serializeCheckboxValue)(value);
    }
    parseValue(value) {
        return (0, __1.parseCheckboxValue)(value);
    }
    parsePlainCellValue(value) {
        return (0, __1.parseCheckboxValue)(value) ? 'Checked' : 'Unchecked';
    }
}
exports.CheckboxHelper = CheckboxHelper;
