// nocodb-sdk/build/main/lib/columnHelper/columns/DateTime.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateTimeHelper = void 0;
const dateTimeHelper_1 = require("../../../lib/dateTimeHelper");
const column_interface_1 = __importDefault(require("../column.interface"));
const utils_1 = require("../utils");
const error_1 = require("../../../lib/error");
const fill_handler_1 = require("../utils/fill-handler");
const helperFunctions_1 = require("../../../lib/helperFunctions");
const dayjs_1 = __importDefault(require("dayjs"));
const is_1 = require("../../../lib/is");
class DateTimeHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            date_format: dateTimeHelper_1.dateFormats[0],
            time_format: dateTimeHelper_1.timeFormats[0],
            is12hrFormat: false,
        };
    }
    serializeValue(value, params) {
        if (params.serializeSearchQuery)
            return null;
        value = (0, utils_1.serializeDateOrDateTimeValue)(value, params.col);
        if (value === null) {
            if (params.isMultipleCellPaste) {
                return null;
            }
            else {
                throw new error_1.SilentTypeConversionError();
            }
        }
        return value;
    }
    parseValue(value, params) {
        return (0, utils_1.parseDateTimeValue)(value, params);
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = (0, utils_1.parseDateTimeValue)(value, params)) !== null && _a !== void 0 ? _a : '';
    }
    populateFillHandle(params) {
        var _a;
        const { column, highlightedData, numberOfRows } = params;
        if (highlightedData.length < 2) {
            return (0, fill_handler_1.populateFillHandleStrictCopy)(params);
        }
        // data is in form like 'YYYY-MM-DD HH:mm' depends on meta dateformat
        const meta = (0, helperFunctions_1.parseProp)(column.meta);
        const metaDateFormat = (_a = meta.date_format) !== null && _a !== void 0 ? _a : 'YYYY-MM-DD';
        const metaTimeFormat = (0, dateTimeHelper_1.constructTimeFormat)(params.column);
        const dateTimeFormat = `${metaDateFormat} ${metaTimeFormat}`;
        let lastData;
        let modifier = undefined;
        let scale = 's';
        let canUseSecondScale = true;
        if (!dateTimeFormat.includes('ss')) {
            canUseSecondScale = false;
        }
        const setModifierForScaleIfRound = (currentData, lastData, diffScale) => {
            const currentModifier = currentData.diff(lastData, diffScale, true);
            if ((0, utils_1.isNumberRound)(currentModifier)) {
                scale = utils_1.DATE_DIFF_TO_SCALE_LABEL_MAP[diffScale];
                modifier = currentModifier;
            }
        };
        // map to dayjs
        const dayJsHighlightedData = [];
        for (const date of highlightedData) {
            if ((0, is_1.ncIsNullOrUndefined)(date) ||
                date === '' ||
                typeof date !== 'string') {
                return (0, fill_handler_1.populateFillHandleStrictCopy)(params);
            }
            const currentData = (0, dayjs_1.default)(date, dateTimeFormat);
            // unlikely on normal case
            if (!currentData.isValid()) {
                return (0, fill_handler_1.populateFillHandleStrictCopy)(params);
            }
            if (!lastData) {
                lastData = currentData;
            }
            else {
                if ((0, is_1.ncIsUndefined)(modifier)) {
                    const modifierInSeconds = currentData.diff(lastData, 'second');
                    if (modifierInSeconds > 60 * 60 * 24 * 354) {
                        setModifierForScaleIfRound(currentData, lastData, 'year');
                    }
                    else if (modifierInSeconds > 60 * 60 * 24 * 28) {
                        setModifierForScaleIfRound(currentData, lastData, 'month');
                    }
                    else if (modifierInSeconds > 60 * 60 * 24) {
                        setModifierForScaleIfRound(currentData, lastData, 'day');
                    }
                    else if (modifierInSeconds > 60 * 60) {
                        setModifierForScaleIfRound(currentData, lastData, 'hour');
                    }
                    else if (modifierInSeconds > 60) {
                        setModifierForScaleIfRound(currentData, lastData, 'minute');
                    }
                    else if (!canUseSecondScale) {
                        modifier = currentData.diff(lastData, 'minute');
                        scale = 'm';
                    }
                    else {
                        modifier = modifierInSeconds;
                        scale = 's';
                    }
                }
                else {
                    const currentModifier = currentData.diff(lastData, utils_1.DATE_SCALE_LABEL_TO_DIFF_MAP[scale]);
                    if (currentModifier !== modifier) {
                        return (0, fill_handler_1.populateFillHandleStrictCopy)(params);
                    }
                }
                lastData = currentData;
            }
            dayJsHighlightedData.push(lastData);
        }
        if (modifier === 0) {
            return (0, fill_handler_1.populateFillHandleStrictCopy)(params);
        }
        const numberToGenerate = numberOfRows - highlightedData.length;
        return Array.from({ length: numberToGenerate }).map(() => {
            const currentData = lastData.add(modifier, utils_1.DATE_SCALE_LABEL_TO_DIFF_MAP[scale]);
            lastData = currentData;
            return currentData.format(dateTimeFormat);
        });
    }
}
exports.DateTimeHelper = DateTimeHelper;
