// nocodb-sdk/build/main/lib/columnHelper/columns/Formula.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormulaHelper = void 0;
const column_interface_1 = __importDefault(require("../column.interface"));
const helperFunctions_1 = require("../../../lib/helperFunctions");
const column_helper_1 = require("../column-helper");
const error_1 = require("../../../lib/error");
const formulaHelpers_1 = require("../../../lib/formulaHelpers");
const is_1 = require("../../../lib/is");
class FormulaHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            display_column_meta: {
                meta: {},
                custom: {},
            },
            display_type: null,
        };
    }
    serializeValue(value, params) {
        var _a, _b, _c, _d;
        if (params.serializeSearchQuery) {
            const dataType = (_d = (_c = (_b = (_a = params.col) === null || _a === void 0 ? void 0 : _a.colOptions) === null || _b === void 0 ? void 0 : _b.parsed_tree) === null || _c === void 0 ? void 0 : _c.dataType) !== null && _d !== void 0 ? _d : formulaHelpers_1.FormulaDataTypes.STRING;
            if (dataType === formulaHelpers_1.FormulaDataTypes.NUMERIC) {
                return (0, is_1.ncIsNaN)(value) ? '' : value;
            }
            return value;
        }
        if (params.isMultipleCellPaste) {
            return undefined;
        }
        else {
            throw new error_1.ComputedTypePasteError();
        }
    }
    parseValue(value, params) {
        var _a;
        const columnMeta = (0, helperFunctions_1.parseProp)((_a = params.col) === null || _a === void 0 ? void 0 : _a.meta);
        const childColumn = Object.assign({ uidt: columnMeta.display_type }, columnMeta.display_column_meta);
        return column_helper_1.ColumnHelper.parseValue(value, Object.assign(Object.assign({}, params), { col: childColumn }));
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = this.parseValue(value, params)) !== null && _a !== void 0 ? _a : '';
    }
}
exports.FormulaHelper = FormulaHelper;
