// nocodb-sdk/build/main/lib/columnHelper/columns/Json.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonHelper = void 0;
const column_interface_1 = __importDefault(require("../../../lib/columnHelper/column.interface"));
const error_1 = require("../../../lib/error");
const utils_1 = require("../utils");
const is_1 = require("../../../lib/is");
class JsonHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {};
    }
    serializeValue(value, params) {
        if (params.serializeSearchQuery) {
            return (0, is_1.ncIsString)(value) ? value : null;
        }
        value = (0, utils_1.serializeJsonValue)(value);
        if (value === null) {
            throw new error_1.SilentTypeConversionError();
        }
        return value;
    }
    parseValue(value) {
        return (0, utils_1.parseJsonValue)(value);
    }
    parsePlainCellValue(value) {
        return (0, utils_1.parseJsonValue)(value);
    }
    equalityComparison(a, b, _param) {
        const aStr = typeof a === 'object'
            ? JSON.stringify(a)
            : typeof a === 'string'
                ? a
                : a.toString();
        const bStr = typeof b === 'object'
            ? JSON.stringify(b)
            : typeof b === 'string'
                ? b
                : b.toString();
        return aStr === bStr;
    }
}
exports.JsonHelper = JsonHelper;
