// nocodb-sdk/build/main/lib/columnHelper/columns/Links.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinksHelper = void 0;
const error_1 = require("../../../lib/error");
const column_interface_1 = __importDefault(require("../column.interface"));
const utils_1 = require("../utils");
const is_1 = require("../../../lib/is");
class LinksHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {};
    }
    serializeValue(value, params) {
        var _a;
        if (params.serializeSearchQuery)
            return null;
        if (!(0, utils_1.isMm)(params.col))
            throw new error_1.SilentTypeConversionError();
        let parsedVal = value;
        try {
            parsedVal = typeof value === 'string' ? JSON.parse(value) : value;
        }
        catch (_b) { }
        if (!(parsedVal &&
            (0, is_1.ncIsObject)(parsedVal) &&
            ['rowId', 'columnId', 'fk_related_model_id', 'value'].every((key) => parsedVal.hasOwnProperty(key))) ||
            (parsedVal === null || parsedVal === void 0 ? void 0 : parsedVal.fk_related_model_id) !==
                ((_a = params.col.colOptions) === null || _a === void 0 ? void 0 : _a.fk_related_model_id)) {
            throw new error_1.SilentTypeConversionError();
        }
        return parsedVal;
    }
    parseValue(value, params) {
        if ((0, utils_1.isMm)(params.col)) {
            return JSON.stringify({
                rowId: params.rowId,
                columnId: params.col.id,
                fk_related_model_id: params.col.colOptions
                    .fk_related_model_id,
                value: !(0, is_1.ncIsNaN)(value) ? +value : 0,
            });
        }
        else if ((0, utils_1.isBt)(params.col) || (0, utils_1.isOo)(params.col)) {
            // fk_related_model_id is used to prevent paste operation in different fk_related_model_id cell
            return JSON.stringify({
                fk_related_model_id: params.col.colOptions
                    .fk_related_model_id,
                value: value || null,
            });
        }
        return value !== null && value !== void 0 ? value : '';
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = (0, utils_1.parseLinksValue)(value, params)) !== null && _a !== void 0 ? _a : '';
    }
}
exports.LinksHelper = LinksHelper;
