// nocodb-sdk/build/main/lib/columnHelper/columns/Rating.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RatingHelper = void 0;
const error_1 = require("../../../lib/error");
const helperFunctions_1 = require("../../../lib/helperFunctions");
const __1 = require("..");
const column_interface_1 = __importDefault(require("../column.interface"));
const is_1 = require("../../../lib/is");
class RatingHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            iconIdx: 0,
            icon: {
                full: 'mdi-star',
                empty: 'mdi-star-outline',
            },
            color: '#fcb401',
            max: 5,
        };
        // simply copy highlighted rows
        // since rating doesn't need increment
        // override populateFillHandle
    }
    serializeValue(value, params) {
        var _a;
        const res = (0, __1.serializeIntValue)(value !== null && value !== void 0 ? value : 0);
        if (res === null) {
            if (params.isMultipleCellPaste || params.serializeSearchQuery) {
                return null;
            }
            else {
                throw new error_1.SilentTypeConversionError();
            }
        }
        if (res) {
            if (params.serializeSearchQuery) {
                return res;
            }
            return Math.min(res, ((_a = (0, helperFunctions_1.parseProp)(params.col.meta)) === null || _a === void 0 ? void 0 : _a.max) || this.columnDefaultMeta.max);
        }
        return res;
    }
    parseValue(value) {
        return (0, __1.parseIntValue)(value);
    }
    parsePlainCellValue(value, params) {
        var _a;
        if (params.isAggregation && (0, is_1.ncIsNaN)(value)) {
            value = 0;
        }
        return `${(_a = (0, __1.parseIntValue)(value, params.col)) !== null && _a !== void 0 ? _a : ''}`;
    }
}
exports.RatingHelper = RatingHelper;
