// nocodb-sdk/build/main/lib/columnHelper/columns/SingleLineText.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SingleLineTextHelper = void 0;
const column_interface_1 = __importDefault(require("../column.interface"));
const utils_1 = require("../utils");
const fill_handler_1 = require("../utils/fill-handler");
class SingleLineTextHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {};
    }
    serializeValue(value, _params) {
        return (0, utils_1.serializeStringValue)(value);
    }
    parseValue(value, _params) {
        if (value === null || value === undefined) {
            return '';
        }
        return value.toString();
    }
    parsePlainCellValue(value, _params) {
        var _a;
        return (_a = value === null || value === void 0 ? void 0 : value.toString()) !== null && _a !== void 0 ? _a : '';
    }
    // using string number fill handler
    populateFillHandle(params) {
        return (0, fill_handler_1.populateFillHandleStringNumber)(params);
    }
}
exports.SingleLineTextHelper = SingleLineTextHelper;
