// nocodb-sdk/build/main/lib/columnHelper/columns/Time.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeHelper = void 0;
const dayjs_1 = __importDefault(require("dayjs"));
const dateTimeHelper_1 = require("../../../lib/dateTimeHelper");
const error_1 = require("../../../lib/error");
const column_interface_1 = __importDefault(require("../column.interface"));
const utils_1 = require("../utils");
class TimeHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            is12hrFormat: false,
        };
    }
    serializeValue(value, params) {
        if (params.serializeSearchQuery) {
            return this.parseValue(value, params);
        }
        value = (0, utils_1.serializeTimeValue)(value, params);
        if (value === null) {
            if (params.isMultipleCellPaste || params.serializeSearchQuery) {
                return null;
            }
            else {
                throw new error_1.SilentTypeConversionError();
            }
        }
        return value;
    }
    parseValue(value, params) {
        return (0, utils_1.parseTimeValue)(value, params);
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = (0, utils_1.parseTimeValue)(value, params)) !== null && _a !== void 0 ? _a : '';
    }
    equalityComparison(a, b, param) {
        const aDayjs = typeof a === 'string' ? (0, dayjs_1.default)((0, utils_1.serializeTimeValue)(a, param)) : (0, dayjs_1.default)(a);
        const bDayjs = typeof b === 'string' ? (0, dayjs_1.default)((0, utils_1.serializeTimeValue)(b, param)) : (0, dayjs_1.default)(b);
        return (aDayjs.format((0, dateTimeHelper_1.constructTimeFormat)(param.col)) ===
            bDayjs.format((0, dateTimeHelper_1.constructTimeFormat)(param.col)));
    }
}
exports.TimeHelper = TimeHelper;
