// nocodb-sdk/build/main/lib/columnHelper/columns/User.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserHelper = void 0;
const error_1 = require("../../../lib/error");
const helperFunctions_1 = require("../../../lib/helperFunctions");
const is_1 = require("../../../lib/is");
const column_interface_1 = __importDefault(require("../column.interface"));
const utils_1 = require("../utils");
class UserHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            is_multi: false,
            notify: false,
        };
    }
    serializeValue(value, params) {
        if (params.serializeSearchQuery) {
            return (0, utils_1.serializeStringValue)(value);
        }
        try {
            value = typeof value === 'string' ? JSON.parse(value) : value;
        }
        catch (_a) { }
        if (value === null) {
            if (params.isMultipleCellPaste) {
                return null;
            }
            else {
                throw new error_1.SilentTypeConversionError();
            }
        }
        const is_multi = (0, helperFunctions_1.parseProp)(params.col.meta).is_multi;
        if ((0, is_1.ncIsArray)(value) || (0, is_1.ncIsObject)(value)) {
            if ((0, is_1.ncIsObject)(value)) {
                value = [value];
            }
            value = value.reduce((acc, user) => {
                if (user.email) {
                    acc.push(user);
                }
                return acc;
            }, []);
            if (!value.length)
                return null;
            if (!is_multi) {
                return value[0];
            }
            return value;
        }
        if (!(0, is_1.ncIsString)(value))
            return null;
        if (is_multi) {
            return value
                .split(',')
                .map((val) => (0, utils_1.serializeEmail)(val))
                .filter(Boolean)
                .join(', ');
        }
        return (0, utils_1.serializeEmail)(value);
    }
    parseValue(value) {
        return (0, utils_1.parseUserValue)(value, true);
    }
    parsePlainCellValue(value) {
        var _a;
        return (_a = (0, utils_1.parseUserValue)(value)) !== null && _a !== void 0 ? _a : '';
    }
}
exports.UserHelper = UserHelper;
