// nocodb-sdk/build/module/lib/dashboard/index.js
import { ChartTypes } from './charts';
export var WidgetTypes;
(function (WidgetTypes) {
    WidgetTypes["CHART"] = "chart";
    WidgetTypes["METRIC"] = "metric";
    WidgetTypes["TEXT"] = "text";
    WidgetTypes["TABLE"] = "table";
    WidgetTypes["IFRAME"] = "iframe";
})(WidgetTypes || (WidgetTypes = {}));
export const WidgetChartLabelMap = {
    [ChartTypes.BAR]: 'Bar Chart',
    [ChartTypes.LINE]: 'Line Chart',
    [ChartTypes.PIE]: 'Pie Chart',
    [ChartTypes.DONUT]: 'Donut Chart',
    [ChartTypes.SCATTER]: 'Scatter Plot',
    [WidgetTypes.TABLE]: 'Table',
    [WidgetTypes.METRIC]: 'Metric',
    [WidgetTypes.TEXT]: 'Text',
    [WidgetTypes.IFRAME]: 'IFrame',
};
export var WidgetDataSourceTypes;
(function (WidgetDataSourceTypes) {
    WidgetDataSourceTypes["VIEW"] = "view";
    WidgetDataSourceTypes["MODEL"] = "model";
    WidgetDataSourceTypes["FILTER"] = "filter";
})(WidgetDataSourceTypes || (WidgetDataSourceTypes = {}));
export var TextWidgetTypes;
(function (TextWidgetTypes) {
    TextWidgetTypes["Markdown"] = "markdown";
    TextWidgetTypes["Text"] = "text";
})(TextWidgetTypes || (TextWidgetTypes = {}));
export * from './charts';
export * from './widgetUtils';
