// nocodb-sdk/build/main/lib/columnHelper/column-helper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColumnHelper = exports.ColumnHelperClass = void 0;
const is_1 = require("../is");
const UITypes_1 = __importDefault(require("../UITypes"));
const columns_1 = require("./columns");
const SingleSelect_1 = require("./columns/SingleSelect");
class ColumnHelperClass {
    constructor() {
        this.defautlHelper = 'defautlHelper';
        this.registry = {
            [this.defautlHelper]: columns_1.DefaultColumnHelper,
            [UITypes_1.default.SingleLineText]: columns_1.SingleLineTextHelper,
            [UITypes_1.default.LongText]: columns_1.LongTextHelper,
            [UITypes_1.default.Number]: columns_1.NumberHelper,
            [UITypes_1.default.Decimal]: columns_1.DecimalHelper,
            [UITypes_1.default.Currency]: columns_1.CurrencyHelper,
            [UITypes_1.default.Duration]: columns_1.DurationHelper,
            [UITypes_1.default.Percent]: columns_1.PercentHelper,
            [UITypes_1.default.Checkbox]: columns_1.CheckboxHelper,
            [UITypes_1.default.Rating]: columns_1.RatingHelper,
            [UITypes_1.default.PhoneNumber]: columns_1.PhoneNumberHelper,
            [UITypes_1.default.Email]: columns_1.EmailHelper,
            [UITypes_1.default.URL]: columns_1.UrlHelper,
            [UITypes_1.default.JSON]: columns_1.JsonHelper,
            [UITypes_1.default.GeoData]: columns_1.GeoDataHelper,
            [UITypes_1.default.Barcode]: columns_1.BarcodeHelper,
            [UITypes_1.default.QrCode]: columns_1.QrCodeHelper,
            [UITypes_1.default.Formula]: columns_1.FormulaHelper,
            [UITypes_1.default.Date]: columns_1.DateHelper,
            [UITypes_1.default.DateTime]: columns_1.DateTimeHelper,
            [UITypes_1.default.Time]: columns_1.TimeHelper,
            [UITypes_1.default.Year]: columns_1.YearHelper,
            [UITypes_1.default.CreatedTime]: columns_1.CreatedTimeHelper,
            [UITypes_1.default.LastModifiedTime]: columns_1.LastModifiedTimeHelper,
            [UITypes_1.default.SingleSelect]: SingleSelect_1.SingleSelectHelper,
            [UITypes_1.default.MultiSelect]: columns_1.MultiSelectHelper,
            [UITypes_1.default.User]: columns_1.UserHelper,
            [UITypes_1.default.Attachment]: columns_1.AttachmentHelper,
            [UITypes_1.default.CreatedBy]: columns_1.UserHelper,
            [UITypes_1.default.LastModifiedBy]: columns_1.UserHelper,
            [UITypes_1.default.Lookup]: columns_1.LookupHelper,
            [UITypes_1.default.Rollup]: columns_1.RollupHelper,
            [UITypes_1.default.Links]: columns_1.LinksHelper,
            [UITypes_1.default.LinkToAnotherRecord]: columns_1.LTARHelper,
            [UITypes_1.default.Button]: columns_1.ButtonHelper,
        };
    }
    // Method to retrieve the specific column class and instantiate it
    getColumn(params) {
        let ColumnClass;
        if ((0, is_1.ncIsString)(params)) {
            ColumnClass = this.registry[params] || this.registry[this.defautlHelper];
        }
        else {
            if (!params.col || !this.registry[params.col.uidt]) {
                ColumnClass = this.registry[this.defautlHelper];
            }
            else {
                ColumnClass = this.registry[params.col.uidt];
            }
        }
        // Instantiate the class with the column data
        return new ColumnClass();
    }
    getColumnDefaultMeta(uidt) {
        const columnInstance = this.getColumn(uidt);
        if (columnInstance) {
            return columnInstance.columnDefaultMeta || {};
        }
        return {};
    }
    /**
     * Parses a stored value back into its original form.
     * Converts a database-stored value into a display-friendly format.
     *
     * @example
     * // Example: Formatting percentage values
     * dbValue = 59; // Stored in DB
     * displayValue = "59%"; // Displayed to users
     *
     * @param value - The value to be parsed from storage format.
     * @param params - Additional parameters related to column parsing.
     * @returns The parsed value in a display-friendly format.
     */
    parseValue(value, params) {
        const columnInstance = this.getColumn(params);
        if (columnInstance) {
            return columnInstance.parseValue(value, params);
        }
    }
    /**
     * Serializes the given value based on column parameters.
     *
     * **WARNING:** This method **can throw errors**. Use a `try-catch` block when calling it.
     *
     * @param value - The value to be serialized.
     * @param params - Additional parameters related to column serialization.
     * @returns The serialized value.
     * @throws {Error} If serialization fails.
     */
    serializeValue(value, params) {
        const columnInstance = this.getColumn(params);
        if (columnInstance) {
            return columnInstance.serializeValue(value, params);
        }
    }
    /**
     * Parses a plain cell value back into its original form.
     * Converts a plain cell value into a display-friendly format.
     *
     * @param value - The value to be parsed from storage format.
     * @param params - Additional parameters related to column parsing.
     * @returns The parsed value in a display-friendly format.
     */
    parsePlainCellValue(value, params) {
        const columnInstance = this.getColumn(params);
        if (columnInstance) {
            return columnInstance.parsePlainCellValue(value, params);
        }
        return value;
    }
    populateFillHandle(params) {
        const columnInstance = this.getColumn({
            col: params.column,
        });
        if (columnInstance) {
            return columnInstance.populateFillHandle(params);
        }
        return undefined;
    }
}
exports.ColumnHelperClass = ColumnHelperClass;
exports.ColumnHelper = new ColumnHelperClass();
