// nocodb-sdk/build/main/lib/audit/v0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.auditOperationSubTypeLabels = exports.auditOperationTypeLabels = exports.AuditOperationSubTypes = exports.AuditOperationTypes = void 0;
var AuditOperationTypes;
(function (AuditOperationTypes) {
    AuditOperationTypes["COMMENT"] = "COMMENT";
    AuditOperationTypes["DATA"] = "DATA";
    AuditOperationTypes["PROJECT"] = "PROJECT";
    AuditOperationTypes["VIRTUAL_RELATION"] = "VIRTUAL_RELATION";
    AuditOperationTypes["RELATION"] = "RELATION";
    AuditOperationTypes["TABLE_VIEW"] = "TABLE_VIEW";
    AuditOperationTypes["TABLE"] = "TABLE";
    AuditOperationTypes["VIEW"] = "VIEW";
    AuditOperationTypes["META"] = "META";
    AuditOperationTypes["TABLE_COLUMN"] = "TABLE_COLUMN";
    AuditOperationTypes["WEBHOOKS"] = "WEBHOOKS";
    AuditOperationTypes["AUTHENTICATION"] = "AUTHENTICATION";
    AuditOperationTypes["ORG_USER"] = "ORG_USER";
})(AuditOperationTypes || (exports.AuditOperationTypes = AuditOperationTypes = {}));
var AuditOperationSubTypes;
(function (AuditOperationSubTypes) {
    AuditOperationSubTypes["INSERT"] = "INSERT";
    AuditOperationSubTypes["CREATE"] = "CREATE";
    AuditOperationSubTypes["UPDATE"] = "UPDATE";
    AuditOperationSubTypes["DELETE"] = "DELETE";
    AuditOperationSubTypes["BULK_INSERT"] = "BULK_INSERT";
    AuditOperationSubTypes["BULK_UPDATE"] = "BULK_UPDATE";
    AuditOperationSubTypes["BULK_DELETE"] = "BULK_DELETE";
    AuditOperationSubTypes["LINK_RECORD"] = "LINK_RECORD";
    AuditOperationSubTypes["UNLINK_RECORD"] = "UNLINK_RECORD";
    AuditOperationSubTypes["RENAME"] = "RENAME";
    AuditOperationSubTypes["IMPORT_FROM_ZIP"] = "IMPORT_FROM_ZIP";
    AuditOperationSubTypes["EXPORT_TO_FS"] = "EXPORT_TO_FS";
    AuditOperationSubTypes["EXPORT_TO_ZIP"] = "EXPORT_TO_ZIP";
    AuditOperationSubTypes["SIGNIN"] = "SIGNIN";
    AuditOperationSubTypes["SIGNUP"] = "SIGNUP";
    AuditOperationSubTypes["PASSWORD_RESET"] = "PASSWORD_RESET";
    AuditOperationSubTypes["PASSWORD_FORGOT"] = "PASSWORD_FORGOT";
    AuditOperationSubTypes["PASSWORD_CHANGE"] = "PASSWORD_CHANGE";
    AuditOperationSubTypes["EMAIL_VERIFICATION"] = "EMAIL_VERIFICATION";
    AuditOperationSubTypes["ROLES_MANAGEMENT"] = "ROLES_MANAGEMENT";
    AuditOperationSubTypes["INVITE"] = "INVITE";
    AuditOperationSubTypes["RESEND_INVITE"] = "RESEND_INVITE";
})(AuditOperationSubTypes || (exports.AuditOperationSubTypes = AuditOperationSubTypes = {}));
const auditOperationTypeLabels = {
    [AuditOperationTypes.COMMENT]: 'Comment',
    [AuditOperationTypes.DATA]: 'Data',
    [AuditOperationTypes.PROJECT]: 'Project',
    [AuditOperationTypes.VIRTUAL_RELATION]: 'Virtual Relation',
    [AuditOperationTypes.RELATION]: 'Relation',
    [AuditOperationTypes.TABLE_VIEW]: 'Table View',
    [AuditOperationTypes.TABLE]: 'Table',
    [AuditOperationTypes.VIEW]: 'View',
    [AuditOperationTypes.META]: 'Meta',
    [AuditOperationTypes.WEBHOOKS]: 'Webhooks',
    [AuditOperationTypes.AUTHENTICATION]: 'Authentication',
    [AuditOperationTypes.TABLE_COLUMN]: 'Table Column',
    [AuditOperationTypes.ORG_USER]: 'Org User',
};
exports.auditOperationTypeLabels = auditOperationTypeLabels;
const auditOperationSubTypeLabels = {
    [AuditOperationSubTypes.UPDATE]: 'Update',
    [AuditOperationSubTypes.INSERT]: 'Insert',
    [AuditOperationSubTypes.DELETE]: 'Delete',
    [AuditOperationSubTypes.BULK_INSERT]: 'Bulk Insert',
    [AuditOperationSubTypes.BULK_UPDATE]: 'Bulk Update',
    [AuditOperationSubTypes.BULK_DELETE]: 'Bulk Delete',
    [AuditOperationSubTypes.LINK_RECORD]: 'Link Record',
    [AuditOperationSubTypes.UNLINK_RECORD]: 'Unlink Record',
    [AuditOperationSubTypes.CREATE]: 'Create',
    [AuditOperationSubTypes.RENAME]: 'Rename',
    [AuditOperationSubTypes.IMPORT_FROM_ZIP]: 'Import From Zip',
    [AuditOperationSubTypes.EXPORT_TO_FS]: 'Export To FS',
    [AuditOperationSubTypes.EXPORT_TO_ZIP]: 'Export To Zip',
    [AuditOperationSubTypes.SIGNIN]: 'Signin',
    [AuditOperationSubTypes.SIGNUP]: 'Signup',
    [AuditOperationSubTypes.PASSWORD_RESET]: 'Password Reset',
    [AuditOperationSubTypes.PASSWORD_FORGOT]: 'Password Forgot',
    [AuditOperationSubTypes.PASSWORD_CHANGE]: 'Password Change',
    [AuditOperationSubTypes.EMAIL_VERIFICATION]: 'Email Verification',
    [AuditOperationSubTypes.ROLES_MANAGEMENT]: 'Roles Management',
    [AuditOperationSubTypes.INVITE]: 'Invite',
    [AuditOperationSubTypes.RESEND_INVITE]: 'Resend Invite',
};
exports.auditOperationSubTypeLabels = auditOperationSubTypeLabels;
