// nocodb-sdk/build/main/lib/audit/v1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuditV1OperationTypes = exports.auditV1OperationsCategory = exports.auditV1OperationTypesAlias = void 0;
exports.auditDescription = auditDescription;
var AuditV1OperationTypes;
(function (AuditV1OperationTypes) {
    AuditV1OperationTypes["USER_SIGNUP"] = "USER_SIGNUP";
    AuditV1OperationTypes["USER_SIGNIN"] = "USER_SIGNIN";
    AuditV1OperationTypes["USER_INVITE"] = "USER_INVITE";
    AuditV1OperationTypes["WORKSPACE_USER_INVITE"] = "WORKSPACE_USER_INVITE";
    AuditV1OperationTypes["WORKSPACE_USER_UPDATE"] = "WORKSPACE_USER_UPDATE";
    AuditV1OperationTypes["WORKSPACE_USER_DELETE"] = "WORKSPACE_USER_DELETE";
    AuditV1OperationTypes["USER_PASSWORD_CHANGE"] = "USER_PASSWORD_CHANGE";
    AuditV1OperationTypes["USER_PASSWORD_RESET"] = "USER_PASSWORD_RESET";
    AuditV1OperationTypes["USER_PASSWORD_FORGOT"] = "USER_PASSWORD_FORGOT";
    AuditV1OperationTypes["USER_EMAIL_VERIFY"] = "USER_EMAIL_VERIFY";
    AuditV1OperationTypes["BASE_USER_INVITE"] = "BASE_USER_INVITE";
    AuditV1OperationTypes["BASE_USER_UPDATE"] = "BASE_USER_UPDATE";
    AuditV1OperationTypes["BASE_USER_INVITE_RESEND"] = "BASE_USER_INVITE_RESEND";
    AuditV1OperationTypes["TABLE_CREATE"] = "TABLE_CREATE";
    AuditV1OperationTypes["TABLE_DELETE"] = "TABLE_DELETE";
    AuditV1OperationTypes["COLUMN_CREATE"] = "FIELD_CREATE";
    AuditV1OperationTypes["COLUMN_UPDATE"] = "FIELD_UPDATE";
    AuditV1OperationTypes["COLUMN_DELETE"] = "FIELD_DELETE";
    AuditV1OperationTypes["COLUMN_RENAME"] = "FIELD_RENAME";
    AuditV1OperationTypes["ORG_USER_INVITE"] = "ORG_USER_INVITE";
    AuditV1OperationTypes["ORG_USER_INVITE_RESEND"] = "ORG_USER_INVITE_RESEND";
    AuditV1OperationTypes["DATA_INSERT"] = "DATA_INSERT";
    AuditV1OperationTypes["DATA_UPDATE"] = "DATA_UPDATE";
    AuditV1OperationTypes["DATA_DELETE"] = "DATA_DELETE";
    AuditV1OperationTypes["DATA_BULK_INSERT"] = "DATA_BULK_INSERT";
    AuditV1OperationTypes["DATA_BULK_UPDATE"] = "DATA_BULK_UPDATE";
    AuditV1OperationTypes["DATA_BULK_DELETE"] = "DATA_BULK_DELETE";
    AuditV1OperationTypes["DATA_BULK_ALL_DELETE"] = "DATA_BULK_ALL_DELETE";
    AuditV1OperationTypes["DATA_BULK_ALL_UPDATE"] = "DATA_BULK_ALL_UPDATE";
    AuditV1OperationTypes["DATA_LINK"] = "DATA_LINK";
    AuditV1OperationTypes["DATA_UNLINK"] = "DATA_UNLINK";
    AuditV1OperationTypes["WORKSPACE_CREATE"] = "WORKSPACE_CREATE";
    AuditV1OperationTypes["WORKSPACE_UPDATE"] = "WORKSPACE_UPDATE";
    AuditV1OperationTypes["WORKSPACE_DELETE"] = "WORKSPACE_DELETE";
    AuditV1OperationTypes["WORKSPACE_RENAME"] = "WORKSPACE_RENAME";
    AuditV1OperationTypes["BASE_CREATE"] = "BASE_CREATE";
    AuditV1OperationTypes["BASE_UPDATE"] = "BASE_UPDATE";
    AuditV1OperationTypes["BASE_DELETE"] = "BASE_DELETE";
    AuditV1OperationTypes["BASE_RENAME"] = "BASE_RENAME";
    AuditV1OperationTypes["VIEW_CREATE"] = "VIEW_CREATE";
    AuditV1OperationTypes["VIEW_UPDATE"] = "VIEW_UPDATE";
    AuditV1OperationTypes["VIEW_DELETE"] = "VIEW_DELETE";
    AuditV1OperationTypes["VIEW_RENAME"] = "VIEW_RENAME";
    AuditV1OperationTypes["HOOK_CREATE"] = "HOOK_CREATE";
    AuditV1OperationTypes["HOOK_UPDATE"] = "HOOK_UPDATE";
    AuditV1OperationTypes["HOOK_DELETE"] = "HOOK_DELETE";
    AuditV1OperationTypes["HOOK_FILTER_CREATE"] = "HOOK_FILTER_CREATE";
    AuditV1OperationTypes["HOOK_FILTER_UPDATE"] = "HOOK_FILTER_UPDATE";
    AuditV1OperationTypes["HOOK_FILTER_DELETE"] = "HOOK_FILTER_DELETE";
    AuditV1OperationTypes["VIEW_FILTER_CREATE"] = "VIEW_FILTER_CREATE";
    AuditV1OperationTypes["VIEW_FILTER_UPDATE"] = "VIEW_FILTER_UPDATE";
    AuditV1OperationTypes["VIEW_FILTER_DELETE"] = "VIEW_FILTER_DELETE";
    AuditV1OperationTypes["LINK_FILTER_CREATE"] = "LINK_FILTER_CREATE";
    AuditV1OperationTypes["LINK_FILTER_UPDATE"] = "LINK_FILTER_UPDATE";
    AuditV1OperationTypes["LINK_FILTER_DELETE"] = "LINK_FILTER_DELETE";
    AuditV1OperationTypes["WIDGET_FILTER_CREATE"] = "WIDGET_FILTER_CREATE";
    AuditV1OperationTypes["WIDGET_FILTER_UPDATE"] = "WIDGET_FILTER_UPDATE";
    AuditV1OperationTypes["WIDGET_FILTER_DELETE"] = "WIDGET_FILTER_DELETE";
    AuditV1OperationTypes["VIEW_SORT_CREATE"] = "VIEW_SORT_CREATE";
    AuditV1OperationTypes["VIEW_SORT_UPDATE"] = "VIEW_SORT_UPDATE";
    AuditV1OperationTypes["VIEW_SORT_DELETE"] = "VIEW_SORT_DELETE";
    AuditV1OperationTypes["SHARED_BASE_CREATE"] = "SHARED_BASE_CREATE";
    AuditV1OperationTypes["SHARED_BASE_DELETE"] = "SHARED_BASE_DELETE";
    AuditV1OperationTypes["SOURCE_CREATE"] = "SOURCE_CREATE";
    AuditV1OperationTypes["SOURCE_UPDATE"] = "SOURCE_UPDATE";
    AuditV1OperationTypes["SOURCE_DELETE"] = "SOURCE_DELETE";
    AuditV1OperationTypes["SOURCE_RENAME"] = "SOURCE_RENAME";
    AuditV1OperationTypes["SHARED_VIEW_DELETE"] = "SHARED_VIEW_DELETE";
    AuditV1OperationTypes["SHARED_VIEW_CREATE"] = "SHARED_VIEW_CREATE";
    AuditV1OperationTypes["SHARED_VIEW_UPDATE"] = "SHARED_VIEW_UPDATE";
    AuditV1OperationTypes["API_TOKEN_DELETE"] = "API_TOKEN_DELETE";
    AuditV1OperationTypes["API_TOKEN_CREATE"] = "API_TOKEN_CREATE";
    AuditV1OperationTypes["BASE_DUPLICATE"] = "BASE_DUPLICATE";
    AuditV1OperationTypes["BASE_DUPLICATE_ERROR"] = "BASE_DUPLICATE_ERROR";
    AuditV1OperationTypes["TABLE_DUPLICATE"] = "TABLE_DUPLICATE";
    AuditV1OperationTypes["TABLE_DUPLICATE_ERROR"] = "TABLE_DUPLICATE_ERROR";
    AuditV1OperationTypes["COLUMN_DUPLICATE"] = "FIELD_DUPLICATE";
    AuditV1OperationTypes["COLUMN_DUPLICATE_ERROR"] = "FIELD_DUPLICATE_ERROR";
    AuditV1OperationTypes["VIEW_DUPLICATE"] = "VIEW_DUPLICATE";
    AuditV1OperationTypes["VIEW_DUPLICATE_ERROR"] = "VIEW_DUPLICATE_ERROR";
    AuditV1OperationTypes["FORM_COLUMN_UPDATE"] = "FORM_FIELD_UPDATE";
    AuditV1OperationTypes["USER_SIGNOUT"] = "USER_SIGNOUT";
    AuditV1OperationTypes["TABLE_UPDATE"] = "TABLE_UPDATE";
    AuditV1OperationTypes["TABLE_RENAME"] = "TABLE_RENAME";
    AuditV1OperationTypes["VIEW_COLUMN_UPDATE"] = "VIEW_FIELD_UPDATE";
    AuditV1OperationTypes["UI_ACL"] = "UI_ACL";
    AuditV1OperationTypes["AIRTABLE_IMPORT"] = "AIRTABLE_IMPORT";
    AuditV1OperationTypes["AIRTABLE_IMPORT_ERROR"] = "AIRTABLE_IMPORT_ERROR";
    AuditV1OperationTypes["INTEGRATION_CREATE"] = "INTEGRATION_CREATE";
    AuditV1OperationTypes["INTEGRATION_UPDATE"] = "INTEGRATION_UPDATE";
    AuditV1OperationTypes["INTEGRATION_DELETE"] = "INTEGRATION_DELETE";
    AuditV1OperationTypes["SNAPSHOT_DELETE"] = "SNAPSHOT_DELETE";
    AuditV1OperationTypes["SNAPSHOT_CREATE"] = "SNAPSHOT_CREATE";
    AuditV1OperationTypes["SNAPSHOT_RESTORE"] = "SNAPSHOT_RESTORE";
    AuditV1OperationTypes["DATA_IMPORT"] = "DATA_IMPORT";
    AuditV1OperationTypes["DATA_EXPORT"] = "DATA_EXPORT";
    AuditV1OperationTypes["USER_PROFILE_UPDATE"] = "USER_PROFILE_UPDATE";
    AuditV1OperationTypes["SCRIPT_CREATE"] = "SCRIPT_CREATE";
    AuditV1OperationTypes["SCRIPT_UPDATE"] = "SCRIPT_UPDATE";
    AuditV1OperationTypes["SCRIPT_DELETE"] = "SCRIPT_DELETE";
    AuditV1OperationTypes["SCRIPT_DUPLICATE"] = "SCRIPT_DUPLICATE";
    AuditV1OperationTypes["DASHBOARD_CREATE"] = "DASHBOARD_CREATE";
    AuditV1OperationTypes["DASHBOARD_UPDATE"] = "DASHBOARD_UPDATE";
    AuditV1OperationTypes["DASHBOARD_DELETE"] = "DASHBOARD_DELETE";
    AuditV1OperationTypes["DASHBOARD_DUPLICATE"] = "DASHBOARD_DUPLICATE";
    AuditV1OperationTypes["DASHBOARD_DUPLICATE_ERROR"] = "DASHBOARD_DUPLICATE_ERROR";
    AuditV1OperationTypes["SHARED_DASHBOARD_CREATE"] = "SHARED_DASHBOARD_CREATE";
    AuditV1OperationTypes["SHARED_DASHBOARD_DELETE"] = "SHARED_DASHBOARD_DELETE";
    AuditV1OperationTypes["SHARED_DASHBOARD_UPDATE"] = "SHARED_DASHBOARD_UPDATE";
    AuditV1OperationTypes["WIDGET_CREATE"] = "WIDGET_CREATE";
    AuditV1OperationTypes["WIDGET_UPDATE"] = "WIDGET_UPDATE";
    AuditV1OperationTypes["WIDGET_DELETE"] = "WIDGET_DELETE";
    AuditV1OperationTypes["WIDGET_DUPLICATE"] = "WIDGET_DUPLICATE";
    AuditV1OperationTypes["PERMISSION_CREATE"] = "PERMISSION_CREATE";
    AuditV1OperationTypes["PERMISSION_UPDATE"] = "PERMISSION_UPDATE";
    AuditV1OperationTypes["PERMISSION_DELETE"] = "PERMISSION_DELETE";
})(AuditV1OperationTypes || (exports.AuditV1OperationTypes = AuditV1OperationTypes = {}));
exports.auditV1OperationTypesAlias = Object.values(AuditV1OperationTypes).reduce((acc, key) => {
    // Convert snake_case or UPPER_SNAKE_CASE to readable format
    // Capitalize each word
    acc[key] = key
        .replace(/_/g, ' ') // Replace underscores with spaces
        .toLowerCase() // Convert to lowercase
        .replace(/\b[a-z]/g, (char) => char.toUpperCase());
    return acc;
}, {});
/**
 * For audit logs table filter usecase only
 */
exports.auditV1OperationsCategory = {
    DATA: {
        label: 'general.data',
        value: 'DATA',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('DATA_') || key.startsWith('AIRTABLE_')),
    },
    FIELD: {
        label: 'objects.column',
        value: 'FIELD',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('FIELD_') ||
            key.startsWith('FORM_COLUMN_') ||
            key.startsWith('VIEW_COLUMN_') ||
            key.startsWith('LINK_FILTER_')),
    },
    VIEW: {
        label: 'objects.view',
        value: 'VIEW',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('VIEW_') && !key.startsWith('VIEW_COLUMN_')),
    },
    TABLE: {
        label: 'objects.table',
        value: 'TABLE',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('TABLE_')),
    },
    BASE: {
        label: 'objects.project',
        value: 'BASE',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('BASE_') ||
            key.startsWith('SNAPSHOT_') ||
            key.startsWith('UI_ACL')),
    },
    SOURCE: {
        label: 'general.source',
        value: 'SOURCE',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('SOURCE_')),
    },
    SHARED: {
        label: 'general.shared',
        value: 'SHARED',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('SHARED_')),
    },
    USER: {
        label: 'objects.user',
        value: 'USER',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('USER_')),
    },
    WORKSPACE: {
        label: 'objects.workspace',
        value: 'WORKSPACE',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('WORKSPACE_')),
    },
    HOOK: {
        label: 'objects.webhook',
        value: 'HOOK',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('HOOK_')),
    },
    INTEGRATION: {
        label: 'general.integration',
        value: 'INTEGRATION',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('INTEGRATION_')),
    },
    API: {
        label: 'title.apiTokens',
        value: 'API',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('API_')),
    },
    ORG: {
        label: 'general.organization',
        value: 'ORG',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('ORG_')),
    },
    SCRIPT: {
        label: 'general.script',
        value: 'SCRIPT',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('SCRIPT_')),
    },
    DASHBOARD: {
        label: 'objects.dashboard',
        value: 'DASHBOARD',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('DASHBOARD_')),
    },
    WIDGET: {
        label: 'objects.widget',
        value: 'WIDGET',
        types: Object.values(AuditV1OperationTypes).filter((key) => key.startsWith('WIDGET_')),
    },
};
const descriptionTemplates = {
    [AuditV1OperationTypes.USER_SIGNUP]: (audit) => `User '${audit.user}' signed up`,
    [AuditV1OperationTypes.USER_SIGNIN]: (audit) => `User '${audit.user}' signed in`,
    [AuditV1OperationTypes.USER_INVITE]: (audit) => `User '${audit.user}' invited '${audit.details.user_email}'`,
    [AuditV1OperationTypes.USER_PASSWORD_CHANGE]: (audit) => `User '${audit.user}' changed password`,
    [AuditV1OperationTypes.USER_PASSWORD_RESET]: (audit) => `User '${audit.user}' reset password`,
    [AuditV1OperationTypes.USER_PASSWORD_FORGOT]: (audit) => `User '${audit.user}' forgot password`,
    [AuditV1OperationTypes.USER_EMAIL_VERIFY]: (audit) => `User '${audit.user}' verified email`,
    [AuditV1OperationTypes.BASE_USER_INVITE]: (audit) => `User '${audit.user}' invited '${audit.details.user_email}' to base`,
    [AuditV1OperationTypes.BASE_USER_INVITE_RESEND]: (audit) => `User '${audit.user}' resent invite to '${audit.details.user_email}'`,
    [AuditV1OperationTypes.BASE_USER_UPDATE]: (audit) => `User '${audit.user}' updated role of '${audit.details.user_email}'`,
    [AuditV1OperationTypes.TABLE_CREATE]: (audit) => `User '${audit.user}' created table '${audit.details.table_title}'`,
    [AuditV1OperationTypes.TABLE_DELETE]: (audit) => `User '${audit.user}' deleted table '${audit.details.table_title}'`,
    [AuditV1OperationTypes.ORG_USER_INVITE]: (audit) => `User '${audit.user}' invited '${audit.details.email}' to organization`,
    [AuditV1OperationTypes.ORG_USER_INVITE_RESEND]: (audit) => `User '${audit.user}' resent invite to '${audit.details.email}'`,
    [AuditV1OperationTypes.DATA_INSERT]: (audit) => `Record with ID [${audit.row_id}] has been inserted`,
    [AuditV1OperationTypes.DATA_UPDATE]: (audit) => `Record with ID [${audit.row_id}] has been updated`,
    [AuditV1OperationTypes.DATA_DELETE]: (audit) => `Record with ID [${audit.row_id}] has been deleted`,
    /*  [AuditV1OperationTypes.DATA_BULK_INSERT]: (
      audit: AuditV1<DataBulkInsertPayload>
    ) =>
      `${audit.details.data.length} ${
        audit.details.data.length > 1 ? 'records have been' : 'record has been'
      } inserted`,
    [AuditV1OperationTypes.DATA_BULK_UPDATE]: (
      audit: AuditV1<DataBulkUpdatePayload>
    ) =>
      `${audit.details.data.length} ${
        audit.details.data.length > 1 ? 'records have been' : 'record has been'
      } updated`,
    [AuditV1OperationTypes.DATA_BULK_DELETE]: (
      audit: AuditV1<DataBulkDeletePayload>
    ) =>
      `${audit.details.data.length} ${
        audit.details.data.length > 1 ? 'records have been' : 'record has been'
      } deleted`,*/
    [AuditV1OperationTypes.DATA_LINK]: (audit) => `Record [id:${audit.details.ref_row_id}] has been linked with record [id:${audit.details.row_id}] in ${audit.details.table_title}`,
    [AuditV1OperationTypes.DATA_UNLINK]: (audit) => `Record [id:${audit.details.ref_row_id}] has been unlinked from record [id:${audit.details.row_id}] in ${audit.details.table_title}`,
    [AuditV1OperationTypes.WORKSPACE_CREATE]: (audit) => `Workspace '${audit.details.workspace_title}' has been created`,
    [AuditV1OperationTypes.WORKSPACE_UPDATE]: (audit) => `Workspace '${audit.details.workspace_title}' has been updated`,
    [AuditV1OperationTypes.WORKSPACE_DELETE]: (audit) => `Workspace '${audit.details.workspace_title}' has been deleted`,
    [AuditV1OperationTypes.WORKSPACE_RENAME]: (audit) => `Workspace '${audit.details.old_workspace_title}' has been renamed to '${audit.details.new_workspace_title}'`,
    [AuditV1OperationTypes.BASE_CREATE]: (audit) => `Base '${audit.details.base_title}' has been created`,
    [AuditV1OperationTypes.BASE_UPDATE]: (audit) => `Base '${audit.details.base_title}' has been updated`,
    [AuditV1OperationTypes.BASE_DELETE]: (audit) => `Base '${audit.details.base_title}' has been deleted`,
    [AuditV1OperationTypes.BASE_RENAME]: (audit) => `Base '${audit.details.old_base_title}' has been renamed to '${audit.details.new_base_title}'`,
    [AuditV1OperationTypes.BASE_DUPLICATE]: (audit) => `Base '${audit.details.source_base_title}' has been duplicated`,
    [AuditV1OperationTypes.VIEW_CREATE]: (audit) => `${audit.details.view_type.replace(/^\w/, (m) => m.toUpperCase())} '${audit.details.view_title}' has been created`,
    [AuditV1OperationTypes.VIEW_UPDATE]: (audit) => `${audit.details.view_type.replace(/^\w/, (m) => m.toUpperCase())} '${audit.details.view_title}' has been updated`,
    [AuditV1OperationTypes.VIEW_DELETE]: (audit) => `${audit.details.view_type.replace(/^\w/, (m) => m.toUpperCase())} '${audit.details.view_title}' has been deleted`,
    [AuditV1OperationTypes.VIEW_RENAME]: (audit) => `${audit.details.view_type.replace(/^\w/, (m) => m.toUpperCase())} '${audit.details.old_view_title}' has been renamed to '${audit.details.new_view_title}'`,
    [AuditV1OperationTypes.VIEW_DUPLICATE]: (m) => `${m.details.view_type.replace(/^\w/, (m) => m.toUpperCase())} '${m.details.source_view_title}' has been duplicated`,
    [AuditV1OperationTypes.COLUMN_CREATE]: (audit) => `Field '${audit.details.view_title}' has been created`,
    [AuditV1OperationTypes.COLUMN_UPDATE]: (audit) => `Field '${audit.details.view_title}' has been updated`,
    [AuditV1OperationTypes.COLUMN_DELETE]: (audit) => `Field '${audit.details.view_title}' has been deleted`,
    [AuditV1OperationTypes.COLUMN_RENAME]: (audit) => `Field '${audit.details.old_view_title}' has been renamed to '${audit.details.new_view_title}'`,
    [AuditV1OperationTypes.COLUMN_DUPLICATE]: (audit) => `Field '${audit.details.source_field_title}' has been duplicated`,
    [AuditV1OperationTypes.HOOK_CREATE]: (audit) => `Hook '${audit.details.hook_title}' has been created`,
    [AuditV1OperationTypes.HOOK_UPDATE]: (audit) => `Hook '${audit.details.hook_title}' has been updated`,
    [AuditV1OperationTypes.HOOK_DELETE]: (audit) => `Hook '${audit.details.hook_title}' has been deleted`,
    [AuditV1OperationTypes.HOOK_FILTER_CREATE]: (audit) => `Filter with column '${audit.details.filter_field_id}' and operation '${audit.details.filter_comparison_op}' has been created`,
    [AuditV1OperationTypes.SCRIPT_CREATE]: (audit) => `Script '${audit.details.script_title}' has been created`,
    [AuditV1OperationTypes.SCRIPT_UPDATE]: (audit) => `Script '${audit.details.script_title}' has been updated`,
    [AuditV1OperationTypes.SCRIPT_DELETE]: (audit) => `Script '${audit.details.script_title}' has been deleted`,
    [AuditV1OperationTypes.SCRIPT_DUPLICATE]: (audit) => `Script '${audit.details.source_script_title}' has been duplicated`,
    [AuditV1OperationTypes.DASHBOARD_CREATE]: (audit) => `Dashboard '${audit.details.dashboard_title}' has been created`,
    [AuditV1OperationTypes.DASHBOARD_UPDATE]: (audit) => `Dashboard '${audit.details.dashboard_title}' has been updated`,
    [AuditV1OperationTypes.DASHBOARD_DELETE]: (audit) => `Dashboard '${audit.details.dashboard_title}' has been deleted`,
    [AuditV1OperationTypes.DASHBOARD_DUPLICATE]: (audit) => `Dashboard '${audit.details.source_dashboard_title}' has been duplicated`,
    [AuditV1OperationTypes.SHARED_DASHBOARD_CREATE]: (audit) => `Shared dashboard '${audit.details.dashboard_title}' has been created`,
    [AuditV1OperationTypes.SHARED_DASHBOARD_DELETE]: (audit) => `Shared dashboard '${audit.details.dashboard_title}' has been deleted`,
    [AuditV1OperationTypes.SHARED_DASHBOARD_UPDATE]: (audit) => `Shared dashboard '${audit.details.dashboard_title}' has been updated`,
    [AuditV1OperationTypes.WIDGET_CREATE]: (audit) => `Widget '${audit.details.widget_title}' has been created`,
    [AuditV1OperationTypes.WIDGET_UPDATE]: (audit) => `Widget '${audit.details.widget_title}' has been updated`,
    [AuditV1OperationTypes.WIDGET_DELETE]: (audit) => `Widget '${audit.details.widget_title}' has been deleted`,
    [AuditV1OperationTypes.WIDGET_DUPLICATE]: (audit) => `Widget '${audit.details.duplicated_widget_title}' has been duplicated`,
    [AuditV1OperationTypes.PERMISSION_CREATE]: (audit) => `Permission '${audit.details.permission}' has been created for entity '${audit.details.entity}' with ID '${audit.details.entity_id}'`,
    [AuditV1OperationTypes.PERMISSION_UPDATE]: (audit) => `Permission '${audit.details.permission}' has been updated for entity '${audit.details.entity}' with ID '${audit.details.entity_id}'`,
    [AuditV1OperationTypes.PERMISSION_DELETE]: (audit) => `Permission '${audit.details.permission}' has been deleted for entity '${audit.details.entity}' with ID '${audit.details.entity_id}'`,
};
function auditDescription(audit) {
    return descriptionTemplates[audit.op_type](audit);
}
