// nocodb-sdk/build/main/lib/aggregationHelper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatAggregation = exports.AllAggregations = exports.AttachmentAggregations = exports.DateAggregations = exports.BooleanAggregations = exports.CommonAggregations = exports.NumericalAggregations = exports.getAvailableAggregations = void 0;
const UITypes_1 = __importDefault(require("./UITypes"));
const formulaHelpers_1 = require("../lib/formulaHelpers");
const dateTimeHelper_1 = require("../lib/dateTimeHelper");
const helperFunctions_1 = require("../lib/helperFunctions");
const durationUtils_1 = require("../lib/durationUtils");
const numberUtils_1 = require("../lib/numberUtils");
const columnHelper_1 = require("./columnHelper");
var NumericalAggregations;
(function (NumericalAggregations) {
    NumericalAggregations["Sum"] = "sum";
    NumericalAggregations["Min"] = "min";
    NumericalAggregations["Max"] = "max";
    NumericalAggregations["Avg"] = "avg";
    NumericalAggregations["Median"] = "median";
    NumericalAggregations["StandardDeviation"] = "std_dev";
    //   Histogram = 'histogram',
    NumericalAggregations["Range"] = "range";
})(NumericalAggregations || (exports.NumericalAggregations = NumericalAggregations = {}));
var CommonAggregations;
(function (CommonAggregations) {
    CommonAggregations["Count"] = "count";
    CommonAggregations["CountEmpty"] = "count_empty";
    CommonAggregations["CountFilled"] = "count_filled";
    CommonAggregations["CountUnique"] = "count_unique";
    CommonAggregations["PercentEmpty"] = "percent_empty";
    CommonAggregations["PercentFilled"] = "percent_filled";
    CommonAggregations["PercentUnique"] = "percent_unique";
    CommonAggregations["None"] = "none";
})(CommonAggregations || (exports.CommonAggregations = CommonAggregations = {}));
var AttachmentAggregations;
(function (AttachmentAggregations) {
    AttachmentAggregations["AttachmentSize"] = "attachment_size";
})(AttachmentAggregations || (exports.AttachmentAggregations = AttachmentAggregations = {}));
var BooleanAggregations;
(function (BooleanAggregations) {
    BooleanAggregations["Checked"] = "checked";
    BooleanAggregations["Unchecked"] = "unchecked";
    BooleanAggregations["PercentChecked"] = "percent_checked";
    BooleanAggregations["PercentUnchecked"] = "percent_unchecked";
})(BooleanAggregations || (exports.BooleanAggregations = BooleanAggregations = {}));
var DateAggregations;
(function (DateAggregations) {
    DateAggregations["EarliestDate"] = "earliest_date";
    DateAggregations["LatestDate"] = "latest_date";
    DateAggregations["DateRange"] = "date_range";
    DateAggregations["MonthRange"] = "month_range";
})(DateAggregations || (exports.DateAggregations = DateAggregations = {}));
const AllAggregations = Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, CommonAggregations), NumericalAggregations), AttachmentAggregations), BooleanAggregations), DateAggregations);
exports.AllAggregations = AllAggregations;
const getAvailableAggregations = (type, parsed_tree) => {
    let returnAggregations = [];
    if (type === UITypes_1.default.Formula && (parsed_tree === null || parsed_tree === void 0 ? void 0 : parsed_tree.dataType)) {
        switch (parsed_tree.dataType) {
            case formulaHelpers_1.FormulaDataTypes.BOOLEAN:
                returnAggregations = [
                    ...Object.values(BooleanAggregations),
                    CommonAggregations.None,
                ];
                break;
            case formulaHelpers_1.FormulaDataTypes.DATE:
                returnAggregations = [
                    ...Object.values(DateAggregations),
                    ...Object.values(CommonAggregations),
                ];
                break;
            case formulaHelpers_1.FormulaDataTypes.NUMERIC:
                returnAggregations = [
                    ...Object.values(NumericalAggregations),
                    ...Object.values(CommonAggregations),
                ];
                break;
            default:
                returnAggregations = [...Object.values(CommonAggregations)];
                break;
        }
    }
    switch (type) {
        case UITypes_1.default.Number:
        case UITypes_1.default.Decimal:
        case UITypes_1.default.Currency:
        case UITypes_1.default.Percent:
        case UITypes_1.default.Duration:
        case UITypes_1.default.Rating:
        case UITypes_1.default.Rollup:
        case UITypes_1.default.Links:
            returnAggregations = [
                ...Object.values(NumericalAggregations),
                ...Object.values(CommonAggregations),
            ];
            break;
        case UITypes_1.default.Checkbox:
            returnAggregations = [
                ...Object.values(BooleanAggregations),
                CommonAggregations.None,
            ];
            break;
        case UITypes_1.default.Date:
        case UITypes_1.default.DateTime:
        case UITypes_1.default.LastModifiedTime:
        case UITypes_1.default.CreatedTime:
            returnAggregations = [
                ...Object.values(DateAggregations),
                ...Object.values(CommonAggregations),
            ];
            break;
        case UITypes_1.default.SpecificDBType:
        case UITypes_1.default.ForeignKey:
            returnAggregations = [CommonAggregations.None];
            break;
        case UITypes_1.default.Button:
        case UITypes_1.default.Attachment:
            return [CommonAggregations.None, AttachmentAggregations.AttachmentSize];
    }
    if (!returnAggregations.length) {
        returnAggregations = [...Object.values(CommonAggregations)];
    }
    return returnAggregations.filter((item) => item !== CommonAggregations.Count);
};
exports.getAvailableAggregations = getAvailableAggregations;
const formatAggregation = (aggregation, value, column, 
/**
 * If columnHelperParams is provided then it will be used to format the aggregation value
 */
columnHelperParams) => {
    var _a, _b, _c, _d, _e, _f;
    if (column.uidt === UITypes_1.default.Formula) {
        if ((_a = column === null || column === void 0 ? void 0 : column.meta) === null || _a === void 0 ? void 0 : _a.display_type) {
            const childColumn = Object.assign({ uidt: (_b = column === null || column === void 0 ? void 0 : column.meta) === null || _b === void 0 ? void 0 : _b.display_type }, (_c = column === null || column === void 0 ? void 0 : column.meta) === null || _c === void 0 ? void 0 : _c.display_column_meta);
            return formatAggregation(aggregation, value, childColumn, Object.assign(Object.assign({}, columnHelperParams), { col: childColumn }));
        }
    }
    if ([DateAggregations.EarliestDate, DateAggregations.LatestDate].includes(aggregation)) {
        if (column.uidt === UITypes_1.default.DateTime) {
            return (0, dateTimeHelper_1.getDateTimeValue)(value, column);
        }
        else if (column.uidt === UITypes_1.default.Date) {
            return (0, dateTimeHelper_1.getDateValue)(value, column);
        }
        return (0, dateTimeHelper_1.getDateTimeValue)(value, column);
    }
    if ([
        CommonAggregations.PercentEmpty,
        CommonAggregations.PercentFilled,
        CommonAggregations.PercentUnique,
        BooleanAggregations.PercentChecked,
        BooleanAggregations.PercentUnchecked,
    ].includes(aggregation)) {
        return `${(_d = (0, numberUtils_1.roundTo)(value, 1)) !== null && _d !== void 0 ? _d : 0}%`;
    }
    if ([DateAggregations.MonthRange, DateAggregations.DateRange].includes(aggregation)) {
        return aggregation === DateAggregations.DateRange
            ? `${value !== null && value !== void 0 ? value : 0} days`
            : `${value !== null && value !== void 0 ? value : 0} months`;
    }
    if ([
        CommonAggregations.Count,
        CommonAggregations.CountEmpty,
        CommonAggregations.CountFilled,
        CommonAggregations.CountUnique,
    ].includes(aggregation)) {
        return value;
    }
    if ([AttachmentAggregations.AttachmentSize].includes(aggregation)) {
        return (0, helperFunctions_1.formatBytes)(value !== null && value !== void 0 ? value : 0);
    }
    if (columnHelperParams &&
        (columnHelperParams === null || columnHelperParams === void 0 ? void 0 : columnHelperParams.col) &&
        Object.values(NumericalAggregations).includes(aggregation) &&
        [
            UITypes_1.default.Number,
            UITypes_1.default.Decimal,
            UITypes_1.default.Currency,
            UITypes_1.default.Percent,
            UITypes_1.default.Duration,
            UITypes_1.default.Rollup,
            UITypes_1.default.Time,
        ].includes(column.uidt)) {
        return columnHelper_1.ColumnHelper.parsePlainCellValue(value, Object.assign(Object.assign({}, columnHelperParams), { isAggregation: true }));
    }
    if (column.uidt === UITypes_1.default.Currency) {
        return (0, numberUtils_1.getCurrencyValue)(value, column);
    }
    if (column.uidt === UITypes_1.default.Percent) {
        return `${(0, numberUtils_1.roundTo)(value, 1)}%`;
    }
    if (column.uidt === UITypes_1.default.Duration) {
        return (0, durationUtils_1.convertMS2Duration)(value, ((_e = (0, helperFunctions_1.parseProp)(column.meta)) === null || _e === void 0 ? void 0 : _e.duration) || 0);
    }
    if (typeof value === 'number') {
        return (_f = (0, numberUtils_1.roundTo)(value, 1)) !== null && _f !== void 0 ? _f : '∞';
    }
    return value;
};
exports.formatAggregation = formatAggregation;
