// nocodb-sdk/build/main/lib/dateTimeHelper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDateValue = exports.getDateTimeValue = exports.isValidTimeFormat = exports.timeAgo = exports.handleTZ = exports.isDateMonthFormat = exports.dateFormats = exports.timeFormats = exports.dateMonthFormats = void 0;
exports.validateDateWithUnknownFormat = validateDateWithUnknownFormat;
exports.getDateFormat = getDateFormat;
exports.getDateTimeFormat = getDateTimeFormat;
exports.parseStringDate = parseStringDate;
exports.parseStringDateTime = parseStringDateTime;
exports.convertToTargetFormat = convertToTargetFormat;
exports.validateDateFormat = validateDateFormat;
exports.constructDateTimeFormat = constructDateTimeFormat;
exports.constructDateFormat = constructDateFormat;
exports.constructTimeFormat = constructTimeFormat;
exports.workerWithTimezone = workerWithTimezone;
const dayjs_1 = __importDefault(require("dayjs"));
const relativeTime_js_1 = __importDefault(require("dayjs/plugin/relativeTime.js"));
const customParseFormat_js_1 = __importDefault(require("dayjs/plugin/customParseFormat.js"));
const duration_js_1 = __importDefault(require("dayjs/plugin/duration.js"));
const utc_js_1 = __importDefault(require("dayjs/plugin/utc.js"));
const weekday_js_1 = __importDefault(require("dayjs/plugin/weekday.js"));
const timezone_js_1 = __importDefault(require("dayjs/plugin/timezone.js"));
const helperFunctions_1 = require("./helperFunctions");
const is_1 = require("./is");
dayjs_1.default.extend(utc_js_1.default);
dayjs_1.default.extend(relativeTime_js_1.default);
dayjs_1.default.extend(customParseFormat_js_1.default);
dayjs_1.default.extend(duration_js_1.default);
dayjs_1.default.extend(weekday_js_1.default);
dayjs_1.default.extend(timezone_js_1.default);
exports.dateMonthFormats = ['YYYY-MM', 'YYYY MM'];
exports.timeFormats = ['HH:mm', 'HH:mm:ss', 'HH:mm:ss.SSS'];
exports.dateFormats = [
    'YYYY-MM-DD',
    'YYYY/MM/DD',
    'DD-MM-YYYY',
    'MM-DD-YYYY',
    'DD/MM/YYYY',
    'MM/DD/YYYY',
    'DD MM YYYY',
    'MM DD YYYY',
    'YYYY MM DD',
    ///added 2 new format#9652
    'DD MMM YYYY',
    'DD MMM YY',
];
const isDateMonthFormat = (format) => exports.dateMonthFormats.includes(format);
exports.isDateMonthFormat = isDateMonthFormat;
function validateDateWithUnknownFormat(v) {
    for (const format of exports.dateFormats) {
        if ((0, dayjs_1.default)(v, format, true).isValid()) {
            return true;
        }
        for (const timeFormat of exports.timeFormats) {
            if ((0, dayjs_1.default)(v, `${format} ${timeFormat}`, true).isValid()) {
                return true;
            }
        }
    }
    return false;
}
function getDateFormat(v) {
    for (const format of exports.dateFormats.concat(exports.dateMonthFormats)) {
        if ((0, dayjs_1.default)(v, format, true).isValid()) {
            return format;
        }
    }
    return 'YYYY/MM/DD';
}
function getDateTimeFormat(v) {
    for (const format of exports.dateFormats) {
        for (const timeFormat of exports.timeFormats) {
            const dateTimeFormat = `${format} ${timeFormat}`;
            if ((0, dayjs_1.default)(v, dateTimeFormat, true).isValid()) {
                return dateTimeFormat;
            }
        }
    }
    return 'YYYY/MM/DD HH:mm';
}
function parseStringDate(v, dateFormat) {
    const dayjsObj = (0, dayjs_1.default)(v);
    if (dayjsObj.isValid()) {
        v = dayjsObj.format('YYYY-MM-DD');
    }
    else {
        v = (0, dayjs_1.default)(v, dateFormat).format('YYYY-MM-DD');
    }
    return v;
}
function parseStringDateTime(v, dateTimeFormat = `${exports.dateFormats[0]} ${exports.timeFormats[0]}`, toLocal = true) {
    const dayjsObj = toLocal ? (0, dayjs_1.default)(v).local() : (0, dayjs_1.default)(v);
    if (dayjsObj.isValid()) {
        v = dayjsObj.format(dateTimeFormat);
    }
    else {
        v = toLocal
            ? (0, dayjs_1.default)(v, dateTimeFormat).local().format(dateTimeFormat)
            : (0, dayjs_1.default)(v, dateTimeFormat).format(dateTimeFormat);
    }
    return v;
}
function convertToTargetFormat(v, oldDataFormat, newDateFormat) {
    if (!exports.dateFormats.includes(oldDataFormat) ||
        !exports.dateFormats.includes(newDateFormat))
        return v;
    return (0, dayjs_1.default)(v, oldDataFormat).format(newDateFormat);
}
const handleTZ = (val) => {
    if (val === undefined || val === null) {
        return;
    }
    if (typeof val !== 'string') {
        return val;
    }
    return val.replace(
    // match and extract dates and times in the ISO 8601 format
    /((?:-?(?:[1-9][0-9]*)?[0-9]{4})-(?:1[0-2]|0[1-9])-(?:3[01]|0[1-9]|[12][0-9])T(?:2[0-3]|[01][0-9]):(?:[0-5][0-9]):(?:[0-5][0-9])(?:\.[0-9]+)?(?:Z|[+-](?:2[0-3]|[01][0-9]):[0-5][0-9]))/g, (_, v) => {
        return (0, dayjs_1.default)(v).format('YYYY-MM-DD HH:mm');
    });
};
exports.handleTZ = handleTZ;
function validateDateFormat(v) {
    return exports.dateFormats.includes(v);
}
const timeAgo = (date) => {
    if (/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$/.test(date)) {
        // if there is no timezone info, consider as UTC
        // e.g. 2023-01-01 08:00:00 (MySQL)
        date += '+00:00';
    }
    // show in local time
    return (0, dayjs_1.default)(date).fromNow();
};
exports.timeAgo = timeAgo;
const isValidTimeFormat = (value, format) => {
    const regexValidator = {
        [exports.timeFormats[0]]: /^([01]\d|2[0-3]):[0-5]\d$/,
        [exports.timeFormats[1]]: /^([01]\d|2[0-3]):[0-5]\d:[0-5]\d$/,
        [exports.timeFormats[2]]: /^([01]\d|2[0-3]):[0-5]\d:[0-5]\d\.\d{3}$/,
    };
    if (regexValidator[format]) {
        return regexValidator[format].test(value);
    }
    return false;
};
exports.isValidTimeFormat = isValidTimeFormat;
function constructDateTimeFormat(column) {
    const dateFormat = constructDateFormat(column);
    const timeFormat = constructTimeFormat(column);
    return `${dateFormat} ${timeFormat}`;
}
function constructDateFormat(column) {
    var _a, _b;
    return (_b = (_a = (0, helperFunctions_1.parseProp)(column === null || column === void 0 ? void 0 : column.meta)) === null || _a === void 0 ? void 0 : _a.date_format) !== null && _b !== void 0 ? _b : exports.dateFormats[0];
}
function constructTimeFormat(column) {
    var _a;
    const columnMeta = (0, helperFunctions_1.parseProp)(column === null || column === void 0 ? void 0 : column.meta);
    const metaTimeFormat = (_a = columnMeta.time_format) !== null && _a !== void 0 ? _a : exports.timeFormats[0];
    return (columnMeta === null || columnMeta === void 0 ? void 0 : columnMeta.is12hrFormat)
        ? metaTimeFormat.replace('HH', 'hh') + ' A' // if 12h, replace HH and add AM/PM at the end
        : metaTimeFormat;
}
function workerWithTimezone(isEeUI, timezone) {
    // Check if the timezone is UTC or GMT (case insensitive)
    const isUtcOrGmt = timezone && /^(utc|gmt)$/i.test(timezone);
    return {
        dayjsTz(value, format) {
            if (!isEeUI) {
                return (0, dayjs_1.default)(value, format);
            }
            if ((0, is_1.ncIsNull)(value) || (0, is_1.ncIsUndefined)(value)) {
                if (timezone) {
                    return dayjs_1.default.tz(undefined, timezone);
                }
                else {
                    return (0, dayjs_1.default)();
                }
            }
            else if (typeof value === 'object' && value.isValid()) {
                return value;
            }
            if (timezone) {
                if (isUtcOrGmt) {
                    const strValue = typeof value === 'object' &&
                        typeof value.isValid === 'function' &&
                        value.isValid()
                        ? value.toISOString()
                        : value;
                    return format
                        ? dayjs_1.default.tz(strValue, format, timezone)
                        : dayjs_1.default.tz(strValue, timezone);
                }
                else {
                    if (!format) {
                        return dayjs_1.default.tz(value, timezone);
                    }
                    else {
                        return dayjs_1.default.tz(value, format, timezone);
                    }
                }
            }
            else {
                return (0, dayjs_1.default)(value, format);
            }
        },
        timezonize(value) {
            if (!timezone) {
                return (0, dayjs_1.default)(value);
            }
            if (!value) {
                return this.dayjsTz();
            }
            let dayjsObject;
            if (typeof value === 'object' &&
                typeof value.isValid === 'function' &&
                value.isValid()) {
                dayjsObject = value.isUTC() ? value : value.utc();
            }
            else {
                dayjsObject = dayjs_1.default.utc(value);
            }
            if (!isEeUI) {
                return dayjsObject.local();
            }
            if (timezone) {
                if (isUtcOrGmt) {
                    return (0, dayjs_1.default)(dayjsObject.toISOString()).tz(timezone);
                }
                else {
                    return dayjsObject.tz(timezone);
                }
            }
            return dayjsObject.local();
        },
    };
}
const getDateTimeValue = (modelValue, col, isXcdbBase) => {
    var _a, _b, _c, _d;
    if (!modelValue || !(0, dayjs_1.default)(modelValue).isValid()) {
        return '';
    }
    const dateFormat = (_b = (_a = (0, helperFunctions_1.parseProp)(col === null || col === void 0 ? void 0 : col.meta)) === null || _a === void 0 ? void 0 : _a.date_format) !== null && _b !== void 0 ? _b : exports.dateFormats[0];
    const timeFormat = (_d = (_c = (0, helperFunctions_1.parseProp)(col === null || col === void 0 ? void 0 : col.meta)) === null || _c === void 0 ? void 0 : _c.time_format) !== null && _d !== void 0 ? _d : exports.timeFormats[0];
    const dateTimeFormat = `${dateFormat} ${timeFormat}`;
    if (!isXcdbBase) {
        return (0, dayjs_1.default)(/^\d+$/.test(modelValue) ? +modelValue : modelValue, dateTimeFormat).format(dateTimeFormat);
    }
    return (0, dayjs_1.default)(modelValue).utc().local().format(dateTimeFormat);
};
exports.getDateTimeValue = getDateTimeValue;
const getDateValue = (modelValue, col, isSystemCol) => {
    var _a, _b;
    const dateFormat = !isSystemCol
        ? (_b = (_a = (0, helperFunctions_1.parseProp)(col.meta)) === null || _a === void 0 ? void 0 : _a.date_format) !== null && _b !== void 0 ? _b : 'YYYY-MM-DD'
        : 'YYYY-MM-DD HH:mm:ss';
    if (!modelValue || !(0, dayjs_1.default)(modelValue).isValid()) {
        return '';
    }
    return (0, dayjs_1.default)(/^\d+$/.test(String(modelValue)) ? +modelValue : modelValue).format(dateFormat);
};
exports.getDateValue = getDateValue;
