// nocodb-sdk/build/main/lib/filterHelpers_withparser.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const filterHelpers_old_spec_1 = require("./filterHelpers_old.spec");
const filterHelpers_withparser_1 = require("./filterHelpers_withparser");
const UITypes_1 = __importDefault(require("./UITypes"));
(0, filterHelpers_old_spec_1.testExtractFilterFromXwhere)('filterHelpers_withparser', filterHelpers_withparser_1.extractFilterFromXwhere);
describe('filterHelpers_withparser_specific', () => {
    describe('extractFilterFromXwhere', () => {
        describe('logical', () => {
            it('will parse basic logical query', () => {
                var _a;
                // isWithin need to have specific suboperator :|
                const query = '(Date,isWithin,pastMonth)~and(Name,like,Hello)';
                const columnAlias = {
                    Date: {
                        id: 'field1',
                        column_name: 'col1',
                        title: 'Date',
                        uidt: UITypes_1.default.DateTime,
                    },
                    Name: {
                        id: 'field2',
                        column_name: 'col2',
                        title: 'Name',
                        uidt: UITypes_1.default.SingleLineText,
                    },
                };
                const result = (0, filterHelpers_withparser_1.extractFilterFromXwhere)(query, columnAlias);
                expect(result).toBeDefined();
                expect(result.filters).toBeDefined();
                expect(result.filters.length).toBe(1);
                expect((_a = result.filters[0].children) === null || _a === void 0 ? void 0 : _a[1].logical_op).toBe('and');
            });
            it('will parse nested logical query', () => {
                var _a;
                // isWithin need to have specific suboperator :|
                const query = '(Date,isWithin,pastMonth)~or((Name,like,Hello)~and(Name,like,World))';
                const columnAlias = {
                    Date: {
                        id: 'field1',
                        column_name: 'col1',
                        title: 'Date',
                        uidt: UITypes_1.default.DateTime,
                    },
                    Name: {
                        id: 'field2',
                        column_name: 'col2',
                        title: 'Name',
                        uidt: UITypes_1.default.SingleLineText,
                    },
                };
                const result = (0, filterHelpers_withparser_1.extractFilterFromXwhere)(query, columnAlias);
                expect(result).toBeDefined();
                expect(result.filters).toBeDefined();
                expect(result.filters.length).toBe(1);
                expect((_a = result.filters[0].children) === null || _a === void 0 ? void 0 : _a[1].logical_op).toBe('or');
            });
        });
    });
});
