// nocodb-sdk/build/main/lib/form.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hideExtraFieldsMetaKey = exports.oppositeValidationTypeMap = exports.AttachmentValidationType = exports.SelectValidationType = exports.YearValidationType = exports.TimeValidationType = exports.DateValidationType = exports.NumberValidationType = exports.StringValidationType = exports.GenericValidationType = void 0;
var GenericValidationType;
(function (GenericValidationType) {
    GenericValidationType["Required"] = "required";
})(GenericValidationType || (exports.GenericValidationType = GenericValidationType = {}));
var StringValidationType;
(function (StringValidationType) {
    StringValidationType["MinLength"] = "minLength";
    StringValidationType["MaxLength"] = "maxLength";
    StringValidationType["StartsWith"] = "startsWith";
    StringValidationType["EndsWith"] = "endsWith";
    StringValidationType["Includes"] = "includes";
    StringValidationType["NotIncludes"] = "notIncludes";
    StringValidationType["Regex"] = "regex";
    StringValidationType["Email"] = "email";
    StringValidationType["BusinessEmail"] = "businessEmail";
    StringValidationType["PhoneNumber"] = "phoneNumber";
    StringValidationType["Url"] = "url";
})(StringValidationType || (exports.StringValidationType = StringValidationType = {}));
var NumberValidationType;
(function (NumberValidationType) {
    NumberValidationType["Min"] = "min";
    NumberValidationType["Max"] = "max";
})(NumberValidationType || (exports.NumberValidationType = NumberValidationType = {}));
var DateValidationType;
(function (DateValidationType) {
    DateValidationType["MinDate"] = "minDate";
    DateValidationType["MaxDate"] = "maxDate";
})(DateValidationType || (exports.DateValidationType = DateValidationType = {}));
var TimeValidationType;
(function (TimeValidationType) {
    TimeValidationType["MinTime"] = "minTime";
    TimeValidationType["MaxTime"] = "maxTime";
})(TimeValidationType || (exports.TimeValidationType = TimeValidationType = {}));
var YearValidationType;
(function (YearValidationType) {
    YearValidationType["MinYear"] = "minYear";
    YearValidationType["MaxYear"] = "maxYear";
})(YearValidationType || (exports.YearValidationType = YearValidationType = {}));
var SelectValidationType;
(function (SelectValidationType) {
    SelectValidationType["MinSelected"] = "minSelected";
    SelectValidationType["MaxSelected"] = "maxSelected";
    SelectValidationType["LimitOptions"] = "limitOptions";
})(SelectValidationType || (exports.SelectValidationType = SelectValidationType = {}));
var AttachmentValidationType;
(function (AttachmentValidationType) {
    AttachmentValidationType["FileTypes"] = "fileTypes";
    AttachmentValidationType["FileSize"] = "fileSize";
    AttachmentValidationType["FileCount"] = "fileCount";
})(AttachmentValidationType || (exports.AttachmentValidationType = AttachmentValidationType = {}));
exports.oppositeValidationTypeMap = {
    [StringValidationType.MaxLength]: StringValidationType.MinLength,
    [StringValidationType.NotIncludes]: StringValidationType.Includes,
    [NumberValidationType.Max]: NumberValidationType.Min,
    [YearValidationType.MaxYear]: YearValidationType.MinYear,
    [DateValidationType.MaxDate]: DateValidationType.MinDate,
    [TimeValidationType.MaxTime]: TimeValidationType.MinTime,
    [SelectValidationType.MaxSelected]: SelectValidationType.MinSelected,
};
exports.hideExtraFieldsMetaKey = 'hideExtraFields';
