// nocodb-sdk/build/main/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemplateGenerator = exports.CustomAPI = exports.UITypesSearchTerms = exports.customLinkSupportedTypes = exports.isDateOrDateTimeCol = exports.isReadOnlyColumn = exports.durationOptions = exports.ratingIconListMap = exports.ratingIconList = exports.checkboxIconListMap = exports.checkboxIconList = exports.columnTypeName = exports.isSupportedDisplayValueColumn = exports.partialUpdateAllowedTypes = exports.readonlyMetaAllowedTypes = exports.getUITypesForFormulaDataType = exports.isOrderCol = exports.isSelectTypeCol = exports.isActionButtonCol = exports.getEquivalentUIType = exports.isHiddenCol = exports.isCreatedOrLastModifiedByCol = exports.isCreatedOrLastModifiedTimeCol = exports.isSelfLinkCol = exports.isLinksOrLTAR = exports.isVirtualCol = exports.isNumericCol = exports.isAIPromptCol = exports.numericUITypes = exports.FieldNameFromUITypes = exports.UITypesName = exports.UITypes = void 0;
__exportStar(require("../lib/XcUIBuilder"), exports);
__exportStar(require("../lib/XcNotification"), exports);
__exportStar(require("../lib/Api"), exports);
__exportStar(require("../lib/columnRules"), exports);
__exportStar(require("../lib/sqlUi"), exports);
__exportStar(require("../lib/globals"), exports);
__exportStar(require("../lib/helperFunctions"), exports);
__exportStar(require("../lib/enums"), exports);
__exportStar(require("../lib/formulaHelpers"), exports);
__exportStar(require("../lib/regex"), exports);
__exportStar(require("../lib/payment"), exports);
__exportStar(require("../lib/permission"), exports);
__exportStar(require("../lib/realtime"), exports);
var UITypes_1 = require("../lib/UITypes");
Object.defineProperty(exports, "UITypes", { enumerable: true, get: function () { return __importDefault(UITypes_1).default; } });
Object.defineProperty(exports, "UITypesName", { enumerable: true, get: function () { return UITypes_1.UITypesName; } });
Object.defineProperty(exports, "FieldNameFromUITypes", { enumerable: true, get: function () { return UITypes_1.FieldNameFromUITypes; } });
Object.defineProperty(exports, "numericUITypes", { enumerable: true, get: function () { return UITypes_1.numericUITypes; } });
Object.defineProperty(exports, "isAIPromptCol", { enumerable: true, get: function () { return UITypes_1.isAIPromptCol; } });
Object.defineProperty(exports, "isNumericCol", { enumerable: true, get: function () { return UITypes_1.isNumericCol; } });
Object.defineProperty(exports, "isVirtualCol", { enumerable: true, get: function () { return UITypes_1.isVirtualCol; } });
Object.defineProperty(exports, "isLinksOrLTAR", { enumerable: true, get: function () { return UITypes_1.isLinksOrLTAR; } });
Object.defineProperty(exports, "isSelfLinkCol", { enumerable: true, get: function () { return UITypes_1.isSelfLinkCol; } });
Object.defineProperty(exports, "isCreatedOrLastModifiedTimeCol", { enumerable: true, get: function () { return UITypes_1.isCreatedOrLastModifiedTimeCol; } });
Object.defineProperty(exports, "isCreatedOrLastModifiedByCol", { enumerable: true, get: function () { return UITypes_1.isCreatedOrLastModifiedByCol; } });
Object.defineProperty(exports, "isHiddenCol", { enumerable: true, get: function () { return UITypes_1.isHiddenCol; } });
Object.defineProperty(exports, "getEquivalentUIType", { enumerable: true, get: function () { return UITypes_1.getEquivalentUIType; } });
Object.defineProperty(exports, "isActionButtonCol", { enumerable: true, get: function () { return UITypes_1.isActionButtonCol; } });
Object.defineProperty(exports, "isSelectTypeCol", { enumerable: true, get: function () { return UITypes_1.isSelectTypeCol; } });
Object.defineProperty(exports, "isOrderCol", { enumerable: true, get: function () { return UITypes_1.isOrderCol; } });
Object.defineProperty(exports, "getUITypesForFormulaDataType", { enumerable: true, get: function () { return UITypes_1.getUITypesForFormulaDataType; } });
Object.defineProperty(exports, "readonlyMetaAllowedTypes", { enumerable: true, get: function () { return UITypes_1.readonlyMetaAllowedTypes; } });
Object.defineProperty(exports, "partialUpdateAllowedTypes", { enumerable: true, get: function () { return UITypes_1.partialUpdateAllowedTypes; } });
Object.defineProperty(exports, "isSupportedDisplayValueColumn", { enumerable: true, get: function () { return UITypes_1.isSupportedDisplayValueColumn; } });
Object.defineProperty(exports, "columnTypeName", { enumerable: true, get: function () { return UITypes_1.columnTypeName; } });
Object.defineProperty(exports, "checkboxIconList", { enumerable: true, get: function () { return UITypes_1.checkboxIconList; } });
Object.defineProperty(exports, "checkboxIconListMap", { enumerable: true, get: function () { return UITypes_1.checkboxIconListMap; } });
Object.defineProperty(exports, "ratingIconList", { enumerable: true, get: function () { return UITypes_1.ratingIconList; } });
Object.defineProperty(exports, "ratingIconListMap", { enumerable: true, get: function () { return UITypes_1.ratingIconListMap; } });
Object.defineProperty(exports, "durationOptions", { enumerable: true, get: function () { return UITypes_1.durationOptions; } });
Object.defineProperty(exports, "isReadOnlyColumn", { enumerable: true, get: function () { return UITypes_1.isReadOnlyColumn; } });
Object.defineProperty(exports, "isDateOrDateTimeCol", { enumerable: true, get: function () { return UITypes_1.isDateOrDateTimeCol; } });
Object.defineProperty(exports, "customLinkSupportedTypes", { enumerable: true, get: function () { return UITypes_1.customLinkSupportedTypes; } });
Object.defineProperty(exports, "UITypesSearchTerms", { enumerable: true, get: function () { return UITypes_1.UITypesSearchTerms; } });
var CustomAPI_1 = require("../lib/CustomAPI");
Object.defineProperty(exports, "CustomAPI", { enumerable: true, get: function () { return __importDefault(CustomAPI_1).default; } });
var TemplateGenerator_1 = require("../lib/TemplateGenerator");
Object.defineProperty(exports, "TemplateGenerator", { enumerable: true, get: function () { return __importDefault(TemplateGenerator_1).default; } });
__exportStar(require("../lib/passwordHelpers"), exports);
__exportStar(require("../lib/mergeSwaggerSchema"), exports);
__exportStar(require("../lib/dateTimeHelper"), exports);
__exportStar(require("../lib/form"), exports);
__exportStar(require("../lib/aggregationHelper"), exports);
__exportStar(require("../lib/connectionConfigUtils"), exports);
__exportStar(require("../lib/currencyHelpers"), exports);
__exportStar(require("../lib/filterHelpers"), exports);
__exportStar(require("../lib/errorUtils"), exports);
__exportStar(require("../lib/formBuilder"), exports);
__exportStar(require("../lib/ai"), exports);
__exportStar(require("../lib/audit"), exports);
__exportStar(require("../lib/ncTypes"), exports);
__exportStar(require("../lib/import-export-data"), exports);
__exportStar(require("../lib/is"), exports);
__exportStar(require("../lib/durationUtils"), exports);
__exportStar(require("../lib/sync"), exports);
__exportStar(require("../lib/columnHelper"), exports);
__exportStar(require("../lib/error"), exports);
__exportStar(require("../lib/error-handler"), exports);
__exportStar(require("../lib/geoDataUtils"), exports);
__exportStar(require("../lib/stringHelpers"), exports);
__exportStar(require("../lib/filter"), exports);
__exportStar(require("../lib/dashboard"), exports);
__exportStar(require("../lib/arrayHelpers"), exports);
__exportStar(require("../lib/roleHelpers"), exports);
