// nocodb-sdk/build/main/lib/numberUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCurrencyValue = exports.roundTo = exports.numberize = void 0;
const helperFunctions_1 = require("../lib/helperFunctions");
const numberize = (value) => {
    if (value === undefined || value === null) {
        return value;
    }
    if (typeof value === 'number') {
        return value;
    }
    else {
        const result = parseInt(value);
        if (isNaN(result)) {
            return undefined;
        }
        return result;
    }
};
exports.numberize = numberize;
const roundTo = (num, precision = 1) => {
    if (!num || Number.isNaN(num))
        return num;
    const factor = 10 ** precision;
    return Math.round(+num * factor) / factor;
};
exports.roundTo = roundTo;
const getCurrencyValue = (modelValue, col) => {
    var _a, _b, _c;
    const currencyMeta = Object.assign({ currency_locale: 'en-US', currency_code: 'USD', precision: 2 }, (0, helperFunctions_1.parseProp)(col.meta));
    try {
        if (modelValue === null ||
            modelValue === undefined ||
            Number.isNaN(modelValue)) {
            return modelValue === null || modelValue === undefined
                ? ''
                : modelValue;
        }
        const roundedValue = (0, helperFunctions_1.roundUpToPrecision)(Number(modelValue), (_a = currencyMeta.precision) !== null && _a !== void 0 ? _a : 2);
        return new Intl.NumberFormat(currencyMeta.currency_locale || 'en-US', {
            style: 'currency',
            currency: currencyMeta.currency_code || 'USD',
            minimumFractionDigits: (_b = currencyMeta.precision) !== null && _b !== void 0 ? _b : 2,
            maximumFractionDigits: (_c = currencyMeta.precision) !== null && _c !== void 0 ? _c : 2,
        }).format(+roundedValue);
    }
    catch (e) {
        return modelValue;
    }
};
exports.getCurrencyValue = getCurrencyValue;
