// nocodb-sdk/build/main/lib/stringHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppendToLengthSuffix = void 0;
exports.getRowColPositionFromIndex = getRowColPositionFromIndex;
exports.appendToLength = appendToLength;
exports.truncateToLength = truncateToLength;
exports.generateUniqueCopyName = generateUniqueCopyName;
exports.escapeRegexString = escapeRegexString;
const is_1 = require("./is");
const AppendToLengthSuffixConfig = {
    _: {
        replacement: '___{index}',
        replacementRegex: /___\s?(\d*)$/,
    },
    dot: {
        replacement: '...{index}',
        replacementRegex: /\.\.\.\s?(\d*)$/,
    },
};
var AppendToLengthSuffix;
(function (AppendToLengthSuffix) {
    AppendToLengthSuffix["_"] = "_";
    AppendToLengthSuffix["dot"] = "dot";
})(AppendToLengthSuffix || (exports.AppendToLengthSuffix = AppendToLengthSuffix = {}));
/**
 * get a row & column given a numeric position
 * row & column start at 0
 */
function getRowColPositionFromIndex({ stack, position, }) {
    const parts = stack.substring(0, position).split('\n');
    return {
        column: parts[parts.length - 1].length,
        row: parts.length - 1,
    };
}
async function appendToLength(param) {
    var _a, _b;
    const { value, appendage, maxLength, isExists } = param;
    const suffixConfig = AppendToLengthSuffixConfig[(_a = param.suffix) !== null && _a !== void 0 ? _a : AppendToLengthSuffix.dot];
    // if it's already in the form of truncated
    // skip with usual append
    if ((value + appendage).length > maxLength) {
        const existingIndex = (_b = value.match(suffixConfig.replacementRegex)) === null || _b === void 0 ? void 0 : _b[1];
        return truncateToLength({
            value,
            maxLength,
            currentIndex: !(0, is_1.ncIsUndefined)(existingIndex) && existingIndex !== ''
                ? Number(existingIndex)
                : undefined,
            isExists,
            suffix: param.suffix,
        });
    }
    let currentIndex = 1;
    let needle = value + appendage;
    while (await isExists(needle)) {
        needle = value + appendage + '_' + currentIndex++;
        if (needle.length >= maxLength) {
            return appendToLength(Object.assign(Object.assign({}, param), { appendage: appendage + '_' + currentIndex++, suffix: param.suffix }));
        }
    }
    return needle;
}
async function truncateToLength(param) {
    var _a;
    const { value, currentIndex, maxLength, isExists } = param;
    const suffixConfig = AppendToLengthSuffixConfig[(_a = param.suffix) !== null && _a !== void 0 ? _a : AppendToLengthSuffix.dot];
    const replacement = (0, is_1.ncIsUndefined)(currentIndex) ? '' : `${currentIndex}`;
    const suffix = suffixConfig.replacement.replace('{index}', replacement);
    const needle = value.substring(0, maxLength - suffix.length) + suffix;
    if (!(await isExists(needle))) {
        return needle;
    }
    else {
        return truncateToLength(Object.assign(Object.assign({}, param), { currentIndex: (0, is_1.ncIsUndefined)(currentIndex) ? 2 : currentIndex + 1 }));
    }
}
/**
 * Generates a unique copy name by checking against existing names/items
 */
function generateUniqueCopyName(originalName, existing, options = {}) {
    const { accessor, prefix = 'Copy of', separator = ' ', counterFormat = '({counter})', } = options;
    // Extract names from the existing array
    const existingNames = existing.map((item) => {
        if (typeof item === 'string')
            return item;
        if (accessor) {
            return typeof accessor === 'function'
                ? accessor(item)
                : String(item[accessor]);
        }
        // Default to 'title' property if no accessor specified
        return String(item.title);
    });
    let newName = `${prefix} ${originalName}`;
    let counter = 1;
    while (existingNames.includes(newName)) {
        const counterText = counterFormat.replace('{counter}', counter.toString());
        newName = `${prefix} ${originalName}${separator}${counterText}`;
        counter++;
    }
    return newName;
}
function escapeRegexString(string) {
    return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
