// nocodb-sdk/build/main/lib/XcUIBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XcActionType = exports.XcType = void 0;
var XcType;
(function (XcType) {
    XcType["ID"] = "ID";
    XcType["ForeignKey"] = "ForeignKey";
    XcType["SingleLineText"] = "SingleLineText";
    XcType["LongText"] = "LongText";
    XcType["Attachment"] = "Attachment";
    XcType["Checkbox"] = "Checkbox";
    XcType["MultiSelect"] = "MultiSelect";
    XcType["SingleSelect"] = "SingleSelect";
    XcType["Collaborator"] = "Collaborator";
    XcType["Date"] = "Date";
    XcType["Year"] = "Year";
    XcType["Time"] = "Time";
    XcType["PhoneNumber"] = "PhoneNumber";
    XcType["Email"] = "Email";
    XcType["URL"] = "URL";
    XcType["Number"] = "Number";
    XcType["Decimal"] = "Decimal";
    XcType["Currency"] = "Currency";
    XcType["Percent"] = "Percent";
    XcType["Duration"] = "Duration";
    XcType["Rating"] = "Rating";
    XcType["Formula"] = "Formula";
    XcType["Rollup"] = "Rollup";
    XcType["Count"] = "Count";
    XcType["Lookup"] = "Lookup";
    XcType["DateTime"] = "DateTime";
    XcType["CreateTime"] = "CreateTime";
    XcType["LastModifiedTime"] = "LastModifiedTime";
    XcType["AutoNumber"] = "AutoNumber";
    XcType["Barcode"] = "Barcode";
    XcType["Button"] = "Button";
    XcType["Password"] = "Password";
})(XcType || (exports.XcType = XcType = {}));
var XcActionType;
(function (XcActionType) {
    XcActionType["SUBMIT"] = "SUBMIT";
    XcActionType["TEST"] = "TEST";
    XcActionType["CLEAR"] = "CLEAR";
    XcActionType["CANCEL"] = "CANCEL";
})(XcActionType || (exports.XcActionType = XcActionType = {}));
