// nocodb-sdk/build/main/lib/permission/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPermissionOptionValue = exports.getPermissionIcon = exports.getPermissionLabel = exports.getPermissionOption = exports.PermissionMeta = exports.PermissionRoleMap = exports.PermissionRolePower = exports.PermissionOptions = exports.PermissionOptionValue = exports.PermissionRole = exports.PermissionEntity = exports.PermissionGrantedType = exports.PermissionKey = void 0;
const enums_1 = require("../enums");
var PermissionKey;
(function (PermissionKey) {
    PermissionKey["TABLE_RECORD_ADD"] = "TABLE_RECORD_ADD";
    PermissionKey["TABLE_RECORD_DELETE"] = "TABLE_RECORD_DELETE";
    PermissionKey["RECORD_FIELD_EDIT"] = "RECORD_FIELD_EDIT";
})(PermissionKey || (exports.PermissionKey = PermissionKey = {}));
var PermissionGrantedType;
(function (PermissionGrantedType) {
    PermissionGrantedType["ROLE"] = "role";
    PermissionGrantedType["USER"] = "user";
    PermissionGrantedType["NOBODY"] = "nobody";
})(PermissionGrantedType || (exports.PermissionGrantedType = PermissionGrantedType = {}));
var PermissionEntity;
(function (PermissionEntity) {
    PermissionEntity["TABLE"] = "table";
    PermissionEntity["FIELD"] = "field";
})(PermissionEntity || (exports.PermissionEntity = PermissionEntity = {}));
var PermissionRole;
(function (PermissionRole) {
    PermissionRole["OWNER"] = "owner";
    PermissionRole["CREATOR"] = "creator";
    PermissionRole["EDITOR"] = "editor";
    PermissionRole["COMMENTER"] = "commenter";
    PermissionRole["VIEWER"] = "viewer";
})(PermissionRole || (exports.PermissionRole = PermissionRole = {}));
// Permission option values used across the application
var PermissionOptionValue;
(function (PermissionOptionValue) {
    PermissionOptionValue["VIEWERS_AND_UP"] = "viewers_and_up";
    PermissionOptionValue["EDITORS_AND_UP"] = "editors_and_up";
    PermissionOptionValue["CREATORS_AND_UP"] = "creators_and_up";
    PermissionOptionValue["SPECIFIC_USERS"] = "specific_users";
    PermissionOptionValue["NOBODY"] = "nobody";
})(PermissionOptionValue || (exports.PermissionOptionValue = PermissionOptionValue = {}));
exports.PermissionOptions = [
    {
        value: PermissionOptionValue.CREATORS_AND_UP,
        label: 'Creators & up',
        description: 'Members with Creator or Owner role',
        icon: 'role_creator',
    },
    {
        value: PermissionOptionValue.EDITORS_AND_UP,
        label: 'Editors & up',
        description: 'Members with Editor, Creator or Owner role',
        icon: 'role_editor',
        isDefault: true,
    },
    {
        value: PermissionOptionValue.VIEWERS_AND_UP,
        label: 'Viewers and up',
        description: 'Members with Viewer, Editor, Creator or Owner role',
        icon: 'role_viewer',
    },
    {
        value: PermissionOptionValue.SPECIFIC_USERS,
        label: 'Specific users',
        description: 'Specific set of members',
        icon: 'ncUsers',
    },
    {
        value: PermissionOptionValue.NOBODY,
        label: 'Nobody',
        description: 'No one can add records',
        icon: 'role_no_access',
    },
];
exports.PermissionRolePower = {
    [PermissionRole.OWNER]: 6,
    [PermissionRole.CREATOR]: 5,
    [PermissionRole.EDITOR]: 4,
    [PermissionRole.COMMENTER]: 3,
    [PermissionRole.VIEWER]: 2,
};
exports.PermissionRoleMap = {
    [enums_1.ProjectRoles.OWNER]: PermissionRole.OWNER,
    [enums_1.ProjectRoles.CREATOR]: PermissionRole.CREATOR,
    [enums_1.ProjectRoles.EDITOR]: PermissionRole.EDITOR,
    [enums_1.ProjectRoles.COMMENTER]: PermissionRole.COMMENTER,
    [enums_1.ProjectRoles.VIEWER]: PermissionRole.VIEWER,
    [enums_1.WorkspaceUserRoles.OWNER]: PermissionRole.OWNER,
    [enums_1.WorkspaceUserRoles.CREATOR]: PermissionRole.CREATOR,
    [enums_1.WorkspaceUserRoles.EDITOR]: PermissionRole.EDITOR,
    [enums_1.WorkspaceUserRoles.COMMENTER]: PermissionRole.COMMENTER,
    [enums_1.WorkspaceUserRoles.VIEWER]: PermissionRole.VIEWER,
};
exports.PermissionMeta = {
    [PermissionKey.TABLE_RECORD_ADD]: {
        minimumRole: PermissionRole.EDITOR,
        label: 'Who can create records',
        description: 'can create records',
        userSelectorDescription: 'Only members selected here will be able to create records.',
    },
    [PermissionKey.TABLE_RECORD_DELETE]: {
        minimumRole: PermissionRole.EDITOR,
        label: 'Who can delete records',
        description: 'can delete records',
        userSelectorDescription: 'Only members selected here will be able to delete records.',
    },
    [PermissionKey.RECORD_FIELD_EDIT]: {
        minimumRole: PermissionRole.EDITOR,
        label: 'Who can edit data in this field',
        description: 'can edit records',
        userSelectorDescription: 'Only members selected here will be able to edit values in the {{field}} field.',
    },
};
// Utility functions for permission management
const getPermissionOption = (value) => {
    return exports.PermissionOptions.find((option) => option.value === value);
};
exports.getPermissionOption = getPermissionOption;
const getPermissionLabel = (value) => {
    var _a;
    return ((_a = (0, exports.getPermissionOption)(value)) === null || _a === void 0 ? void 0 : _a.label) || 'Editors & up';
};
exports.getPermissionLabel = getPermissionLabel;
const getPermissionIcon = (value) => {
    var _a;
    return ((_a = (0, exports.getPermissionOption)(value)) === null || _a === void 0 ? void 0 : _a.icon) || 'role_editor';
};
exports.getPermissionIcon = getPermissionIcon;
const getPermissionOptionValue = (grantedType, grantedRole) => {
    if (grantedType === PermissionGrantedType.ROLE) {
        if (grantedRole === PermissionRole.VIEWER) {
            return PermissionOptionValue.VIEWERS_AND_UP;
        }
        else if (grantedRole === PermissionRole.CREATOR) {
            return PermissionOptionValue.CREATORS_AND_UP;
        }
        else {
            return PermissionOptionValue.EDITORS_AND_UP;
        }
    }
    else if (grantedType === PermissionGrantedType.USER) {
        return PermissionOptionValue.SPECIFIC_USERS;
    }
    else if (grantedType === PermissionGrantedType.NOBODY) {
        return PermissionOptionValue.NOBODY;
    }
    return PermissionOptionValue.EDITORS_AND_UP;
};
exports.getPermissionOptionValue = getPermissionOptionValue;
