// nocodb-sdk/build/module/lib/aggregationHelper.js
import UITypes from './UITypes';
import { FormulaDataTypes } from '../lib/formulaHelpers';
import { getDateTimeValue, getDateValue } from '../lib/dateTimeHelper';
import { formatBytes, parseProp } from '../lib/helperFunctions';
import { convertMS2Duration } from '../lib/durationUtils';
import { getCurrencyValue, roundTo } from '../lib/numberUtils';
import { ColumnHelper } from './columnHelper';
var NumericalAggregations;
(function (NumericalAggregations) {
    NumericalAggregations["Sum"] = "sum";
    NumericalAggregations["Min"] = "min";
    NumericalAggregations["Max"] = "max";
    NumericalAggregations["Avg"] = "avg";
    NumericalAggregations["Median"] = "median";
    NumericalAggregations["StandardDeviation"] = "std_dev";
    //   Histogram = 'histogram',
    NumericalAggregations["Range"] = "range";
})(NumericalAggregations || (NumericalAggregations = {}));
var CommonAggregations;
(function (CommonAggregations) {
    CommonAggregations["Count"] = "count";
    CommonAggregations["CountEmpty"] = "count_empty";
    CommonAggregations["CountFilled"] = "count_filled";
    CommonAggregations["CountUnique"] = "count_unique";
    CommonAggregations["PercentEmpty"] = "percent_empty";
    CommonAggregations["PercentFilled"] = "percent_filled";
    CommonAggregations["PercentUnique"] = "percent_unique";
    CommonAggregations["None"] = "none";
})(CommonAggregations || (CommonAggregations = {}));
var AttachmentAggregations;
(function (AttachmentAggregations) {
    AttachmentAggregations["AttachmentSize"] = "attachment_size";
})(AttachmentAggregations || (AttachmentAggregations = {}));
var BooleanAggregations;
(function (BooleanAggregations) {
    BooleanAggregations["Checked"] = "checked";
    BooleanAggregations["Unchecked"] = "unchecked";
    BooleanAggregations["PercentChecked"] = "percent_checked";
    BooleanAggregations["PercentUnchecked"] = "percent_unchecked";
})(BooleanAggregations || (BooleanAggregations = {}));
var DateAggregations;
(function (DateAggregations) {
    DateAggregations["EarliestDate"] = "earliest_date";
    DateAggregations["LatestDate"] = "latest_date";
    DateAggregations["DateRange"] = "date_range";
    DateAggregations["MonthRange"] = "month_range";
})(DateAggregations || (DateAggregations = {}));
const AllAggregations = Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, CommonAggregations), NumericalAggregations), AttachmentAggregations), BooleanAggregations), DateAggregations);
const getAvailableAggregations = (type, parsed_tree) => {
    let returnAggregations = [];
    if (type === UITypes.Formula && (parsed_tree === null || parsed_tree === void 0 ? void 0 : parsed_tree.dataType)) {
        switch (parsed_tree.dataType) {
            case FormulaDataTypes.BOOLEAN:
                returnAggregations = [
                    ...Object.values(BooleanAggregations),
                    CommonAggregations.None,
                ];
                break;
            case FormulaDataTypes.DATE:
                returnAggregations = [
                    ...Object.values(DateAggregations),
                    ...Object.values(CommonAggregations),
                ];
                break;
            case FormulaDataTypes.NUMERIC:
                returnAggregations = [
                    ...Object.values(NumericalAggregations),
                    ...Object.values(CommonAggregations),
                ];
                break;
            default:
                returnAggregations = [...Object.values(CommonAggregations)];
                break;
        }
    }
    switch (type) {
        case UITypes.Number:
        case UITypes.Decimal:
        case UITypes.Currency:
        case UITypes.Percent:
        case UITypes.Duration:
        case UITypes.Rating:
        case UITypes.Rollup:
        case UITypes.Links:
            returnAggregations = [
                ...Object.values(NumericalAggregations),
                ...Object.values(CommonAggregations),
            ];
            break;
        case UITypes.Checkbox:
            returnAggregations = [
                ...Object.values(BooleanAggregations),
                CommonAggregations.None,
            ];
            break;
        case UITypes.Date:
        case UITypes.DateTime:
        case UITypes.LastModifiedTime:
        case UITypes.CreatedTime:
            returnAggregations = [
                ...Object.values(DateAggregations),
                ...Object.values(CommonAggregations),
            ];
            break;
        case UITypes.SpecificDBType:
        case UITypes.ForeignKey:
            returnAggregations = [CommonAggregations.None];
            break;
        case UITypes.Button:
        case UITypes.Attachment:
            return [CommonAggregations.None, AttachmentAggregations.AttachmentSize];
    }
    if (!returnAggregations.length) {
        returnAggregations = [...Object.values(CommonAggregations)];
    }
    return returnAggregations.filter((item) => item !== CommonAggregations.Count);
};
const formatAggregation = (aggregation, value, column, 
/**
 * If columnHelperParams is provided then it will be used to format the aggregation value
 */
columnHelperParams) => {
    var _a, _b, _c, _d, _e, _f;
    if (column.uidt === UITypes.Formula) {
        if ((_a = column === null || column === void 0 ? void 0 : column.meta) === null || _a === void 0 ? void 0 : _a.display_type) {
            const childColumn = Object.assign({ uidt: (_b = column === null || column === void 0 ? void 0 : column.meta) === null || _b === void 0 ? void 0 : _b.display_type }, (_c = column === null || column === void 0 ? void 0 : column.meta) === null || _c === void 0 ? void 0 : _c.display_column_meta);
            return formatAggregation(aggregation, value, childColumn, Object.assign(Object.assign({}, columnHelperParams), { col: childColumn }));
        }
    }
    if ([DateAggregations.EarliestDate, DateAggregations.LatestDate].includes(aggregation)) {
        if (column.uidt === UITypes.DateTime) {
            return getDateTimeValue(value, column);
        }
        else if (column.uidt === UITypes.Date) {
            return getDateValue(value, column);
        }
        return getDateTimeValue(value, column);
    }
    if ([
        CommonAggregations.PercentEmpty,
        CommonAggregations.PercentFilled,
        CommonAggregations.PercentUnique,
        BooleanAggregations.PercentChecked,
        BooleanAggregations.PercentUnchecked,
    ].includes(aggregation)) {
        return `${(_d = roundTo(value, 1)) !== null && _d !== void 0 ? _d : 0}%`;
    }
    if ([DateAggregations.MonthRange, DateAggregations.DateRange].includes(aggregation)) {
        return aggregation === DateAggregations.DateRange
            ? `${value !== null && value !== void 0 ? value : 0} days`
            : `${value !== null && value !== void 0 ? value : 0} months`;
    }
    if ([
        CommonAggregations.Count,
        CommonAggregations.CountEmpty,
        CommonAggregations.CountFilled,
        CommonAggregations.CountUnique,
    ].includes(aggregation)) {
        return value;
    }
    if ([AttachmentAggregations.AttachmentSize].includes(aggregation)) {
        return formatBytes(value !== null && value !== void 0 ? value : 0);
    }
    if (columnHelperParams &&
        (columnHelperParams === null || columnHelperParams === void 0 ? void 0 : columnHelperParams.col) &&
        Object.values(NumericalAggregations).includes(aggregation) &&
        [
            UITypes.Number,
            UITypes.Decimal,
            UITypes.Currency,
            UITypes.Percent,
            UITypes.Duration,
            UITypes.Rollup,
            UITypes.Time,
        ].includes(column.uidt)) {
        return ColumnHelper.parsePlainCellValue(value, Object.assign(Object.assign({}, columnHelperParams), { isAggregation: true }));
    }
    if (column.uidt === UITypes.Currency) {
        return getCurrencyValue(value, column);
    }
    if (column.uidt === UITypes.Percent) {
        return `${roundTo(value, 1)}%`;
    }
    if (column.uidt === UITypes.Duration) {
        return convertMS2Duration(value, ((_e = parseProp(column.meta)) === null || _e === void 0 ? void 0 : _e.duration) || 0);
    }
    if (typeof value === 'number') {
        return (_f = roundTo(value, 1)) !== null && _f !== void 0 ? _f : '∞';
    }
    return value;
};
export { getAvailableAggregations, NumericalAggregations, CommonAggregations, BooleanAggregations, DateAggregations, AttachmentAggregations, AllAggregations, formatAggregation, };
