// nocodb-sdk/build/module/lib/form.js
export var GenericValidationType;
(function (GenericValidationType) {
    GenericValidationType["Required"] = "required";
})(GenericValidationType || (GenericValidationType = {}));
export var StringValidationType;
(function (StringValidationType) {
    StringValidationType["MinLength"] = "minLength";
    StringValidationType["MaxLength"] = "maxLength";
    StringValidationType["StartsWith"] = "startsWith";
    StringValidationType["EndsWith"] = "endsWith";
    StringValidationType["Includes"] = "includes";
    StringValidationType["NotIncludes"] = "notIncludes";
    StringValidationType["Regex"] = "regex";
    StringValidationType["Email"] = "email";
    StringValidationType["BusinessEmail"] = "businessEmail";
    StringValidationType["PhoneNumber"] = "phoneNumber";
    StringValidationType["Url"] = "url";
})(StringValidationType || (StringValidationType = {}));
export var NumberValidationType;
(function (NumberValidationType) {
    NumberValidationType["Min"] = "min";
    NumberValidationType["Max"] = "max";
})(NumberValidationType || (NumberValidationType = {}));
export var DateValidationType;
(function (DateValidationType) {
    DateValidationType["MinDate"] = "minDate";
    DateValidationType["MaxDate"] = "maxDate";
})(DateValidationType || (DateValidationType = {}));
export var TimeValidationType;
(function (TimeValidationType) {
    TimeValidationType["MinTime"] = "minTime";
    TimeValidationType["MaxTime"] = "maxTime";
})(TimeValidationType || (TimeValidationType = {}));
export var YearValidationType;
(function (YearValidationType) {
    YearValidationType["MinYear"] = "minYear";
    YearValidationType["MaxYear"] = "maxYear";
})(YearValidationType || (YearValidationType = {}));
export var SelectValidationType;
(function (SelectValidationType) {
    SelectValidationType["MinSelected"] = "minSelected";
    SelectValidationType["MaxSelected"] = "maxSelected";
    SelectValidationType["LimitOptions"] = "limitOptions";
})(SelectValidationType || (SelectValidationType = {}));
export var AttachmentValidationType;
(function (AttachmentValidationType) {
    AttachmentValidationType["FileTypes"] = "fileTypes";
    AttachmentValidationType["FileSize"] = "fileSize";
    AttachmentValidationType["FileCount"] = "fileCount";
})(AttachmentValidationType || (AttachmentValidationType = {}));
export const oppositeValidationTypeMap = {
    [StringValidationType.MaxLength]: StringValidationType.MinLength,
    [StringValidationType.NotIncludes]: StringValidationType.Includes,
    [NumberValidationType.Max]: NumberValidationType.Min,
    [YearValidationType.MaxYear]: YearValidationType.MinYear,
    [DateValidationType.MaxDate]: DateValidationType.MinDate,
    [TimeValidationType.MaxTime]: TimeValidationType.MinTime,
    [SelectValidationType.MaxSelected]: SelectValidationType.MinSelected,
};
export const hideExtraFieldsMetaKey = 'hideExtraFields';
