// nocodb-sdk/build/module/lib/globals.js
import { ProjectRoles, WorkspaceUserRoles } from './enums';
export const enumColors = {
    light: [
        '#cfdffe',
        '#d0f1fd',
        '#c2f5e8',
        '#ffdaf6',
        '#ffdce5',
        '#fee2d5',
        '#ffeab6',
        '#d1f7c4',
        '#ede2fe',
        '#eeeeee',
    ],
    dark: [
        '#2d7ff999',
        '#18bfff99',
        '#20d9d299',
        '#ff08c299',
        '#f82b6099',
        '#ff6f2c99',
        '#fcb40099',
        '#20c93399',
        '#8b46ff99',
        '#666',
    ],
    get: (theme, index) => {
        index = Math.abs(index) % enumColors[theme].length;
        return enumColors[theme][index];
    },
};
export var ViewTypes;
(function (ViewTypes) {
    ViewTypes[ViewTypes["FORM"] = 1] = "FORM";
    ViewTypes[ViewTypes["GALLERY"] = 2] = "GALLERY";
    ViewTypes[ViewTypes["GRID"] = 3] = "GRID";
    ViewTypes[ViewTypes["KANBAN"] = 4] = "KANBAN";
    ViewTypes[ViewTypes["MAP"] = 5] = "MAP";
    ViewTypes[ViewTypes["CALENDAR"] = 6] = "CALENDAR";
})(ViewTypes || (ViewTypes = {}));
export const viewTypeAlias = {
    [ViewTypes.FORM]: 'form',
    [ViewTypes.GALLERY]: 'gallery',
    [ViewTypes.GRID]: 'grid',
    [ViewTypes.KANBAN]: 'kanban',
    [ViewTypes.MAP]: 'map',
    [ViewTypes.CALENDAR]: 'calendar',
};
export const viewTypeToStringMap = Object.assign({}, viewTypeAlias);
// Generate reverse mapping from the original viewTypeAlias
export const stringToViewTypeMap = Object.entries(viewTypeAlias).reduce((acc, [key, value]) => {
    acc[value] = Number(key);
    return acc;
}, {});
export var ProjectTypes;
(function (ProjectTypes) {
    ProjectTypes["DATABASE"] = "database";
    ProjectTypes["DOCUMENTATION"] = "documentation";
    ProjectTypes["COWRITER"] = "cowriter";
    ProjectTypes["DASHBOARD"] = "dashboard";
})(ProjectTypes || (ProjectTypes = {}));
export var RelationTypes;
(function (RelationTypes) {
    RelationTypes["HAS_MANY"] = "hm";
    RelationTypes["BELONGS_TO"] = "bt";
    RelationTypes["MANY_TO_MANY"] = "mm";
    RelationTypes["ONE_TO_ONE"] = "oo";
})(RelationTypes || (RelationTypes = {}));
export const ExpandedFormMode = {
    FIELD: 'field',
    ATTACHMENT: 'attachment',
    DISCUSSION: 'discussion',
};
export var ExportTypes;
(function (ExportTypes) {
    ExportTypes["EXCEL"] = "excel";
    ExportTypes["CSV"] = "csv";
})(ExportTypes || (ExportTypes = {}));
export var PluginCategory;
(function (PluginCategory) {
    PluginCategory["STORAGE"] = "Storage";
    PluginCategory["EMAIL"] = "Email";
})(PluginCategory || (PluginCategory = {}));
export var ModelTypes;
(function (ModelTypes) {
    ModelTypes["TABLE"] = "table";
    ModelTypes["VIEW"] = "view";
})(ModelTypes || (ModelTypes = {}));
export var ProjectStatus;
(function (ProjectStatus) {
    ProjectStatus["JOB"] = "job";
})(ProjectStatus || (ProjectStatus = {}));
export var TiptapNodesTypes;
(function (TiptapNodesTypes) {
    TiptapNodesTypes["doc"] = "doc";
    TiptapNodesTypes["sec"] = "sec";
    TiptapNodesTypes["paragraph"] = "paragraph";
    TiptapNodesTypes["text"] = "text";
    TiptapNodesTypes["heading"] = "heading";
    TiptapNodesTypes["bullet"] = "bullet";
    TiptapNodesTypes["ordered"] = "ordered";
    TiptapNodesTypes["task"] = "task";
    TiptapNodesTypes["quote"] = "quote";
    TiptapNodesTypes["divider"] = "divider";
    TiptapNodesTypes["codeBlock"] = "codeBlock";
    TiptapNodesTypes["image"] = "image";
    TiptapNodesTypes["callout"] = "callout";
    TiptapNodesTypes["tipCallout"] = "tipCallout";
    TiptapNodesTypes["table"] = "table";
    TiptapNodesTypes["tableRow"] = "tableRow";
    TiptapNodesTypes["tableCell"] = "tableCell";
    TiptapNodesTypes["embed"] = "embed";
    TiptapNodesTypes["collapsable"] = "collapsable";
    TiptapNodesTypes["collapsableContent"] = "collapsable_content";
    TiptapNodesTypes["collapsableHeader"] = "collapsable_header";
    TiptapNodesTypes["column"] = "column";
    TiptapNodesTypes["columnContent"] = "columnContent";
    TiptapNodesTypes["linkToPage"] = "linkToPage";
    TiptapNodesTypes["attachment"] = "attachment";
})(TiptapNodesTypes || (TiptapNodesTypes = {}));
export var TiptapMarksTypes;
(function (TiptapMarksTypes) {
    TiptapMarksTypes["strike"] = "strike";
    TiptapMarksTypes["bold"] = "bold";
    TiptapMarksTypes["italic"] = "italic";
    TiptapMarksTypes["link"] = "link";
    TiptapMarksTypes["code"] = "code";
    TiptapMarksTypes["underline"] = "underline";
})(TiptapMarksTypes || (TiptapMarksTypes = {}));
export var NcDataErrorCodes;
(function (NcDataErrorCodes) {
    NcDataErrorCodes["NC_ERR_MM_MODEL_NOT_FOUND"] = "NC_ERR_MM_MODEL_NOT_FOUND";
})(NcDataErrorCodes || (NcDataErrorCodes = {}));
export var NcErrorType;
(function (NcErrorType) {
    NcErrorType["AUTHENTICATION_REQUIRED"] = "AUTHENTICATION_REQUIRED";
    NcErrorType["FORBIDDEN"] = "FORBIDDEN";
    NcErrorType["API_TOKEN_NOT_ALLOWED"] = "API_TOKEN_NOT_ALLOWED";
    NcErrorType["WORKSPACE_NOT_FOUND"] = "WORKSPACE_NOT_FOUND";
    NcErrorType["BASE_NOT_FOUND"] = "BASE_NOT_FOUND";
    NcErrorType["SOURCE_NOT_FOUND"] = "SOURCE_NOT_FOUND";
    NcErrorType["TABLE_NOT_FOUND"] = "TABLE_NOT_FOUND";
    NcErrorType["VIEW_NOT_FOUND"] = "VIEW_NOT_FOUND";
    NcErrorType["FIELD_NOT_FOUND"] = "FIELD_NOT_FOUND";
    NcErrorType["RECORD_NOT_FOUND"] = "RECORD_NOT_FOUND";
    NcErrorType["GENERIC_NOT_FOUND"] = "GENERIC_NOT_FOUND";
    NcErrorType["HOOK_NOT_FOUND"] = "HOOK_NOT_FOUND";
    NcErrorType["REQUIRED_FIELD_MISSING"] = "REQUIRED_FIELD_MISSING";
    NcErrorType["ERROR_DUPLICATE_RECORD"] = "ERROR_DUPLICATE_RECORD";
    NcErrorType["USER_NOT_FOUND"] = "USER_NOT_FOUND";
    NcErrorType["INVALID_OFFSET_VALUE"] = "INVALID_OFFSET_VALUE";
    NcErrorType["INVALID_PAGE_VALUE"] = "INVALID_PAGE_VALUE";
    NcErrorType["INVALID_LIMIT_VALUE"] = "INVALID_LIMIT_VALUE";
    NcErrorType["INVALID_FILTER"] = "INVALID_FILTER";
    NcErrorType["INVALID_SHARED_VIEW_PASSWORD"] = "INVALID_SHARED_VIEW_PASSWORD";
    NcErrorType["INVALID_ATTACHMENT_JSON"] = "INVALID_ATTACHMENT_JSON";
    NcErrorType["NOT_IMPLEMENTED"] = "NOT_IMPLEMENTED";
    NcErrorType["INTERNAL_SERVER_ERROR"] = "INTERNAL_SERVER_ERROR";
    NcErrorType["DATABASE_ERROR"] = "DATABASE_ERROR";
    NcErrorType["UNKNOWN_ERROR"] = "UNKNOWN_ERROR";
    NcErrorType["BAD_JSON"] = "BAD_JSON";
    NcErrorType["INVALID_PK_VALUE"] = "INVALID_PK_VALUE";
    NcErrorType["COLUMN_ASSOCIATED_WITH_LINK"] = "COLUMN_ASSOCIATED_WITH_LINK";
    NcErrorType["TABLE_ASSOCIATED_WITH_LINK"] = "TABLE_ASSOCIATED_WITH_LINK";
    NcErrorType["INTEGRATION_NOT_FOUND"] = "INTEGRATION_NOT_FOUND";
    NcErrorType["INTEGRATION_LINKED_WITH_BASES"] = "INTEGRATION_LINKED_WITH_BASES";
    NcErrorType["FORMULA_ERROR"] = "FORMULA_ERROR";
    NcErrorType["FORMULA_CIRCULAR_REF_ERROR"] = "FORMULA_CIRCULAR_REF_ERROR";
    NcErrorType["PERMISSION_DENIED"] = "PERMISSION_DENIED";
    NcErrorType["INVALID_ATTACHMENT_UPLOAD_SCOPE"] = "INVALID_ATTACHMENT_UPLOAD_SCOPE";
    NcErrorType["CANNOT_CALCULATE_INTERMEDIATE_ORDER"] = "CANNOT_CALCULATE_INTERMEDIATE_ORDER";
    NcErrorType["REORDER_FAILED"] = "REORDER_FAILED";
    NcErrorType["PLAN_LIMIT_EXCEEDED"] = "PLAN_LIMIT_EXCEEDED";
    NcErrorType["FEATURE_NOT_SUPPORTED"] = "FEATURE_NOT_SUPPORTED";
    NcErrorType["SSO_LOGIN_REQUIRED"] = "SSO_LOGIN_REQUIRED";
    NcErrorType["SSO_GENERATED_TOKEN_REQUIRED"] = "SSO_GENERATED_TOKEN_REQUIRED";
    NcErrorType["MAX_INSERT_LIMIT_EXCEEDED"] = "MAX_INSERT_LIMIT_EXCEEDED";
    NcErrorType["INVALID_VALUE_FOR_FIELD"] = "INVALID_VALUE_FOR_FIELD";
    NcErrorType["MAX_WORKSPACE_LIMIT_REACHED"] = "MAX_WORKSPACE_LIMIT_REACHED";
    NcErrorType["BASE_USER_ERROR"] = "BASE_USER_ERROR";
    NcErrorType["PROHIBITED_SYNC_TABLE_OPERATION"] = "PROHIBITED_SYNC_TABLE_OPERATION";
    NcErrorType["INVALID_REQUEST_BODY"] = "INVALID_REQUEST_BODY";
    NcErrorType["DASHBOARD_NOT_FOUND"] = "DASHBOARD_NOT_FOUND";
    NcErrorType["WIDGET_NOT_FOUND"] = "WIDGET_NOT_FOUND";
    NcErrorType["INVALID_SHARED_DASHBOARD_PASSWORD"] = "INVALID_SHARED_DASHBOARD_PASSWORD";
})(NcErrorType || (NcErrorType = {}));
export var ROW_COLORING_MODE;
(function (ROW_COLORING_MODE) {
    ROW_COLORING_MODE["FILTER"] = "filter";
    ROW_COLORING_MODE["SELECT"] = "select";
})(ROW_COLORING_MODE || (ROW_COLORING_MODE = {}));
export const LongTextAiMetaProp = 'ai';
export const NO_SCOPE = 'nc';
export const NON_SEAT_ROLES = [
    WorkspaceUserRoles.NO_ACCESS,
    WorkspaceUserRoles.VIEWER,
    WorkspaceUserRoles.COMMENTER,
    ProjectRoles.NO_ACCESS,
    ProjectRoles.VIEWER,
    ProjectRoles.COMMENTER,
];
export const DURATION_TYPE_MAP = {
    0: 'h:mm',
    1: 'h:mm:ss',
    2: 'h:mm:ss.s',
    3: 'h:mm:ss.ss',
    4: 'h:mm:ss.sss',
    'h:mm': 0,
    'h:mm:ss': 1,
    'h:mm:ss.s': 2,
    'h:mm:ss.ss': 3,
    'h:mm:ss.sss': 4,
};
export const CURRENT_USER_TOKEN = '@me';
