// nocodb-sdk/build/module/lib/numberUtils.js
import { parseProp, roundUpToPrecision } from '../lib/helperFunctions';
export const numberize = (value) => {
    if (value === undefined || value === null) {
        return value;
    }
    if (typeof value === 'number') {
        return value;
    }
    else {
        const result = parseInt(value);
        if (isNaN(result)) {
            return undefined;
        }
        return result;
    }
};
export const roundTo = (num, precision = 1) => {
    if (!num || Number.isNaN(num))
        return num;
    const factor = Math.pow(10, precision);
    return Math.round(+num * factor) / factor;
};
export const getCurrencyValue = (modelValue, col) => {
    var _a, _b, _c;
    const currencyMeta = Object.assign({ currency_locale: 'en-US', currency_code: 'USD', precision: 2 }, parseProp(col.meta));
    try {
        if (modelValue === null ||
            modelValue === undefined ||
            Number.isNaN(modelValue)) {
            return modelValue === null || modelValue === undefined
                ? ''
                : modelValue;
        }
        const roundedValue = roundUpToPrecision(Number(modelValue), (_a = currencyMeta.precision) !== null && _a !== void 0 ? _a : 2);
        return new Intl.NumberFormat(currencyMeta.currency_locale || 'en-US', {
            style: 'currency',
            currency: currencyMeta.currency_code || 'USD',
            minimumFractionDigits: (_b = currencyMeta.precision) !== null && _b !== void 0 ? _b : 2,
            maximumFractionDigits: (_c = currencyMeta.precision) !== null && _c !== void 0 ? _c : 2,
        }).format(+roundedValue);
    }
    catch (e) {
        return modelValue;
    }
};
