// nocodb-sdk/build/module/lib/stringHelpers.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { ncIsUndefined } from './is';
const AppendToLengthSuffixConfig = {
    _: {
        replacement: '___{index}',
        replacementRegex: /___\s?(\d*)$/,
    },
    dot: {
        replacement: '...{index}',
        replacementRegex: /\.\.\.\s?(\d*)$/,
    },
};
export var AppendToLengthSuffix;
(function (AppendToLengthSuffix) {
    AppendToLengthSuffix["_"] = "_";
    AppendToLengthSuffix["dot"] = "dot";
})(AppendToLengthSuffix || (AppendToLengthSuffix = {}));
/**
 * get a row & column given a numeric position
 * row & column start at 0
 */
export function getRowColPositionFromIndex({ stack, position, }) {
    const parts = stack.substring(0, position).split('\n');
    return {
        column: parts[parts.length - 1].length,
        row: parts.length - 1,
    };
}
export function appendToLength(param) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        const { value, appendage, maxLength, isExists } = param;
        const suffixConfig = AppendToLengthSuffixConfig[(_a = param.suffix) !== null && _a !== void 0 ? _a : AppendToLengthSuffix.dot];
        // if it's already in the form of truncated
        // skip with usual append
        if ((value + appendage).length > maxLength) {
            const existingIndex = (_b = value.match(suffixConfig.replacementRegex)) === null || _b === void 0 ? void 0 : _b[1];
            return truncateToLength({
                value,
                maxLength,
                currentIndex: !ncIsUndefined(existingIndex) && existingIndex !== ''
                    ? Number(existingIndex)
                    : undefined,
                isExists,
                suffix: param.suffix,
            });
        }
        let currentIndex = 1;
        let needle = value + appendage;
        while (yield isExists(needle)) {
            needle = value + appendage + '_' + currentIndex++;
            if (needle.length >= maxLength) {
                return appendToLength(Object.assign(Object.assign({}, param), { appendage: appendage + '_' + currentIndex++, suffix: param.suffix }));
            }
        }
        return needle;
    });
}
export function truncateToLength(param) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        const { value, currentIndex, maxLength, isExists } = param;
        const suffixConfig = AppendToLengthSuffixConfig[(_a = param.suffix) !== null && _a !== void 0 ? _a : AppendToLengthSuffix.dot];
        const replacement = ncIsUndefined(currentIndex) ? '' : `${currentIndex}`;
        const suffix = suffixConfig.replacement.replace('{index}', replacement);
        const needle = value.substring(0, maxLength - suffix.length) + suffix;
        if (!(yield isExists(needle))) {
            return needle;
        }
        else {
            return truncateToLength(Object.assign(Object.assign({}, param), { currentIndex: ncIsUndefined(currentIndex) ? 2 : currentIndex + 1 }));
        }
    });
}
/**
 * Generates a unique copy name by checking against existing names/items
 */
export function generateUniqueCopyName(originalName, existing, options = {}) {
    const { accessor, prefix = 'Copy of', separator = ' ', counterFormat = '({counter})', } = options;
    // Extract names from the existing array
    const existingNames = existing.map((item) => {
        if (typeof item === 'string')
            return item;
        if (accessor) {
            return typeof accessor === 'function'
                ? accessor(item)
                : String(item[accessor]);
        }
        // Default to 'title' property if no accessor specified
        return String(item.title);
    });
    let newName = `${prefix} ${originalName}`;
    let counter = 1;
    while (existingNames.includes(newName)) {
        const counterText = counterFormat.replace('{counter}', counter.toString());
        newName = `${prefix} ${originalName}${separator}${counterText}`;
        counter++;
    }
    return newName;
}
export function escapeRegexString(string) {
    return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
