// nocodb-sdk/build/module/lib/error-handler/nc-error-codex-manager.js
import { NcBaseErrorv2 } from '../../lib/error/nc-base.error';
import { presetErrorCodexMap } from '../../lib/error-handler/preset-error-codex-map';
export class NcErrorCodexManager {
    constructor() {
        this.errorCodexMap = {};
        this.setErrorCodexes(presetErrorCodexMap);
    }
    setErrorCodex(errorType, handler) {
        this.errorCodexMap[errorType] = handler;
    }
    setErrorCodexes(handlers) {
        this.errorCodexMap = Object.assign(Object.assign({}, this.errorCodexMap), handlers);
    }
    generateError(error, args) {
        const errorHelper = this.errorCodexMap[error];
        const { params, customMessage, details } = args || {};
        if (!errorHelper) {
            return {
                message: 'An error occurred',
                code: 500,
                details: details,
            };
        }
        let message;
        const messageHelper = customMessage || errorHelper.message;
        if (typeof messageHelper === 'function') {
            message = messageHelper(...(Array.isArray(params) ? params : [params]));
        }
        else {
            message = messageHelper;
        }
        return new NcBaseErrorv2(message, errorHelper.code, error, {
            details: details,
        });
    }
}
