// nocodb-sdk/build/main/lib/error-handler/nc-error-codex-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NcErrorCodexManager = void 0;
const nc_base_error_1 = require("../../lib/error/nc-base.error");
const preset_error_codex_map_1 = require("../../lib/error-handler/preset-error-codex-map");
class NcErrorCodexManager {
    constructor() {
        this.errorCodexMap = {};
        this.setErrorCodexes(preset_error_codex_map_1.presetErrorCodexMap);
    }
    setErrorCodex(errorType, handler) {
        this.errorCodexMap[errorType] = handler;
    }
    setErrorCodexes(handlers) {
        this.errorCodexMap = Object.assign(Object.assign({}, this.errorCodexMap), handlers);
    }
    generateError(error, args) {
        const errorHelper = this.errorCodexMap[error];
        const { params, customMessage, details } = args || {};
        if (!errorHelper) {
            return {
                message: 'An error occurred',
                code: 500,
                details: details,
            };
        }
        let message;
        const messageHelper = customMessage || errorHelper.message;
        if (typeof messageHelper === 'function') {
            message = messageHelper(...(Array.isArray(params) ? params : [params]));
        }
        else {
            message = messageHelper;
        }
        return new nc_base_error_1.NcBaseErrorv2(message, errorHelper.code, error, {
            details: details,
        });
    }
}
exports.NcErrorCodexManager = NcErrorCodexManager;
