// nocodb-sdk/build/main/lib/formBuilder/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FORM_BUILDER_NON_CATEGORIZED = exports.FormBuilderValidatorType = exports.FormBuilderInputType = void 0;
var FormBuilderInputType;
(function (FormBuilderInputType) {
    FormBuilderInputType["Input"] = "input";
    FormBuilderInputType["Select"] = "select";
    FormBuilderInputType["Switch"] = "switch";
    FormBuilderInputType["Space"] = "space";
    FormBuilderInputType["Password"] = "password";
    FormBuilderInputType["SelectIntegration"] = "integration";
    FormBuilderInputType["SelectBase"] = "select-base";
    FormBuilderInputType["OAuth"] = "oauth";
})(FormBuilderInputType || (exports.FormBuilderInputType = FormBuilderInputType = {}));
var FormBuilderValidatorType;
(function (FormBuilderValidatorType) {
    FormBuilderValidatorType["Required"] = "required";
})(FormBuilderValidatorType || (exports.FormBuilderValidatorType = FormBuilderValidatorType = {}));
exports.FORM_BUILDER_NON_CATEGORIZED = 'form-builder-non-categorized';
