// nocodb-sdk/build/module/lib/parser/queryFilter/query-filter-cst-parser.js
import { InvalidFilterError } from '../../../lib/error/invalid-filter.error';
import { parseVariableAsArray, parseVariableAsString, } from '../common-cst-parser';
export const parseExpressionArguments = (cst) => {
    if (cst.children.VARIABLE) {
        return parseVariableAsArray(cst.children.VARIABLE);
    }
    return undefined;
};
export const parseMultiClause = (cst, opt) => {
    var _a;
    const result = {
        is_group: true,
        logical_op: (_a = opt === null || opt === void 0 ? void 0 : opt.logicalOperator) !== null && _a !== void 0 ? _a : 'and',
        children: [],
    };
    for (let index = 0; index < cst.children.clause.length; index++) {
        const clause = cst.children.clause[index];
        result.children.push(clause.name === 'and_or_clause'
            ? parseAndOrClause(clause)
            : clause.name === 'not_clause'
                ? parseNotClause(clause)
                : parseParenClause(clause));
    }
    return result;
};
export const parseNotClause = (cst) => {
    return parseParenClause(cst.children.clause[0], {
        logicalOperator: 'not',
    });
};
export const parseAndOrClause = (cst) => {
    if (!['~and', '~or'].includes(cst.children.operator[0].image)) {
        throw new InvalidFilterError({
            message: `Invalid filter expression. Expected a valid logical operator like '~or' or '~and', but found '${cst.children.operator[0].image}'`,
        });
    }
    return parseParenClause(cst.children.clause[0], {
        logicalOperator: cst.children.operator[0].image.replace('~', ''),
    });
};
export const parseParenClause = (cst, opt) => {
    const clause = cst.children.clause[0];
    if (clause.name === 'multi_clause') {
        return parseMultiClause(clause, opt);
    }
    else {
        return parseCallExpression(clause, opt);
    }
};
export const parseCallExpression = (cst, opt) => {
    const operator = cst.children.OPERATOR[0].image;
    const result = {
        is_group: false,
        field: parseVariableAsString(cst.children.VARIABLE),
        comparison_op: operator,
        logical_op: opt === null || opt === void 0 ? void 0 : opt.logicalOperator,
    };
    if (cst.children.expression_arguments &&
        cst.children.expression_arguments[0]) {
        const variables = parseExpressionArguments(cst.children.expression_arguments[0]);
        result.value = variables;
    }
    handleBlankOperator(result);
    handleInOperator(result);
    handleOperatorAndValue(result);
    return result;
};
const handleBlankOperator = (filter) => {
    var _a, _b;
    switch (filter.comparison_op) {
        case 'is':
            if (((_a = filter.value) === null || _a === void 0 ? void 0 : _a[0]) === 'blank') {
                filter.comparison_op = 'blank';
                filter.value = undefined;
            }
            else if (((_b = filter.value) === null || _b === void 0 ? void 0 : _b[0]) === 'notblank') {
                filter.comparison_op = 'notblank';
                filter.value = undefined;
            }
            break;
        case 'isblank':
        case 'is_blank':
            filter.comparison_op = 'blank';
            break;
        case 'isnotblank':
        case 'is_not_blank':
        case 'is_notblank':
            filter.comparison_op = 'notblank';
            break;
    }
};
const handleInOperator = (filter) => {
    var _a;
    if (filter.comparison_op === 'in' && !Array.isArray(filter.value)) {
        filter.value = (_a = filter.value) === null || _a === void 0 ? void 0 : _a.split(',');
    }
};
const handleOperatorAndValue = (filter) => {
    if (Array.isArray(filter.value) &&
        [
            'eq',
            'neq',
            'not',
            'like',
            'nlike',
            'empty',
            'notempty',
            'null',
            'notnull',
            'checked',
            'notchecked',
            'blank',
            'notblank',
            'allof',
            'anyof',
            'nallof',
            'nanyof',
            'gt',
            'lt',
            'gte',
            'lte',
            'ge',
            'le',
            'isnot',
            'is',
            'gb_eq',
        ].includes(filter.comparison_op)) {
        if (filter.value.length === filter.value.filter((k) => k === null).length) {
            filter.value = null;
        }
        else {
            filter.value = filter.value.filter((k) => k).join(',');
        }
    }
    // for equality, replace with empty string if value is undefined
    else if (filter.value === undefined &&
        ['eq', 'neq', 'gb_eq'].includes(filter.comparison_op)) {
        filter.value = '';
    }
};
export const parseCst = (cst) => {
    return parseMultiClause(cst);
};
