// nocodb-sdk/build/main/lib/formula/hooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsepCurlyHook = exports.jsepIndexHook = void 0;
// opening and closing string code
const OCURLY_CODE = 123; // '{'
const CCURLY_CODE = 125; // '}'
exports.jsepIndexHook = {
    name: 'indexing',
    init(jsep) {
        // Match identifier in following pattern: {abc-cde}
        jsep.hooks.add('after-expression', function escapedIdentifier(env) {
            var _a, _b, _c;
            if (!env.node) {
                return;
            }
            // search needle, either the callee name, identifier raw, etc
            const needle = (_b = (_a = env.node.name) !== null && _a !== void 0 ? _a : env.node.raw) !== null && _b !== void 0 ? _b : (_c = env.node.callee) === null || _c === void 0 ? void 0 : _c.name;
            // current token index
            let lastIndex = this.index;
            if (env.node.arguments) {
                // because arguments are always at the right side, we start from earliest argument
                lastIndex = env.node.arguments
                    .map((k) => k.indexStart)
                    .filter((k) => k)
                    .reduce((acc, cur) => (acc < cur ? acc : cur), lastIndex);
            }
            // get the last index of needle
            const startIndex = env.context.expr.lastIndexOf(needle, lastIndex);
            env.node.indexStart = startIndex;
            env.node.nodeLength = lastIndex - startIndex;
            env.node.indexEnd = this.index;
            return env.node;
        });
    },
};
exports.jsepCurlyHook = {
    name: 'curly',
    init(jsep) {
        // Match identifier in following pattern: {abc-cde}
        jsep.hooks.add('gobble-token', function escapedIdentifier(env) {
            // check if the current token is an opening curly bracket
            if (this.code === OCURLY_CODE) {
                const patternIndex = this.index;
                // move to the next character until we find a closing curly bracket
                while (this.index < this.expr.length) {
                    ++this.index;
                    if (this.code === CCURLY_CODE) {
                        let identifier = this.expr.slice(patternIndex, ++this.index);
                        // if starting with double curley brace then check for ending double curley brace
                        // if found include with the identifier
                        if (identifier.startsWith('{{') &&
                            this.expr.slice(patternIndex, this.index + 1).endsWith('}')) {
                            identifier = this.expr.slice(patternIndex, ++this.index);
                        }
                        env.node = {
                            type: jsep.IDENTIFIER,
                            name: /^{{.*}}$/.test(identifier)
                                ? // start would be the position of the first curly bracket
                                    // add 2 to point to the first character for expressions like {{col1}}
                                    identifier.slice(2, -2)
                                : // start would be the position of the first curly bracket
                                    // add 1 to point to the first character for expressions like {col1}
                                    identifier.slice(1, -1),
                            raw: identifier,
                        };
                        // env.node = this.gobbleTokenProperty(env.node);
                        return env.node;
                    }
                }
                this.throwError('Unclosed }');
            }
        });
    },
};
