// nocodb-sdk/build/module/lib/columnHelper/columns/LTAR/index.js
import AbstractColumnHelper from '../../column.interface';
import { isBt, isHm, isMm, isOo } from '../../utils';
import { DefaultColumnHelper } from '../DefaultColumnHelper';
import { BelongsToHelper } from './BelongsTo';
import { HasManyHelper } from './HasMany';
import { ManyToManyHelper } from './ManyToMany';
import { OneToOneHelper } from './OneToOne';
export class LTARHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {};
    }
    getLtarHelperColumn(params) {
        let columnHelper = DefaultColumnHelper;
        if (isHm(params.col))
            columnHelper = HasManyHelper;
        if (isMm(params.col))
            columnHelper = ManyToManyHelper;
        if (isBt(params.col))
            columnHelper = BelongsToHelper;
        if (isOo(params.col))
            columnHelper = OneToOneHelper;
        return new columnHelper();
    }
    serializeValue(value, params) {
        return this.getLtarHelperColumn(params).serializeValue(value, params);
    }
    parseValue(value, params) {
        return this.getLtarHelperColumn(params).parseValue(value, params);
    }
    parsePlainCellValue(value, params) {
        return this.getLtarHelperColumn(params).parsePlainCellValue(value, params);
    }
}
export { BelongsToHelper, HasManyHelper, ManyToManyHelper, OneToOneHelper };
