// nocodb-sdk/build/module/lib/columnHelper/columns/Date.js
import dayjs from 'dayjs';
import { dateFormats } from '../../../lib/dateTimeHelper';
import { DATE_DIFF_TO_SCALE_LABEL_MAP, DATE_SCALE_LABEL_TO_DIFF_MAP, isNumberRound, parseDateValue, serializeDateOrDateTimeValue, } from '..';
import AbstractColumnHelper from '../column.interface';
import { SilentTypeConversionError } from '../../../lib/error';
import { parseProp } from '../../../lib/helperFunctions';
import { ncIsNullOrUndefined, ncIsUndefined } from '../../../lib/is';
import { populateFillHandleStrictCopy } from '../utils/fill-handler';
export class DateHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            date_format: dateFormats[0],
        };
    }
    serializeValue(value, params) {
        if (params.serializeSearchQuery)
            return null;
        value = serializeDateOrDateTimeValue(value, params.col);
        if (value === null) {
            if (params.isMultipleCellPaste) {
                return null;
            }
            else {
                throw new SilentTypeConversionError();
            }
        }
        return value;
    }
    parseValue(value, params) {
        return parseDateValue(value, params.col, params.isSystemCol);
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = parseDateValue(value, params.col, params.isSystemCol)) !== null && _a !== void 0 ? _a : '';
    }
    populateFillHandle(params) {
        var _a;
        const { column, highlightedData, numberOfRows } = params;
        if (highlightedData.length < 2) {
            return populateFillHandleStrictCopy(params);
        }
        // data is in form like 'YYYY-MM-DD' depends on meta dateformat
        const meta = parseProp(column.meta);
        const metaDateFormat = (_a = meta.date_format) !== null && _a !== void 0 ? _a : 'YYYY-MM-DD';
        const dateFormat = metaDateFormat + ' HH:mm:ss';
        let lastData;
        let modifier = undefined;
        let scale = 'D';
        const setModifierForScaleIfRound = (currentData, lastData, diffScale) => {
            const currentModifier = currentData.diff(lastData, diffScale, true);
            if (isNumberRound(currentModifier)) {
                scale = DATE_DIFF_TO_SCALE_LABEL_MAP[diffScale];
                modifier = currentModifier;
            }
        };
        // map to dayjs
        const dayJsHighlightedData = [];
        for (const date of highlightedData) {
            if (ncIsNullOrUndefined(date) ||
                date === '' ||
                typeof date !== 'string') {
                return populateFillHandleStrictCopy(params);
            }
            const currentData = dayjs(date + ' 00:00:00', dateFormat);
            // unlikely on normal case
            if (!currentData.isValid()) {
                return populateFillHandleStrictCopy(params);
            }
            if (!lastData) {
                lastData = currentData;
            }
            else {
                if (ncIsUndefined(modifier)) {
                    const modifierInDays = currentData.diff(lastData, 'day');
                    if (modifierInDays > 354) {
                        setModifierForScaleIfRound(currentData, lastData, 'year');
                    }
                    else if (modifierInDays > 28) {
                        setModifierForScaleIfRound(currentData, lastData, 'month');
                    }
                    else {
                        modifier = modifierInDays;
                        scale = 'D';
                    }
                }
                else {
                    const currentModifier = currentData.diff(lastData, DATE_SCALE_LABEL_TO_DIFF_MAP[scale]);
                    if (currentModifier !== modifier) {
                        return populateFillHandleStrictCopy(params);
                    }
                }
                lastData = currentData;
            }
            dayJsHighlightedData.push(lastData);
        }
        if (modifier === 0) {
            return populateFillHandleStrictCopy(params);
        }
        const numberToGenerate = numberOfRows - highlightedData.length;
        return Array.from({ length: numberToGenerate }).map(() => {
            const currentData = lastData.add(modifier, DATE_SCALE_LABEL_TO_DIFF_MAP[scale]);
            lastData = currentData;
            return currentData.format(metaDateFormat);
        });
    }
}
