// nocodb-sdk/build/module/lib/columnHelper/columns/Duration.js
import { SilentTypeConversionError } from '../../../lib/error';
import { parseDurationValue, serializeDurationValue } from '..';
import AbstractColumnHelper from '../column.interface';
import { ncIsNaN } from '../../../lib/is';
export class DurationHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            duration: 0,
        };
    }
    serializeValue(value, params) {
        value = serializeDurationValue(value, params.col);
        if (value === null) {
            if (params.isMultipleCellPaste || params.serializeSearchQuery) {
                return null;
            }
            else {
                throw new SilentTypeConversionError();
            }
        }
        return value;
    }
    parseValue(value, params) {
        return parseDurationValue(value, params.col);
    }
    parsePlainCellValue(value, params) {
        var _a;
        if (params.isAggregation && ncIsNaN(value)) {
            value = 0;
        }
        return (_a = parseDurationValue(value, params.col)) !== null && _a !== void 0 ? _a : '';
    }
}
