// nocodb-sdk/build/module/lib/columnHelper/columns/Formula.js
import AbstractColumnHelper from '../column.interface';
import { parseProp } from '../../../lib/helperFunctions';
import { ColumnHelper } from '../column-helper';
import { ComputedTypePasteError } from '../../../lib/error';
import { FormulaDataTypes } from '../../../lib/formulaHelpers';
import { ncIsNaN } from '../../../lib/is';
export class FormulaHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            display_column_meta: {
                meta: {},
                custom: {},
            },
            display_type: null,
        };
    }
    serializeValue(value, params) {
        var _a, _b, _c, _d;
        if (params.serializeSearchQuery) {
            const dataType = (_d = (_c = (_b = (_a = params.col) === null || _a === void 0 ? void 0 : _a.colOptions) === null || _b === void 0 ? void 0 : _b.parsed_tree) === null || _c === void 0 ? void 0 : _c.dataType) !== null && _d !== void 0 ? _d : FormulaDataTypes.STRING;
            if (dataType === FormulaDataTypes.NUMERIC) {
                return ncIsNaN(value) ? '' : value;
            }
            return value;
        }
        if (params.isMultipleCellPaste) {
            return undefined;
        }
        else {
            throw new ComputedTypePasteError();
        }
    }
    parseValue(value, params) {
        var _a;
        const columnMeta = parseProp((_a = params.col) === null || _a === void 0 ? void 0 : _a.meta);
        const childColumn = Object.assign({ uidt: columnMeta.display_type }, columnMeta.display_column_meta);
        return ColumnHelper.parseValue(value, Object.assign(Object.assign({}, params), { col: childColumn }));
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = this.parseValue(value, params)) !== null && _a !== void 0 ? _a : '';
    }
}
