// nocodb-sdk/build/module/lib/columnHelper/columns/Json.js
import AbstractColumnHelper from '../../../lib/columnHelper/column.interface';
import { SilentTypeConversionError } from '../../../lib/error';
import { parseJsonValue, serializeJsonValue } from '../utils';
import { ncIsString } from '../../../lib/is';
export class JsonHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {};
    }
    serializeValue(value, params) {
        if (params.serializeSearchQuery) {
            return ncIsString(value) ? value : null;
        }
        value = serializeJsonValue(value);
        if (value === null) {
            throw new SilentTypeConversionError();
        }
        return value;
    }
    parseValue(value) {
        return parseJsonValue(value);
    }
    parsePlainCellValue(value) {
        return parseJsonValue(value);
    }
    equalityComparison(a, b, _param) {
        const aStr = typeof a === 'object'
            ? JSON.stringify(a)
            : typeof a === 'string'
                ? a
                : a.toString();
        const bStr = typeof b === 'object'
            ? JSON.stringify(b)
            : typeof b === 'string'
                ? b
                : b.toString();
        return aStr === bStr;
    }
}
