// nocodb-sdk/build/module/lib/columnHelper/columns/Links.js
import { SilentTypeConversionError } from '../../../lib/error';
import AbstractColumnHelper from '../column.interface';
import { isBt, isMm, isOo, parseLinksValue } from '../utils';
import { ncIsNaN, ncIsObject } from '../../../lib/is';
export class LinksHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {};
    }
    serializeValue(value, params) {
        var _a;
        if (params.serializeSearchQuery)
            return null;
        if (!isMm(params.col))
            throw new SilentTypeConversionError();
        let parsedVal = value;
        try {
            parsedVal = typeof value === 'string' ? JSON.parse(value) : value;
        }
        catch (_b) { }
        if (!(parsedVal &&
            ncIsObject(parsedVal) &&
            ['rowId', 'columnId', 'fk_related_model_id', 'value'].every((key) => parsedVal.hasOwnProperty(key))) ||
            (parsedVal === null || parsedVal === void 0 ? void 0 : parsedVal.fk_related_model_id) !==
                ((_a = params.col.colOptions) === null || _a === void 0 ? void 0 : _a.fk_related_model_id)) {
            throw new SilentTypeConversionError();
        }
        return parsedVal;
    }
    parseValue(value, params) {
        if (isMm(params.col)) {
            return JSON.stringify({
                rowId: params.rowId,
                columnId: params.col.id,
                fk_related_model_id: params.col.colOptions
                    .fk_related_model_id,
                value: !ncIsNaN(value) ? +value : 0,
            });
        }
        else if (isBt(params.col) || isOo(params.col)) {
            // fk_related_model_id is used to prevent paste operation in different fk_related_model_id cell
            return JSON.stringify({
                fk_related_model_id: params.col.colOptions
                    .fk_related_model_id,
                value: value || null,
            });
        }
        return value !== null && value !== void 0 ? value : '';
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = parseLinksValue(value, params)) !== null && _a !== void 0 ? _a : '';
    }
}
