// nocodb-sdk/build/module/lib/columnHelper/columns/Rollup.js
import AbstractColumnHelper from '../column.interface';
import { ColumnHelper } from '../column-helper';
import { getRenderAsTextFunForUiType, parseProp } from '../../../lib/helperFunctions';
import UITypes from '../../../lib/UITypes';
import { ComputedTypePasteError } from '../../../lib/error';
import { precisionFormats } from '../utils';
import { isValidValue } from '../../../lib/is';
export class RollupHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            precision: precisionFormats[0],
            isLocaleString: false,
        };
    }
    serializeValue(_value, params) {
        if (params.isMultipleCellPaste || params.serializeSearchQuery) {
            return undefined;
        }
        else {
            throw new ComputedTypePasteError();
        }
    }
    parseValue(value, params) {
        var _a, _b, _c;
        if (!isValidValue(value))
            return null;
        const { col, meta, metas } = params;
        const colOptions = col.colOptions;
        const relationColumnOptions = colOptions.fk_relation_column_id
            ? (_b = (_a = meta === null || meta === void 0 ? void 0 : meta.columns) === null || _a === void 0 ? void 0 : _a.find((c) => c.id === colOptions.fk_relation_column_id)) === null || _b === void 0 ? void 0 : _b.colOptions
            : null;
        const relatedTableMeta = (relationColumnOptions === null || relationColumnOptions === void 0 ? void 0 : relationColumnOptions.fk_related_model_id) &&
            (metas === null || metas === void 0 ? void 0 : metas[relationColumnOptions.fk_related_model_id]);
        const childColumn = relatedTableMeta === null || relatedTableMeta === void 0 ? void 0 : relatedTableMeta.columns.find((c) => c.id === colOptions.fk_rollup_column_id);
        if (!childColumn)
            return value;
        const renderAsTextFun = getRenderAsTextFunForUiType(((_c = childColumn.uidt) !== null && _c !== void 0 ? _c : UITypes.SingleLineText));
        childColumn.meta = Object.assign(Object.assign({}, parseProp(childColumn === null || childColumn === void 0 ? void 0 : childColumn.meta)), parseProp(col === null || col === void 0 ? void 0 : col.meta));
        if (renderAsTextFun.includes(colOptions.rollup_function)) {
            childColumn.uidt = UITypes.Decimal;
        }
        return ColumnHelper.parseValue(value, Object.assign(Object.assign({}, params), { col: childColumn }));
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = this.parseValue(value, params)) !== null && _a !== void 0 ? _a : '';
    }
}
