// nocodb-sdk/build/module/lib/columnHelper/columns/Time.js
import dayjs from 'dayjs';
import { constructTimeFormat } from '../../../lib/dateTimeHelper';
import { SilentTypeConversionError } from '../../../lib/error';
import AbstractColumnHelper from '../column.interface';
import { parseTimeValue, serializeTimeValue } from '../utils';
export class TimeHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            is12hrFormat: false,
        };
    }
    serializeValue(value, params) {
        if (params.serializeSearchQuery) {
            return this.parseValue(value, params);
        }
        value = serializeTimeValue(value, params);
        if (value === null) {
            if (params.isMultipleCellPaste || params.serializeSearchQuery) {
                return null;
            }
            else {
                throw new SilentTypeConversionError();
            }
        }
        return value;
    }
    parseValue(value, params) {
        return parseTimeValue(value, params);
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = parseTimeValue(value, params)) !== null && _a !== void 0 ? _a : '';
    }
    equalityComparison(a, b, param) {
        const aDayjs = typeof a === 'string' ? dayjs(serializeTimeValue(a, param)) : dayjs(a);
        const bDayjs = typeof b === 'string' ? dayjs(serializeTimeValue(b, param)) : dayjs(b);
        return (aDayjs.format(constructTimeFormat(param.col)) ===
            bDayjs.format(constructTimeFormat(param.col)));
    }
}
