// nocodb-sdk/build/module/lib/columnHelper/columns/User.js
import { SilentTypeConversionError } from '../../../lib/error';
import { parseProp } from '../../../lib/helperFunctions';
import { ncIsArray, ncIsObject, ncIsString } from '../../../lib/is';
import AbstractColumnHelper from '../column.interface';
import { parseUserValue, serializeEmail, serializeStringValue } from '../utils';
export class UserHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            is_multi: false,
            notify: false,
        };
    }
    serializeValue(value, params) {
        if (params.serializeSearchQuery) {
            return serializeStringValue(value);
        }
        try {
            value = typeof value === 'string' ? JSON.parse(value) : value;
        }
        catch (_a) { }
        if (value === null) {
            if (params.isMultipleCellPaste) {
                return null;
            }
            else {
                throw new SilentTypeConversionError();
            }
        }
        const is_multi = parseProp(params.col.meta).is_multi;
        if (ncIsArray(value) || ncIsObject(value)) {
            if (ncIsObject(value)) {
                value = [value];
            }
            value = value.reduce((acc, user) => {
                if (user.email) {
                    acc.push(user);
                }
                return acc;
            }, []);
            if (!value.length)
                return null;
            if (!is_multi) {
                return value[0];
            }
            return value;
        }
        if (!ncIsString(value))
            return null;
        if (is_multi) {
            return value
                .split(',')
                .map((val) => serializeEmail(val))
                .filter(Boolean)
                .join(', ');
        }
        return serializeEmail(value);
    }
    parseValue(value) {
        return parseUserValue(value, true);
    }
    parsePlainCellValue(value) {
        var _a;
        return (_a = parseUserValue(value)) !== null && _a !== void 0 ? _a : '';
    }
}
