// nocodb-sdk/build/module/lib/columnHelper/columns/Year.js
import { SilentTypeConversionError } from '../../../lib/error';
import AbstractColumnHelper from '../column.interface';
import { parseYearValue, serializeYearValue } from '../utils';
import { populateFillHandleStringNumber } from '../utils/fill-handler';
export class YearHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {};
    }
    serializeValue(value, params) {
        if (params.serializeSearchQuery) {
            return this.parseValue(value);
        }
        value = serializeYearValue(value);
        if (value === null) {
            if (params.isMultipleCellPaste) {
                return null;
            }
            else {
                throw new SilentTypeConversionError();
            }
        }
        return value;
    }
    parseValue(value) {
        return parseYearValue(value);
    }
    parsePlainCellValue(value) {
        var _a;
        return `${(_a = parseYearValue(value)) !== null && _a !== void 0 ? _a : ''}`;
    }
    // using string number fill handler
    populateFillHandle(params) {
        return populateFillHandleStringNumber(params);
    }
}
