// nocodb-sdk/build/main/lib/columnHelper/utils/cell.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNumericFieldType = exports.renderValue = exports.isPrimaryKey = exports.isPrimary = exports.isManualSaved = exports.isAutoSaved = exports.isAI = exports.isScriptButton = exports.isAiButton = exports.isButton = exports.isMultiUser = exports.isUser = exports.isGeometry = exports.isSpecificDBType = exports.isPercent = exports.isGeoData = exports.isDuration = exports.isDecimal = exports.isPhoneNumber = exports.isCurrency = exports.isRating = exports.isAttachment = exports.isEmail = exports.isURL = exports.isMultiSelect = exports.isSet = exports.isSingleSelect = exports.isEnum = exports.isJSON = exports.isReadonlyUser = exports.isReadonlyDateTime = exports.isDateTime = exports.isTime = exports.isYear = exports.isDate = exports.isFloat = exports.isInt = exports.isRichText = exports.isTextArea = exports.isString = exports.isBoolean = exports.dataTypeLow = void 0;
const dayjs_1 = __importDefault(require("dayjs"));
const Api_1 = require("../../../lib/Api");
const globals_1 = require("../../../lib/globals");
const helperFunctions_1 = require("../../../lib/helperFunctions");
const UITypes_1 = __importDefault(require("../../../lib/UITypes"));
const dataTypeLow = (column) => { var _a; return (_a = column.dt) === null || _a === void 0 ? void 0 : _a.toLowerCase(); };
exports.dataTypeLow = dataTypeLow;
const isBoolean = (column, abstractType) => column.uidt === UITypes_1.default.Checkbox || abstractType === 'boolean';
exports.isBoolean = isBoolean;
const isString = (column, abstractType) => column.uidt === UITypes_1.default.SingleLineText || abstractType === 'string';
exports.isString = isString;
const isTextArea = (column) => column.uidt === UITypes_1.default.LongText;
exports.isTextArea = isTextArea;
const isRichText = (column) => column.uidt === UITypes_1.default.LongText && !!(0, helperFunctions_1.parseProp)(column === null || column === void 0 ? void 0 : column.meta).richMode;
exports.isRichText = isRichText;
const isInt = (_column, abstractType) => abstractType === 'integer';
exports.isInt = isInt;
const isFloat = (column, abstractType) => abstractType === 'float' || column.uidt === UITypes_1.default.Number;
exports.isFloat = isFloat;
const isDate = (column, abstractType) => abstractType === 'date' || column.uidt === UITypes_1.default.Date;
exports.isDate = isDate;
const isYear = (column, abstractType) => abstractType === 'year' || column.uidt === UITypes_1.default.Year;
exports.isYear = isYear;
const isTime = (column, abstractType) => abstractType === 'time' || column.uidt === UITypes_1.default.Time;
exports.isTime = isTime;
const isDateTime = (column, abstractType) => abstractType === 'datetime' || column.uidt === UITypes_1.default.DateTime;
exports.isDateTime = isDateTime;
const isReadonlyDateTime = (column, _abstractType) => column.uidt === UITypes_1.default.CreatedTime ||
    column.uidt === UITypes_1.default.LastModifiedTime;
exports.isReadonlyDateTime = isReadonlyDateTime;
const isReadonlyUser = (column, _abstractType) => column.uidt === UITypes_1.default.CreatedBy || column.uidt === UITypes_1.default.LastModifiedBy;
exports.isReadonlyUser = isReadonlyUser;
const isJSON = (column) => column.uidt === UITypes_1.default.JSON;
exports.isJSON = isJSON;
const isEnum = (column) => column.uidt === UITypes_1.default.SingleSelect;
exports.isEnum = isEnum;
const isSingleSelect = (column) => column.uidt === UITypes_1.default.SingleSelect;
exports.isSingleSelect = isSingleSelect;
const isSet = (column) => column.uidt === UITypes_1.default.MultiSelect;
exports.isSet = isSet;
const isMultiSelect = (column) => column.uidt === UITypes_1.default.MultiSelect;
exports.isMultiSelect = isMultiSelect;
const isURL = (column) => column.uidt === UITypes_1.default.URL;
exports.isURL = isURL;
const isEmail = (column) => column.uidt === UITypes_1.default.Email;
exports.isEmail = isEmail;
const isAttachment = (column) => column.uidt === UITypes_1.default.Attachment;
exports.isAttachment = isAttachment;
const isRating = (column) => column.uidt === UITypes_1.default.Rating;
exports.isRating = isRating;
const isCurrency = (column) => column.uidt === UITypes_1.default.Currency;
exports.isCurrency = isCurrency;
const isPhoneNumber = (column) => column.uidt === UITypes_1.default.PhoneNumber;
exports.isPhoneNumber = isPhoneNumber;
const isDecimal = (column) => column.uidt === UITypes_1.default.Decimal;
exports.isDecimal = isDecimal;
const isDuration = (column) => column.uidt === UITypes_1.default.Duration;
exports.isDuration = isDuration;
const isGeoData = (column) => column.uidt === UITypes_1.default.GeoData;
exports.isGeoData = isGeoData;
const isPercent = (column) => column.uidt === UITypes_1.default.Percent;
exports.isPercent = isPercent;
const isSpecificDBType = (column) => column.uidt === UITypes_1.default.SpecificDBType;
exports.isSpecificDBType = isSpecificDBType;
const isGeometry = (column) => column.uidt === UITypes_1.default.Geometry;
exports.isGeometry = isGeometry;
const isUser = (column) => column.uidt === UITypes_1.default.User;
exports.isUser = isUser;
const isMultiUser = (column) => (0, exports.isUser)(column) && !!(0, helperFunctions_1.parseProp)(column.meta).is_multi;
exports.isMultiUser = isMultiUser;
const isButton = (column) => column.uidt === UITypes_1.default.Button;
exports.isButton = isButton;
const isAiButton = (column) => {
    var _a;
    return column.uidt === UITypes_1.default.Button &&
        ((_a = column === null || column === void 0 ? void 0 : column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === Api_1.ButtonActionsType.Ai;
};
exports.isAiButton = isAiButton;
const isScriptButton = (column) => {
    var _a;
    return column.uidt === UITypes_1.default.Button &&
        ((_a = column === null || column === void 0 ? void 0 : column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === Api_1.ButtonActionsType.Script;
};
exports.isScriptButton = isScriptButton;
const isAI = (column) => {
    var _a;
    return column.uidt === UITypes_1.default.LongText &&
        ((_a = (0, helperFunctions_1.parseProp)(column === null || column === void 0 ? void 0 : column.meta)) === null || _a === void 0 ? void 0 : _a[globals_1.LongTextAiMetaProp]) === true;
};
exports.isAI = isAI;
const isAutoSaved = (column) => [
    UITypes_1.default.SingleLineText,
    UITypes_1.default.LongText,
    UITypes_1.default.PhoneNumber,
    UITypes_1.default.Email,
    UITypes_1.default.URL,
    UITypes_1.default.Number,
    UITypes_1.default.Decimal,
    UITypes_1.default.Percent,
    UITypes_1.default.Count,
    UITypes_1.default.AutoNumber,
    UITypes_1.default.SpecificDBType,
    UITypes_1.default.Geometry,
    UITypes_1.default.GeoData,
    UITypes_1.default.Duration,
].includes(column.uidt);
exports.isAutoSaved = isAutoSaved;
const isManualSaved = (column) => [UITypes_1.default.Currency].includes(column.uidt);
exports.isManualSaved = isManualSaved;
const isPrimary = (column) => !!column.pv;
exports.isPrimary = isPrimary;
const isPrimaryKey = (column) => !!column.pk;
exports.isPrimaryKey = isPrimaryKey;
// used for LTAR and Formula
const renderValue = (result) => {
    if (!result || typeof result !== 'string') {
        return result;
    }
    // convert ISO string (e.g. in MSSQL) to YYYY-MM-DD hh:mm:ssZ
    // e.g. 2023-05-18T05:30:00.000Z -> 2023-05-18 11:00:00+05:30
    result = result.replace(/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z/g, (d) => {
        return (0, dayjs_1.default)(d).isValid() ? (0, dayjs_1.default)(d).format('YYYY-MM-DD HH:mm:ssZ') : d;
    });
    // convert all date time values to local time
    // the datetime is either YYYY-MM-DD hh:mm:ss (xcdb)
    // or YYYY-MM-DD hh:mm:ss+/-xx:yy (ext)
    return result.replace(/\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}(?:[+-]\d{2}:\d{2})?/g, (d) => {
        // TODO(timezone): retrieve the format from the corresponding column meta
        // assume HH:mm at this moment
        return (0, dayjs_1.default)(d).isValid() ? (0, dayjs_1.default)(d).format('YYYY-MM-DD HH:mm') : d;
    });
};
exports.renderValue = renderValue;
const isNumericFieldType = (column, abstractType) => {
    return ((0, exports.isInt)(column, abstractType) ||
        (0, exports.isFloat)(column, abstractType) ||
        (0, exports.isDecimal)(column) ||
        (0, exports.isCurrency)(column) ||
        (0, exports.isPercent)(column) ||
        (0, exports.isDuration)(column));
};
exports.isNumericFieldType = isNumericFieldType;
