// nocodb-sdk/build/main/lib/columnHelper/utils/date-time.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeDateOrDateTimeValue = exports.parseDateTimeValue = exports.parseDateValue = exports.DATE_DIFF_TO_SCALE_LABEL_MAP = exports.DATE_SCALE_LABEL_TO_DIFF_MAP = void 0;
const dayjs_1 = __importDefault(require("dayjs"));
const dateTimeHelper_1 = require("../../../lib/dateTimeHelper");
const helperFunctions_1 = require("../../../lib/helperFunctions");
const UITypes_1 = __importDefault(require("../../../lib/UITypes"));
exports.DATE_SCALE_LABEL_TO_DIFF_MAP = {
    Y: 'year',
    M: 'month',
    D: 'day',
    H: 'hour',
    m: 'minute',
    s: 'second',
};
exports.DATE_DIFF_TO_SCALE_LABEL_MAP = {
    year: 'Y',
    month: 'M',
    day: 'D',
    hour: 'H',
    minute: 'm',
    second: 's',
};
const parseDateValue = (value, col, isSystemCol) => {
    var _a, _b;
    value = value === null || value === void 0 ? void 0 : value.toString().trim();
    const dateFormat = !isSystemCol
        ? (_b = (_a = (0, helperFunctions_1.parseProp)(col.meta)) === null || _a === void 0 ? void 0 : _a.date_format) !== null && _b !== void 0 ? _b : 'YYYY-MM-DD'
        : 'YYYY-MM-DD HH:mm:ss';
    if (!value || !(0, dayjs_1.default)(value).isValid()) {
        return null;
    }
    else {
        value = value === null || value === void 0 ? void 0 : value.toString().trim();
        return (0, dayjs_1.default)(/^\d+$/.test(value) ? +value : value).format(dateFormat);
    }
};
exports.parseDateValue = parseDateValue;
const parseDateTimeValue = (value, params) => {
    var _a;
    // remove `"`
    // e.g. "2023-05-12T08:03:53.000Z" -> 2023-05-12T08:03:53.000Z
    value = value === null || value === void 0 ? void 0 : value.replace(/["']/g, '');
    if (!value) {
        return;
    }
    const isMySQL = (_a = params.isMysql) === null || _a === void 0 ? void 0 : _a.call(params, params.col.source_id);
    let d = (0, dayjs_1.default)(value);
    if (!d.isValid()) {
        // insert a datetime value, copy the value without refreshing
        // e.g. value = 2023-05-12T03:49:25.000Z
        // feed custom parse format
        d = (0, dayjs_1.default)(value, isMySQL ? 'YYYY-MM-DD HH:mm:ss' : 'YYYY-MM-DD HH:mm:ssZ');
    }
    // users can change the datetime format in UI
    // `value` would be always in YYYY-MM-DD HH:mm:ss(Z / +xx:yy) format
    // therefore, here we reformat to the correct datetime format based on the meta
    value = d.format((0, dateTimeHelper_1.constructDateTimeFormat)(params.col));
    if (!d.isValid()) {
        // return empty string for invalid datetime
        return null;
    }
    return value;
};
exports.parseDateTimeValue = parseDateTimeValue;
const serializeDateOrDateTimeValue = (value, col) => {
    var _a;
    if (!value)
        return null;
    value = value === null || value === void 0 ? void 0 : value.toString().trim();
    let isDateOnly = false;
    if (typeof value === 'string' && value.length < 11) {
        isDateOnly = true;
    }
    const formatting = isDateOnly
        ? (_a = (0, helperFunctions_1.parseProp)(col.meta).date_format) !== null && _a !== void 0 ? _a : 'YYYY-MM-DD'
        : (0, dateTimeHelper_1.constructDateTimeFormat)(col);
    let parsedDateOrDateTime = (0, dayjs_1.default)(value, formatting);
    if (!parsedDateOrDateTime.isValid()) {
        parsedDateOrDateTime = (0, dayjs_1.default)(value, (0, dateTimeHelper_1.getDateTimeFormat)(value));
        if (!parsedDateOrDateTime.isValid()) {
            parsedDateOrDateTime = (0, dayjs_1.default)(value, (0, dateTimeHelper_1.getDateFormat)(value));
        }
    }
    if (!parsedDateOrDateTime.isValid()) {
        return null;
    }
    return col.uidt === UITypes_1.default.Date
        ? parsedDateOrDateTime.format('YYYY-MM-DD')
        : parsedDateOrDateTime.utc().format('YYYY-MM-DD HH:mm:ssZ');
};
exports.serializeDateOrDateTimeValue = serializeDateOrDateTimeValue;
