// nocodb-sdk/build/main/lib/columnHelper/utils/decimal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.floatDivDumb = exports.floatMultDumb = exports.floatSubDumb = exports.floatAddDumb = exports.isNumberRound = exports.composeNewDecimalValue = exports.extractDecimalFromString = exports.allowedDecimalRegex = void 0;
exports.allowedDecimalRegex = /[-0123456789.]/g;
const extractDecimalFromString = (value) => {
    const extractedArrFromRegex = value.match(exports.allowedDecimalRegex);
    if (!extractedArrFromRegex || extractedArrFromRegex.length === 0)
        return '';
    let dotExists = false;
    const isMinus = extractedArrFromRegex[0] === '-';
    const extractedValueFromRegex = extractedArrFromRegex
        .filter((k) => {
        if (k === '.' && dotExists) {
            return false;
        }
        else if (k === '.') {
            dotExists = true;
        }
        else if (k === '-') {
            return false;
        }
        return true;
    })
        .join('');
    return (isMinus ? '-' : '') + extractedValueFromRegex;
};
exports.extractDecimalFromString = extractDecimalFromString;
const composeNewDecimalValue = (props) => {
    var _a, _b, _c;
    return (0, exports.extractDecimalFromString)([
        props.lastValue.substring(0, (_a = props.selectionStart) !== null && _a !== void 0 ? _a : 0),
        props.newValue,
        props.lastValue.substring((_c = (_b = props.selectionEnd) !== null && _b !== void 0 ? _b : props.selectionStart) !== null && _c !== void 0 ? _c : 0),
    ].join(''));
};
exports.composeNewDecimalValue = composeNewDecimalValue;
const isNumberRound = (a) => {
    return a - Math.floor(a) === 0;
};
exports.isNumberRound = isNumberRound;
const floatAddDumb = (a, b) => {
    return Math.round(a * 100 + b * 100) / 100;
};
exports.floatAddDumb = floatAddDumb;
const floatSubDumb = (a, b) => {
    return Math.round(a * 100 - b * 100) / 100;
};
exports.floatSubDumb = floatSubDumb;
const floatMultDumb = (a, b) => {
    return Math.round(a * b * 100) / 100;
};
exports.floatMultDumb = floatMultDumb;
const floatDivDumb = (a, b) => {
    return Math.round((a / b) * 100) / 100;
};
exports.floatDivDumb = floatDivDumb;
