// nocodb-sdk/build/main/lib/columnHelper/utils/fill-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.populateFillHandleStringNumber = populateFillHandleStringNumber;
exports.populateFillHandleStrictCopy = populateFillHandleStrictCopy;
const is_1 = require("../../../lib/is");
const decimal_1 = require("./decimal");
const fillHandlerSuffixRegex = /(.*)(\d+)$/;
const getModifierForNumber = ({ group, nullPrefix, }) => {
    var _a, _b;
    let sumOfModifier = 0;
    let lastData = Number(group[nullPrefix].data[1]);
    sumOfModifier = (0, decimal_1.floatSubDumb)(lastData, Number(group[nullPrefix].data[0]));
    for (const current of group[nullPrefix].data.slice(2)) {
        sumOfModifier += (0, decimal_1.floatSubDumb)(Number(current), lastData);
        lastData = Number(current);
    }
    return {
        modifier: (0, decimal_1.floatDivDumb)(sumOfModifier, ((_b = (_a = group[nullPrefix]) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.length) - 1),
        lastData,
    };
};
// follow google sheet behavior
function populateFillHandleStringNumber({ highlightedData, numberOfRows, }) {
    var _a, _b;
    const nullPrefix = Symbol('nullPrefix');
    const emptyCell = Symbol('empty');
    const group = {};
    const groupIndex = {};
    const rowToFill = numberOfRows - highlightedData.length;
    if (rowToFill <= 0) {
        return [];
    }
    let highLightIndex = 0;
    // group row values by prefix / suffix
    for (const highlightedRow of highlightedData) {
        let index;
        let groupKey;
        if (typeof highlightedRow === 'string') {
            // check if number
            if (!(0, is_1.ncIsNullOrUndefined)(highlightedRow) &&
                highlightedRow !== '' &&
                (0, is_1.ncIsNumber)(Number(highlightedRow))) {
                groupKey = nullPrefix;
                index = highlightedRow;
            }
            else {
                const regexResult = highlightedRow.match(fillHandlerSuffixRegex);
                if (!regexResult) {
                    if (highlightedRow !== '') {
                        groupKey = highlightedRow;
                    }
                }
                else {
                    const [_match, ...rest] = regexResult;
                    groupKey = rest[0];
                    index = rest[1];
                    if (groupKey === '') {
                        groupKey = nullPrefix;
                    }
                }
            }
        }
        else if (typeof highlightedRow === 'number') {
            groupKey = nullPrefix;
            index = highlightedRow;
        }
        if ((0, is_1.ncIsNullOrUndefined)(groupKey) && (0, is_1.ncIsNullOrUndefined)(index)) {
            groupIndex[highLightIndex++] = {
                group: {
                    data: [],
                    modifier: 0,
                    prefix: emptyCell,
                    lastIndex: undefined,
                },
                row: undefined,
            };
            continue;
        }
        if (!group[groupKey !== null && groupKey !== void 0 ? groupKey : nullPrefix]) {
            group[groupKey !== null && groupKey !== void 0 ? groupKey : nullPrefix] = {
                prefix: groupKey !== null && groupKey !== void 0 ? groupKey : nullPrefix,
                data: [],
                modifier: 0,
                lastIndex: index,
            };
        }
        const groupData = group[groupKey !== null && groupKey !== void 0 ? groupKey : nullPrefix].data;
        groupData.push(index);
        groupIndex[highLightIndex++] = {
            group: group[groupKey !== null && groupKey !== void 0 ? groupKey : nullPrefix],
            row: index,
        };
    }
    // loop once per group to calculate / set modifier
    for (const [_prefix, groupRecord] of Object.entries(group)) {
        if (groupRecord.data.length > 1) {
            let lastData = Number(groupRecord.data[1]);
            let modifier = lastData - Number(groupRecord.data[0]);
            for (const current of groupRecord.data.slice(2)) {
                if (Number(current) - lastData !== modifier) {
                    modifier = 0;
                    break;
                }
                lastData = Number(current);
            }
            groupRecord.modifier = modifier;
            groupRecord.lastIndex = lastData;
        }
    }
    // for pure number, the logic is different following google sheet
    if (((_b = (_a = group[nullPrefix]) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.length) > 1) {
        const { lastData, modifier } = getModifierForNumber({
            group,
            nullPrefix,
        });
        group[nullPrefix].modifier = modifier;
        group[nullPrefix].lastIndex = lastData;
    }
    const result = [];
    for (let fillingIndex = 0; fillingIndex < rowToFill; fillingIndex++) {
        const groupOfIndex = groupIndex[fillingIndex % highlightedData.length];
        if (emptyCell === groupOfIndex.group.prefix) {
            result.push(null);
        }
        else if (typeof groupOfIndex.group.prefix === 'string' &&
            groupOfIndex.group.modifier === 0) {
            const rowStr = (0, is_1.ncIsUndefined)(groupOfIndex.row) ? '' : groupOfIndex.row;
            result.push(`${groupOfIndex.group.prefix}${rowStr}`);
        }
        else {
            // if the modifier has decimal, round to nearest decimal
            if (groupOfIndex.group.modifier - Math.floor(groupOfIndex.group.modifier) >
                0) {
                groupOfIndex.group.lastIndex = (0, decimal_1.floatAddDumb)(groupOfIndex.group.lastIndex, groupOfIndex.group.modifier);
            }
            else {
                groupOfIndex.group.lastIndex += groupOfIndex.group.modifier;
            }
            result.push(`${groupOfIndex.group.prefix === nullPrefix
                ? ''
                : groupOfIndex.group.prefix}${Math.abs(groupOfIndex.group.lastIndex)}`);
        }
    }
    return result;
}
// simply copy the value repeatedly
function populateFillHandleStrictCopy({ highlightedData, numberOfRows, }) {
    const result = [];
    for (let i = 0; i < numberOfRows - highlightedData.length; i++) {
        result.push(highlightedData[i % highlightedData.length]);
    }
    return result;
}
