// nocodb-sdk/build/main/lib/columnHelper/utils/recursive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.traverseNested = exports.arrayToNested = void 0;
const arrayToNested = ({ data, getIdHandler, getFkHandler, childAssignHandler, parentFkId, level = 0, maxLevel = 5, }) => {
    let recordsOnLevel;
    if (level === 0) {
        recordsOnLevel = data.filter((row) => !getFkHandler(row));
    }
    else {
        recordsOnLevel = data.filter((row) => getFkHandler(row) === parentFkId);
    }
    if (level < maxLevel) {
        for (const row of recordsOnLevel) {
            childAssignHandler(row, (0, exports.arrayToNested)({
                data,
                getIdHandler,
                parentFkId: getIdHandler(row),
                childAssignHandler,
                getFkHandler,
                level: level + 1,
                maxLevel,
            }));
        }
    }
    return recordsOnLevel;
};
exports.arrayToNested = arrayToNested;
const traverseNested = (params) => {
    for (const each of params.nestedData) {
        params.onEach(each);
        (0, exports.traverseNested)({
            nestedData: params.getChildField(each),
            getChildField: params.getChildField,
            onEach: params.onEach,
        });
    }
};
exports.traverseNested = traverseNested;
