// nocodb-sdk/build/main/lib/columnHelper/utils/virtualCell.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.isReadonlyVirtualColumn = exports.isLink = exports.isCount = exports.isBarcode = exports.isQrCode = exports.isFormula = exports.isRollup = exports.isLookup = exports.isOo = exports.isBt = exports.isMm = exports.isHm = exports.isLTAR = void 0;
exports.isReadOnlyVirtualCell = isReadOnlyVirtualCell;
const globals_1 = require("../../../lib/globals");
const helperFunctions_1 = require("../../../lib/helperFunctions");
const UITypes_1 = __importStar(require("../../../lib/UITypes"));
const cell_1 = require("./cell");
const isLTAR = (uidt, _colOptions) => {
    if (!uidt)
        return false;
    return (0, UITypes_1.isLinksOrLTAR)(uidt);
};
exports.isLTAR = isLTAR;
const isHm = (column) => {
    var _a;
    return (0, exports.isLTAR)(column.uidt, column.colOptions) &&
        ((_a = column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === globals_1.RelationTypes.HAS_MANY;
};
exports.isHm = isHm;
const isMm = (column) => {
    var _a;
    return (0, exports.isLTAR)(column.uidt, column.colOptions) &&
        ((_a = column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === globals_1.RelationTypes.MANY_TO_MANY;
};
exports.isMm = isMm;
const isBt = (column) => {
    var _a;
    return (0, exports.isLTAR)(column.uidt, column.colOptions) &&
        ((_a = column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === globals_1.RelationTypes.BELONGS_TO;
};
exports.isBt = isBt;
const isOo = (column) => {
    var _a;
    return (0, exports.isLTAR)(column.uidt, column.colOptions) &&
        ((_a = column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === globals_1.RelationTypes.ONE_TO_ONE;
};
exports.isOo = isOo;
const isLookup = (column) => column.uidt === UITypes_1.default.Lookup;
exports.isLookup = isLookup;
const isRollup = (column) => column.uidt === UITypes_1.default.Rollup;
exports.isRollup = isRollup;
const isFormula = (column) => column.uidt === UITypes_1.default.Formula;
exports.isFormula = isFormula;
const isQrCode = (column) => column.uidt === UITypes_1.default.QrCode;
exports.isQrCode = isQrCode;
const isBarcode = (column) => column.uidt === UITypes_1.default.Barcode;
exports.isBarcode = isBarcode;
const isCount = (column) => column.uidt === UITypes_1.default.Count;
exports.isCount = isCount;
const isLink = (column) => column.uidt === UITypes_1.default.Links;
exports.isLink = isLink;
function isReadOnlyVirtualCell(col) {
    return ((0, exports.isRollup)(col) ||
        (0, exports.isFormula)(col) ||
        (0, exports.isBarcode)(col) ||
        (0, exports.isLookup)(col) ||
        (0, exports.isQrCode)(col) ||
        (0, helperFunctions_1.isSystemColumn)(col) ||
        (0, UITypes_1.isCreatedOrLastModifiedTimeCol)(col) ||
        (0, UITypes_1.isCreatedOrLastModifiedByCol)(col));
}
const isReadonlyVirtualColumn = (col) => {
    return ((0, helperFunctions_1.isSystemColumn)(col) ||
        (0, exports.isLookup)(col) ||
        (0, exports.isRollup)(col) ||
        (0, exports.isFormula)(col) ||
        (0, cell_1.isButton)(col) ||
        (0, UITypes_1.isVirtualCol)(col) ||
        (0, UITypes_1.isCreatedOrLastModifiedTimeCol)(col) ||
        (0, UITypes_1.isCreatedOrLastModifiedByCol)(col));
};
exports.isReadonlyVirtualColumn = isReadonlyVirtualColumn;
