// nocodb-sdk/build/module/lib/columnHelper/utils/cell.js
import dayjs from 'dayjs';
import { ButtonActionsType } from '../../../lib/Api';
import { LongTextAiMetaProp } from '../../../lib/globals';
import { parseProp } from '../../../lib/helperFunctions';
import UITypes from '../../../lib/UITypes';
export const dataTypeLow = (column) => { var _a; return (_a = column.dt) === null || _a === void 0 ? void 0 : _a.toLowerCase(); };
export const isBoolean = (column, abstractType) => column.uidt === UITypes.Checkbox || abstractType === 'boolean';
export const isString = (column, abstractType) => column.uidt === UITypes.SingleLineText || abstractType === 'string';
export const isTextArea = (column) => column.uidt === UITypes.LongText;
export const isRichText = (column) => column.uidt === UITypes.LongText && !!parseProp(column === null || column === void 0 ? void 0 : column.meta).richMode;
export const isInt = (_column, abstractType) => abstractType === 'integer';
export const isFloat = (column, abstractType) => abstractType === 'float' || column.uidt === UITypes.Number;
export const isDate = (column, abstractType) => abstractType === 'date' || column.uidt === UITypes.Date;
export const isYear = (column, abstractType) => abstractType === 'year' || column.uidt === UITypes.Year;
export const isTime = (column, abstractType) => abstractType === 'time' || column.uidt === UITypes.Time;
export const isDateTime = (column, abstractType) => abstractType === 'datetime' || column.uidt === UITypes.DateTime;
export const isReadonlyDateTime = (column, _abstractType) => column.uidt === UITypes.CreatedTime ||
    column.uidt === UITypes.LastModifiedTime;
export const isReadonlyUser = (column, _abstractType) => column.uidt === UITypes.CreatedBy || column.uidt === UITypes.LastModifiedBy;
export const isJSON = (column) => column.uidt === UITypes.JSON;
export const isEnum = (column) => column.uidt === UITypes.SingleSelect;
export const isSingleSelect = (column) => column.uidt === UITypes.SingleSelect;
export const isSet = (column) => column.uidt === UITypes.MultiSelect;
export const isMultiSelect = (column) => column.uidt === UITypes.MultiSelect;
export const isURL = (column) => column.uidt === UITypes.URL;
export const isEmail = (column) => column.uidt === UITypes.Email;
export const isAttachment = (column) => column.uidt === UITypes.Attachment;
export const isRating = (column) => column.uidt === UITypes.Rating;
export const isCurrency = (column) => column.uidt === UITypes.Currency;
export const isPhoneNumber = (column) => column.uidt === UITypes.PhoneNumber;
export const isDecimal = (column) => column.uidt === UITypes.Decimal;
export const isDuration = (column) => column.uidt === UITypes.Duration;
export const isGeoData = (column) => column.uidt === UITypes.GeoData;
export const isPercent = (column) => column.uidt === UITypes.Percent;
export const isSpecificDBType = (column) => column.uidt === UITypes.SpecificDBType;
export const isGeometry = (column) => column.uidt === UITypes.Geometry;
export const isUser = (column) => column.uidt === UITypes.User;
export const isMultiUser = (column) => isUser(column) && !!parseProp(column.meta).is_multi;
export const isButton = (column) => column.uidt === UITypes.Button;
export const isAiButton = (column) => {
    var _a;
    return column.uidt === UITypes.Button &&
        ((_a = column === null || column === void 0 ? void 0 : column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === ButtonActionsType.Ai;
};
export const isScriptButton = (column) => {
    var _a;
    return column.uidt === UITypes.Button &&
        ((_a = column === null || column === void 0 ? void 0 : column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === ButtonActionsType.Script;
};
export const isAI = (column) => {
    var _a;
    return column.uidt === UITypes.LongText &&
        ((_a = parseProp(column === null || column === void 0 ? void 0 : column.meta)) === null || _a === void 0 ? void 0 : _a[LongTextAiMetaProp]) === true;
};
export const isAutoSaved = (column) => [
    UITypes.SingleLineText,
    UITypes.LongText,
    UITypes.PhoneNumber,
    UITypes.Email,
    UITypes.URL,
    UITypes.Number,
    UITypes.Decimal,
    UITypes.Percent,
    UITypes.Count,
    UITypes.AutoNumber,
    UITypes.SpecificDBType,
    UITypes.Geometry,
    UITypes.GeoData,
    UITypes.Duration,
].includes(column.uidt);
export const isManualSaved = (column) => [UITypes.Currency].includes(column.uidt);
export const isPrimary = (column) => !!column.pv;
export const isPrimaryKey = (column) => !!column.pk;
// used for LTAR and Formula
export const renderValue = (result) => {
    if (!result || typeof result !== 'string') {
        return result;
    }
    // convert ISO string (e.g. in MSSQL) to YYYY-MM-DD hh:mm:ssZ
    // e.g. 2023-05-18T05:30:00.000Z -> 2023-05-18 11:00:00+05:30
    result = result.replace(/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z/g, (d) => {
        return dayjs(d).isValid() ? dayjs(d).format('YYYY-MM-DD HH:mm:ssZ') : d;
    });
    // convert all date time values to local time
    // the datetime is either YYYY-MM-DD hh:mm:ss (xcdb)
    // or YYYY-MM-DD hh:mm:ss+/-xx:yy (ext)
    return result.replace(/\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}(?:[+-]\d{2}:\d{2})?/g, (d) => {
        // TODO(timezone): retrieve the format from the corresponding column meta
        // assume HH:mm at this moment
        return dayjs(d).isValid() ? dayjs(d).format('YYYY-MM-DD HH:mm') : d;
    });
};
export const isNumericFieldType = (column, abstractType) => {
    return (isInt(column, abstractType) ||
        isFloat(column, abstractType) ||
        isDecimal(column) ||
        isCurrency(column) ||
        isPercent(column) ||
        isDuration(column));
};
