// nocodb-sdk/build/module/lib/columnHelper/utils/date-time.js
import dayjs from 'dayjs';
import { constructDateTimeFormat, getDateFormat, getDateTimeFormat, } from '../../../lib/dateTimeHelper';
import { parseProp } from '../../../lib/helperFunctions';
import UITypes from '../../../lib/UITypes';
export const DATE_SCALE_LABEL_TO_DIFF_MAP = {
    Y: 'year',
    M: 'month',
    D: 'day',
    H: 'hour',
    m: 'minute',
    s: 'second',
};
export const DATE_DIFF_TO_SCALE_LABEL_MAP = {
    year: 'Y',
    month: 'M',
    day: 'D',
    hour: 'H',
    minute: 'm',
    second: 's',
};
export const parseDateValue = (value, col, isSystemCol) => {
    var _a, _b;
    value = value === null || value === void 0 ? void 0 : value.toString().trim();
    const dateFormat = !isSystemCol
        ? (_b = (_a = parseProp(col.meta)) === null || _a === void 0 ? void 0 : _a.date_format) !== null && _b !== void 0 ? _b : 'YYYY-MM-DD'
        : 'YYYY-MM-DD HH:mm:ss';
    if (!value || !dayjs(value).isValid()) {
        return null;
    }
    else {
        value = value === null || value === void 0 ? void 0 : value.toString().trim();
        return dayjs(/^\d+$/.test(value) ? +value : value).format(dateFormat);
    }
};
export const parseDateTimeValue = (value, params) => {
    var _a;
    // remove `"`
    // e.g. "2023-05-12T08:03:53.000Z" -> 2023-05-12T08:03:53.000Z
    value = value === null || value === void 0 ? void 0 : value.replace(/["']/g, '');
    if (!value) {
        return;
    }
    const isMySQL = (_a = params.isMysql) === null || _a === void 0 ? void 0 : _a.call(params, params.col.source_id);
    let d = dayjs(value);
    if (!d.isValid()) {
        // insert a datetime value, copy the value without refreshing
        // e.g. value = 2023-05-12T03:49:25.000Z
        // feed custom parse format
        d = dayjs(value, isMySQL ? 'YYYY-MM-DD HH:mm:ss' : 'YYYY-MM-DD HH:mm:ssZ');
    }
    // users can change the datetime format in UI
    // `value` would be always in YYYY-MM-DD HH:mm:ss(Z / +xx:yy) format
    // therefore, here we reformat to the correct datetime format based on the meta
    value = d.format(constructDateTimeFormat(params.col));
    if (!d.isValid()) {
        // return empty string for invalid datetime
        return null;
    }
    return value;
};
export const serializeDateOrDateTimeValue = (value, col) => {
    var _a;
    if (!value)
        return null;
    value = value === null || value === void 0 ? void 0 : value.toString().trim();
    let isDateOnly = false;
    if (typeof value === 'string' && value.length < 11) {
        isDateOnly = true;
    }
    const formatting = isDateOnly
        ? (_a = parseProp(col.meta).date_format) !== null && _a !== void 0 ? _a : 'YYYY-MM-DD'
        : constructDateTimeFormat(col);
    let parsedDateOrDateTime = dayjs(value, formatting);
    if (!parsedDateOrDateTime.isValid()) {
        parsedDateOrDateTime = dayjs(value, getDateTimeFormat(value));
        if (!parsedDateOrDateTime.isValid()) {
            parsedDateOrDateTime = dayjs(value, getDateFormat(value));
        }
    }
    if (!parsedDateOrDateTime.isValid()) {
        return null;
    }
    return col.uidt === UITypes.Date
        ? parsedDateOrDateTime.format('YYYY-MM-DD')
        : parsedDateOrDateTime.utc().format('YYYY-MM-DD HH:mm:ssZ');
};
