// nocodb-sdk/build/module/lib/columnHelper/utils/decimal.js
export const allowedDecimalRegex = /[-0123456789.]/g;
export const extractDecimalFromString = (value) => {
    const extractedArrFromRegex = value.match(allowedDecimalRegex);
    if (!extractedArrFromRegex || extractedArrFromRegex.length === 0)
        return '';
    let dotExists = false;
    const isMinus = extractedArrFromRegex[0] === '-';
    const extractedValueFromRegex = extractedArrFromRegex
        .filter((k) => {
        if (k === '.' && dotExists) {
            return false;
        }
        else if (k === '.') {
            dotExists = true;
        }
        else if (k === '-') {
            return false;
        }
        return true;
    })
        .join('');
    return (isMinus ? '-' : '') + extractedValueFromRegex;
};
export const composeNewDecimalValue = (props) => {
    var _a, _b, _c;
    return extractDecimalFromString([
        props.lastValue.substring(0, (_a = props.selectionStart) !== null && _a !== void 0 ? _a : 0),
        props.newValue,
        props.lastValue.substring((_c = (_b = props.selectionEnd) !== null && _b !== void 0 ? _b : props.selectionStart) !== null && _c !== void 0 ? _c : 0),
    ].join(''));
};
export const isNumberRound = (a) => {
    return a - Math.floor(a) === 0;
};
export const floatAddDumb = (a, b) => {
    return Math.round(a * 100 + b * 100) / 100;
};
export const floatSubDumb = (a, b) => {
    return Math.round(a * 100 - b * 100) / 100;
};
export const floatMultDumb = (a, b) => {
    return Math.round(a * b * 100) / 100;
};
export const floatDivDumb = (a, b) => {
    return Math.round((a / b) * 100) / 100;
};
