// nocodb-sdk/build/module/lib/columnHelper/utils/virtualCell.js
import { RelationTypes } from '../../../lib/globals';
import { isSystemColumn } from '../../../lib/helperFunctions';
import UITypes, { isCreatedOrLastModifiedByCol, isCreatedOrLastModifiedTimeCol, isLinksOrLTAR, isVirtualCol, } from '../../../lib/UITypes';
import { isButton } from './cell';
export const isLTAR = (uidt, _colOptions) => {
    if (!uidt)
        return false;
    return isLinksOrLTAR(uidt);
};
export const isHm = (column) => {
    var _a;
    return isLTAR(column.uidt, column.colOptions) &&
        ((_a = column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === RelationTypes.HAS_MANY;
};
export const isMm = (column) => {
    var _a;
    return isLTAR(column.uidt, column.colOptions) &&
        ((_a = column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === RelationTypes.MANY_TO_MANY;
};
export const isBt = (column) => {
    var _a;
    return isLTAR(column.uidt, column.colOptions) &&
        ((_a = column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === RelationTypes.BELONGS_TO;
};
export const isOo = (column) => {
    var _a;
    return isLTAR(column.uidt, column.colOptions) &&
        ((_a = column.colOptions) === null || _a === void 0 ? void 0 : _a.type) === RelationTypes.ONE_TO_ONE;
};
export const isLookup = (column) => column.uidt === UITypes.Lookup;
export const isRollup = (column) => column.uidt === UITypes.Rollup;
export const isFormula = (column) => column.uidt === UITypes.Formula;
export const isQrCode = (column) => column.uidt === UITypes.QrCode;
export const isBarcode = (column) => column.uidt === UITypes.Barcode;
export const isCount = (column) => column.uidt === UITypes.Count;
export const isLink = (column) => column.uidt === UITypes.Links;
export function isReadOnlyVirtualCell(col) {
    return (isRollup(col) ||
        isFormula(col) ||
        isBarcode(col) ||
        isLookup(col) ||
        isQrCode(col) ||
        isSystemColumn(col) ||
        isCreatedOrLastModifiedTimeCol(col) ||
        isCreatedOrLastModifiedByCol(col));
}
export const isReadonlyVirtualColumn = (col) => {
    return (isSystemColumn(col) ||
        isLookup(col) ||
        isRollup(col) ||
        isFormula(col) ||
        isButton(col) ||
        isVirtualCol(col) ||
        isCreatedOrLastModifiedTimeCol(col) ||
        isCreatedOrLastModifiedByCol(col));
};
