// nocodb-sdk/build/main/lib/parser/queryFilter/query-filter-lexer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryFilterLexer = exports.QUERY_FILTER_TOKENS = exports.NOT_OPERATOR = exports.TOKEN_OPERATOR = exports.COMPARISON_SUB_OPS = exports.IS_WITHIN_COMPARISON_SUB_OPS = exports.COMPARISON_OPS_ALIAS = exports.COMPARISON_OPS = exports.GROUPBY_COMPARISON_OPS = void 0;
const chevrotain_1 = require("chevrotain");
const common_token_1 = require("../common-token");
exports.GROUPBY_COMPARISON_OPS = [
    // these are used for groupby
    'gb_eq',
    'gb_null',
];
exports.COMPARISON_OPS = [
    'eq',
    'neq',
    'not',
    'like',
    'nlike',
    'empty',
    'notempty',
    'null',
    'notnull',
    'checked',
    'notchecked',
    'blank',
    'notblank',
    'allof',
    'anyof',
    'nallof',
    'nanyof',
    'gt',
    'lt',
    'gte',
    'lte',
    'ge',
    'le',
    'in',
    'isnot',
    'is',
    'isWithin',
    'btw',
    'nbtw',
];
exports.COMPARISON_OPS_ALIAS = [
    'isblank',
    'is_blank',
    'isnotblank',
    'is_not_blank',
    'is_notblank',
];
exports.IS_WITHIN_COMPARISON_SUB_OPS = [
    'pastWeek',
    'pastMonth',
    'pastYear',
    'nextWeek',
    'nextMonth',
    'nextYear',
    'pastNumberOfDays',
    'nextNumberOfDays',
];
exports.COMPARISON_SUB_OPS = [
    'today',
    'tomorrow',
    'yesterday',
    'oneWeekAgo',
    'oneWeekFromNow',
    'oneMonthAgo',
    'oneMonthFromNow',
    'daysAgo',
    'daysFromNow',
    'exactDate',
    ...exports.IS_WITHIN_COMPARISON_SUB_OPS,
];
exports.TOKEN_OPERATOR = (0, chevrotain_1.createToken)({
    name: 'OPERATOR',
    pattern: new RegExp(`\\b(${[
        ...exports.COMPARISON_OPS,
        ...exports.COMPARISON_OPS_ALIAS,
        ...exports.GROUPBY_COMPARISON_OPS,
    ].join('|')})\\b`),
    longer_alt: common_token_1.COMMON_TOKEN.IDENTIFIER,
    categories: [common_token_1.COMMON_TOKEN.IDENTIFIER],
});
exports.NOT_OPERATOR = (0, chevrotain_1.createToken)({
    name: 'NOT_OPERATOR',
    pattern: /~not/,
    longer_alt: common_token_1.COMMON_TOKEN.IDENTIFIER,
    categories: [common_token_1.COMMON_TOKEN.IDENTIFIER],
});
exports.QUERY_FILTER_TOKENS = [
    common_token_1.COMMON_TOKEN.WHITESPACE_SKIP,
    exports.TOKEN_OPERATOR,
    exports.NOT_OPERATOR,
    common_token_1.COMMON_TOKEN.PAREN_START,
    common_token_1.COMMON_TOKEN.PAREN_END,
    common_token_1.COMMON_TOKEN.COMMA,
    common_token_1.COMMON_TOKEN.EMPTY_QUOTED_IDENTIFIER,
    common_token_1.COMMON_TOKEN.SUP_SGL_QUOTE_IDENTIFIER,
    common_token_1.COMMON_TOKEN.SUP_DBL_QUOTE_IDENTIFIER,
    common_token_1.COMMON_TOKEN.IDENTIFIER,
];
exports.QueryFilterLexer = new chevrotain_1.Lexer(exports.QUERY_FILTER_TOKENS);
