// nocodb-sdk/build/main/lib/parser/queryFilter/query-filter-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryFilterParser = void 0;
const query_filter_lexer_1 = require("./query-filter-lexer");
const common_token_1 = require("../common-token");
const common_cst_parser_1 = require("../common-cst-parser");
const query_filter_cst_parser_1 = require("./query-filter-cst-parser");
class QueryFilterParser extends common_cst_parser_1.CommonCstParser {
    constructor() {
        super(query_filter_lexer_1.QUERY_FILTER_TOKENS);
        this.initializeRule();
        // very important to call this after all the rules have been defined.
        // otherwise the parser may not work correctly as it will lack information
        // derived during the self analysis phase.
        this.performSelfAnalysis();
    }
    initializeRule() {
        // not mandatory, using $ (or any other sign) to reduce verbosity (this. this. this. this. .......)
        const $ = this;
        // the parsing methods
        // we define the "rules" of how the syntax will be defined
        $.RULE('multi_clause', () => {
            $.OR([
                { ALT: () => $.SUBRULE($['not_clause'], { LABEL: 'clause' }) },
                { ALT: () => $.SUBRULE($['paren_clause'], { LABEL: 'clause' }) },
            ]);
            $.MANY({
                DEF: () => $.SUBRULE($['and_or_clause'], { LABEL: 'clause' }),
            });
        });
        $.RULE('and_or_clause', () => {
            $.CONSUME(common_token_1.COMMON_TOKEN.IDENTIFIER, { LABEL: 'operator' });
            $.SUBRULE($['paren_clause'], { LABEL: 'clause' });
        });
        $.RULE('not_clause', () => {
            $.CONSUME(query_filter_lexer_1.NOT_OPERATOR);
            $.SUBRULE($['paren_clause'], { LABEL: 'clause' });
        });
        $.RULE('paren_clause', () => {
            $.CONSUME(common_token_1.COMMON_TOKEN.PAREN_START);
            $.OR([
                { ALT: () => $.SUBRULE($['multi_clause'], { LABEL: 'clause' }) },
                { ALT: () => $.SUBRULE($['call_expression'], { LABEL: 'clause' }) },
            ]);
            $.CONSUME(common_token_1.COMMON_TOKEN.PAREN_END);
        });
        $.RULE('call_expression', () => {
            $.SUBRULE($['VARIABLE']);
            $.CONSUME2(common_token_1.COMMON_TOKEN.COMMA);
            $.CONSUME(query_filter_lexer_1.TOKEN_OPERATOR);
            $.OPTION(() => {
                $.CONSUME3(common_token_1.COMMON_TOKEN.COMMA);
                // even after comma, the expression arguments can be blank
                $.OPTION2(() => {
                    $.SUBRULE($['expression_arguments']);
                });
            });
        });
        $.RULE('expression_arguments', () => {
            $.MANY_SEP({
                SEP: common_token_1.COMMON_TOKEN.COMMA,
                DEF: () => {
                    $.SUBRULE2($['VARIABLE']);
                },
            });
        });
    }
    parse() {
        return this.multi_clause();
    }
    static parse(text) {
        const parser = new QueryFilterParser();
        const lexResult = query_filter_lexer_1.QueryFilterLexer.tokenize(text);
        // setting a new input will RESET the parser instance's state.
        parser.input = lexResult.tokens;
        // any top level rule may be used as an entry point
        const cst = parser.parse();
        return {
            cst: cst,
            lexErrors: lexResult.errors,
            parseErrors: parser.errors,
            parsedCst: parser.errors && parser.errors.length === 0 ? (0, query_filter_cst_parser_1.parseCst)(cst) : undefined,
        };
    }
}
exports.QueryFilterParser = QueryFilterParser;
