// nocodb-sdk/build/main/lib/parser/common-cst-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonCstParser = exports.trimQuote = exports.parseVariableAsString = exports.parseVariableAsArray = exports.parseVariable = exports.getVariableRuleToken = void 0;
const chevrotain_1 = require("chevrotain");
const common_token_1 = require("./common-token");
const getVariableRuleToken = (variable) => {
    var _a, _b, _c, _d, _e;
    const childToken = (_d = (_b = (_a = variable.children.SUP_SGL_QUOTE_IDENTIFIER) === null || _a === void 0 ? void 0 : _a[0]) !== null && _b !== void 0 ? _b : (_c = variable.children.SUP_DBL_QUOTE_IDENTIFIER) === null || _c === void 0 ? void 0 : _c[0]) !== null && _d !== void 0 ? _d : (_e = variable.children.IDENTIFIER) === null || _e === void 0 ? void 0 : _e[0];
    return Object.assign(Object.assign({}, variable), { startOffset: childToken === null || childToken === void 0 ? void 0 : childToken.startOffset, endOffset: childToken === null || childToken === void 0 ? void 0 : childToken.endOffset, startLine: childToken === null || childToken === void 0 ? void 0 : childToken.startLine, endLine: childToken === null || childToken === void 0 ? void 0 : childToken.endLine, startColumn: childToken === null || childToken === void 0 ? void 0 : childToken.startColumn, endColumn: childToken === null || childToken === void 0 ? void 0 : childToken.endColumn });
};
exports.getVariableRuleToken = getVariableRuleToken;
const parseVariable = (variable) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
    if (Array.isArray(variable)) {
        if (variable.length === 1) {
            return (0, exports.parseVariable)(variable[0]);
        }
        return variable.map((eachVar) => (0, exports.parseVariable)(eachVar));
    }
    else {
        if (((_b = (_a = variable.children.IDENTIFIER) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.image) &&
            (((_d = (_c = variable.children.IDENTIFIER) === null || _c === void 0 ? void 0 : _c[0]) === null || _d === void 0 ? void 0 : _d.image) === 'NULL' ||
                ((_f = (_e = variable.children.IDENTIFIER) === null || _e === void 0 ? void 0 : _e[0]) === null || _f === void 0 ? void 0 : _f.image) === 'null')) {
            return null;
        }
        else if (variable.children.EMPTY_QUOTED_IDENTIFIER &&
            ((_h = (_g = variable.children.EMPTY_QUOTED_IDENTIFIER) === null || _g === void 0 ? void 0 : _g[0]) === null || _h === void 0 ? void 0 : _h.image)) {
            return '';
        }
        return ((_p = (_l = (_k = (_j = variable.children.IDENTIFIER) === null || _j === void 0 ? void 0 : _j[0]) === null || _k === void 0 ? void 0 : _k.image) !== null && _l !== void 0 ? _l : (0, exports.trimQuote)((_o = (_m = variable.children.SUP_SGL_QUOTE_IDENTIFIER) === null || _m === void 0 ? void 0 : _m[0]) === null || _o === void 0 ? void 0 : _o.image)) !== null && _p !== void 0 ? _p : (0, exports.trimQuote)((_r = (_q = variable.children.SUP_DBL_QUOTE_IDENTIFIER) === null || _q === void 0 ? void 0 : _q[0]) === null || _r === void 0 ? void 0 : _r.image));
    }
};
exports.parseVariable = parseVariable;
const parseVariableAsArray = (variable) => {
    const result = (0, exports.parseVariable)(variable);
    if (!Array.isArray(result)) {
        return [result];
    }
    else {
        return result;
    }
};
exports.parseVariableAsArray = parseVariableAsArray;
const parseVariableAsString = (variable) => {
    const result = (0, exports.parseVariable)(variable);
    if (Array.isArray(result)) {
        if (result.filter((k) => k === null).length === result.length) {
            return null;
        }
        else {
            return result.join(' ');
        }
    }
    else {
        return result;
    }
};
exports.parseVariableAsString = parseVariableAsString;
const trimQuote = (value) => {
    if (value === undefined) {
        return value;
    }
    return value === null || value === void 0 ? void 0 : value.substring(1, value.length - 1);
};
exports.trimQuote = trimQuote;
class CommonCstParser extends chevrotain_1.CstParser {
    constructor(tokens, opt = {
        enabledRules: {
            VARIABLE: 'VARIABLE',
        },
    }) {
        super(tokens);
        this.prepareCommonRules(opt);
    }
    prepareCommonRules(opt) {
        const $ = this;
        if (opt.enabledRules.VARIABLE) {
            $.RULE(opt.enabledRules.VARIABLE, () => {
                $.OR([
                    {
                        ALT: () => {
                            $.CONSUME(common_token_1.COMMON_TOKEN.EMPTY_QUOTED_IDENTIFIER);
                        },
                    },
                    {
                        ALT: () => {
                            $.CONSUME(common_token_1.COMMON_TOKEN.SUP_SGL_QUOTE_IDENTIFIER);
                        },
                    },
                    {
                        ALT: () => {
                            $.CONSUME(common_token_1.COMMON_TOKEN.SUP_DBL_QUOTE_IDENTIFIER);
                        },
                    },
                    {
                        ALT: () => {
                            $.CONSUME(common_token_1.COMMON_TOKEN.COMMA_SUPPORTED_IDENTIFIER);
                        },
                    },
                    {
                        ALT: () => {
                            $.CONSUME(common_token_1.COMMON_TOKEN.IDENTIFIER);
                        },
                    },
                ]);
            });
        }
    }
}
exports.CommonCstParser = CommonCstParser;
