// nocodb-sdk/build/main/lib/parser/common-token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMMON_TOKEN = void 0;
const chevrotain_1 = require("chevrotain");
const IDENTIFIER = (0, chevrotain_1.createToken)({
    name: 'IDENTIFIER',
    pattern: /[^()'",]+/,
});
const COMMA_SUPPORTED_IDENTIFIER = (0, chevrotain_1.createToken)({
    name: 'COMMA_SUPPORTED_IDENTIFIER',
    pattern: /[^()'"]+/,
});
exports.COMMON_TOKEN = {
    PAREN_START: (0, chevrotain_1.createToken)({ name: 'PAREN_START', pattern: /\(/ }),
    PAREN_END: (0, chevrotain_1.createToken)({ name: 'PAREN_END', pattern: /\)/ }),
    SQ_BRACKET_START: (0, chevrotain_1.createToken)({ name: 'SQ_BRACKET_START', pattern: /\[/ }),
    SQ_BRACKET_END: (0, chevrotain_1.createToken)({ name: 'SQ_BRACKET_END', pattern: /\]/ }),
    CURLY_START: (0, chevrotain_1.createToken)({ name: 'CURLY_START', pattern: /\{/ }),
    CURLY_END: (0, chevrotain_1.createToken)({ name: 'CURLY_END', pattern: /\}/ }),
    NUMBER_LITERAL: (0, chevrotain_1.createToken)({ name: 'NUMBER_LITERAL', pattern: /[1-9]\d*/ }),
    ALPHABET_LITERAL: (0, chevrotain_1.createToken)({
        name: 'ALPHABET_LITERAL',
        pattern: /[a-zA-Z]\d*/,
    }),
    COMMA: (0, chevrotain_1.createToken)({ name: 'COMMA', pattern: /,/ }),
    WHITESPACE: (0, chevrotain_1.createToken)({
        name: 'WHITESPACE',
        // do not delete the space inside aregex
        pattern: /[ \t\n\r]+/,
    }),
    WHITESPACE_SKIP: (0, chevrotain_1.createToken)({
        name: 'WHITESPACE',
        // do not delete the space inside aregex
        pattern: /[ \t\n\r]+/,
        group: chevrotain_1.Lexer.SKIPPED,
    }),
    PLUS: (0, chevrotain_1.createToken)({ name: 'PLUS', pattern: /\+/ }),
    MINUS: (0, chevrotain_1.createToken)({ name: 'MINUS', pattern: /-/ }),
    MULTI: (0, chevrotain_1.createToken)({ name: 'MULTI', pattern: /\*/ }),
    DIVIDE: (0, chevrotain_1.createToken)({ name: 'DIVIDE', pattern: /\// }),
    COLON: (0, chevrotain_1.createToken)({ name: 'COLON', pattern: /:/ }),
    SEMICOLON: (0, chevrotain_1.createToken)({ name: 'SEMICOLON', pattern: /;/ }),
    PIPE: (0, chevrotain_1.createToken)({ name: 'PIPE', pattern: /|/ }),
    SLASH: (0, chevrotain_1.createToken)({ name: 'SLASH', pattern: /\// }),
    BACKSLASH: (0, chevrotain_1.createToken)({ name: 'BACKSLASH', pattern: /\\/ }),
    DBL_QUOTE: (0, chevrotain_1.createToken)({ name: 'DBL_QUOTE', pattern: /"/ }),
    SGL_QUOTE: (0, chevrotain_1.createToken)({ name: 'SGL_QUOTE', pattern: /'/ }),
    IDENTIFIER: IDENTIFIER,
    COMMA_SUPPORTED_IDENTIFIER: COMMA_SUPPORTED_IDENTIFIER,
    EMPTY_QUOTED_IDENTIFIER: (0, chevrotain_1.createToken)({
        name: 'EMPTY_QUOTED_IDENTIFIER',
        pattern: /''|""|``/,
        longer_alt: IDENTIFIER,
    }),
    SUP_SGL_QUOTE_IDENTIFIER: (0, chevrotain_1.createToken)({
        name: 'SUP_SGL_QUOTE_IDENTIFIER',
        pattern: /'((?:\\'|[^'])+)'/,
        longer_alt: IDENTIFIER,
    }),
    SUP_DBL_QUOTE_IDENTIFIER: (0, chevrotain_1.createToken)({
        name: 'SUP_DBL_QUOTE_IDENTIFIER',
        pattern: /"((?:\\"|[^"])+)"/,
        longer_alt: IDENTIFIER,
    }),
    SUP_BACK_QUOTE_IDENTIFIER: (0, chevrotain_1.createToken)({
        name: 'SUP_BACK_QUOTE_IDENTIFIER',
        pattern: /`((?:\\`|[^`])+)`/,
        longer_alt: IDENTIFIER,
    }),
};
