// nocodb-sdk/build/main/lib/columnHelper/columns/LTAR/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneToOneHelper = exports.ManyToManyHelper = exports.HasManyHelper = exports.BelongsToHelper = exports.LTARHelper = void 0;
const column_interface_1 = __importDefault(require("../../column.interface"));
const utils_1 = require("../../utils");
const DefaultColumnHelper_1 = require("../DefaultColumnHelper");
const BelongsTo_1 = require("./BelongsTo");
Object.defineProperty(exports, "BelongsToHelper", { enumerable: true, get: function () { return BelongsTo_1.BelongsToHelper; } });
const HasMany_1 = require("./HasMany");
Object.defineProperty(exports, "HasManyHelper", { enumerable: true, get: function () { return HasMany_1.HasManyHelper; } });
const ManyToMany_1 = require("./ManyToMany");
Object.defineProperty(exports, "ManyToManyHelper", { enumerable: true, get: function () { return ManyToMany_1.ManyToManyHelper; } });
const OneToOne_1 = require("./OneToOne");
Object.defineProperty(exports, "OneToOneHelper", { enumerable: true, get: function () { return OneToOne_1.OneToOneHelper; } });
class LTARHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {};
    }
    getLtarHelperColumn(params) {
        let columnHelper = DefaultColumnHelper_1.DefaultColumnHelper;
        if ((0, utils_1.isHm)(params.col))
            columnHelper = HasMany_1.HasManyHelper;
        if ((0, utils_1.isMm)(params.col))
            columnHelper = ManyToMany_1.ManyToManyHelper;
        if ((0, utils_1.isBt)(params.col))
            columnHelper = BelongsTo_1.BelongsToHelper;
        if ((0, utils_1.isOo)(params.col))
            columnHelper = OneToOne_1.OneToOneHelper;
        return new columnHelper();
    }
    serializeValue(value, params) {
        return this.getLtarHelperColumn(params).serializeValue(value, params);
    }
    parseValue(value, params) {
        return this.getLtarHelperColumn(params).parseValue(value, params);
    }
    parsePlainCellValue(value, params) {
        return this.getLtarHelperColumn(params).parsePlainCellValue(value, params);
    }
}
exports.LTARHelper = LTARHelper;
