// nocodb-sdk/build/main/lib/dashboard/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextWidgetTypes = exports.WidgetDataSourceTypes = exports.WidgetChartLabelMap = exports.WidgetTypes = void 0;
const charts_1 = require("./charts");
var WidgetTypes;
(function (WidgetTypes) {
    WidgetTypes["CHART"] = "chart";
    WidgetTypes["METRIC"] = "metric";
    WidgetTypes["TEXT"] = "text";
    WidgetTypes["TABLE"] = "table";
    WidgetTypes["IFRAME"] = "iframe";
})(WidgetTypes || (exports.WidgetTypes = WidgetTypes = {}));
exports.WidgetChartLabelMap = {
    [charts_1.ChartTypes.BAR]: 'Bar Chart',
    [charts_1.ChartTypes.LINE]: 'Line Chart',
    [charts_1.ChartTypes.PIE]: 'Pie Chart',
    [charts_1.ChartTypes.DONUT]: 'Donut Chart',
    [charts_1.ChartTypes.SCATTER]: 'Scatter Plot',
    [WidgetTypes.TABLE]: 'Table',
    [WidgetTypes.METRIC]: 'Metric',
    [WidgetTypes.TEXT]: 'Text',
    [WidgetTypes.IFRAME]: 'IFrame',
};
var WidgetDataSourceTypes;
(function (WidgetDataSourceTypes) {
    WidgetDataSourceTypes["VIEW"] = "view";
    WidgetDataSourceTypes["MODEL"] = "model";
    WidgetDataSourceTypes["FILTER"] = "filter";
})(WidgetDataSourceTypes || (exports.WidgetDataSourceTypes = WidgetDataSourceTypes = {}));
var TextWidgetTypes;
(function (TextWidgetTypes) {
    TextWidgetTypes["Markdown"] = "markdown";
    TextWidgetTypes["Text"] = "text";
})(TextWidgetTypes || (exports.TextWidgetTypes = TextWidgetTypes = {}));
__exportStar(require("./charts"), exports);
__exportStar(require("./widgetUtils"), exports);
