// nocodb-sdk/build/main/lib/sync/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TARGET_TABLES_META = exports.TARGET_TABLES = exports.SyncCategoryMeta = exports.SyncTypeMeta = exports.OnDeleteActionMeta = exports.SyncTriggerMeta = exports.SyncCategory = exports.OnDeleteAction = exports.SyncTrigger = exports.SyncType = void 0;
var SyncType;
(function (SyncType) {
    SyncType["Full"] = "full";
    SyncType["Incremental"] = "incremental";
})(SyncType || (exports.SyncType = SyncType = {}));
var SyncTrigger;
(function (SyncTrigger) {
    SyncTrigger["Manual"] = "manual";
    SyncTrigger["Schedule"] = "schedule";
    SyncTrigger["Webhook"] = "webhook";
})(SyncTrigger || (exports.SyncTrigger = SyncTrigger = {}));
var OnDeleteAction;
(function (OnDeleteAction) {
    OnDeleteAction["Delete"] = "delete";
    OnDeleteAction["MarkDeleted"] = "mark_deleted";
})(OnDeleteAction || (exports.OnDeleteAction = OnDeleteAction = {}));
var SyncCategory;
(function (SyncCategory) {
    SyncCategory["TICKETING"] = "ticketing";
    SyncCategory["CRM"] = "crm";
    SyncCategory["FILE_STORAGE"] = "file_storage";
    SyncCategory["CUSTOM"] = "custom";
})(SyncCategory || (exports.SyncCategory = SyncCategory = {}));
exports.SyncTriggerMeta = {
    [SyncTrigger.Manual]: {
        value: SyncTrigger.Manual,
        label: 'Manual',
        description: 'Sync data manually',
    },
    [SyncTrigger.Schedule]: {
        value: SyncTrigger.Schedule,
        label: 'Scheduled',
        description: 'Sync data on a schedule',
    },
    [SyncTrigger.Webhook]: {
        value: SyncTrigger.Webhook,
        label: 'Webhook',
        description: 'Sync data via a webhook',
    },
};
exports.OnDeleteActionMeta = {
    [OnDeleteAction.Delete]: {
        value: OnDeleteAction.Delete,
        label: 'Delete',
        description: 'Delete data permanently in NocoDB',
    },
    [OnDeleteAction.MarkDeleted]: {
        value: OnDeleteAction.MarkDeleted,
        label: 'Mark Deleted',
        description: 'Mark data as deleted in NocoDB',
    },
};
exports.SyncTypeMeta = {
    [SyncType.Full]: {
        value: SyncType.Full,
        label: 'Full',
        description: 'Sync all data',
    },
    [SyncType.Incremental]: {
        value: SyncType.Incremental,
        label: 'Incremental',
        description: 'Sync only new and updated data',
    },
};
exports.SyncCategoryMeta = {
    [SyncCategory.TICKETING]: {
        value: SyncCategory.TICKETING,
        label: 'Ticketing',
        description: 'Sync data from a ticketing system',
        icon: 'ncBookOpen',
    },
    [SyncCategory.CRM]: {
        value: SyncCategory.CRM,
        label: 'CRM',
        description: 'Sync data from a CRM',
        icon: 'ncUsers',
    },
    [SyncCategory.FILE_STORAGE]: {
        value: SyncCategory.FILE_STORAGE,
        label: 'File Storage',
        description: 'Sync data from a file storage system',
        icon: 'ncFolder',
    },
    [SyncCategory.CUSTOM]: {
        value: SyncCategory.CUSTOM,
        label: 'Custom',
        description: 'Sync data from a dynamic source',
        icon: 'ncDatabase',
    },
};
var TARGET_TABLES;
(function (TARGET_TABLES) {
    TARGET_TABLES["TICKETING_TICKET"] = "ticketing_ticket";
    TARGET_TABLES["TICKETING_USER"] = "ticketing_user";
    TARGET_TABLES["TICKETING_COMMENT"] = "ticketing_comment";
    TARGET_TABLES["TICKETING_TEAM"] = "ticketing_team";
})(TARGET_TABLES || (exports.TARGET_TABLES = TARGET_TABLES = {}));
exports.TARGET_TABLES_META = {
    [TARGET_TABLES.TICKETING_TICKET]: {
        category: SyncCategory.TICKETING,
        value: TARGET_TABLES.TICKETING_TICKET,
        icon: 'ncBookOpen',
        label: 'Ticket',
        description: 'Sync all ticket data from the source',
        required: true,
    },
    [TARGET_TABLES.TICKETING_USER]: {
        category: SyncCategory.TICKETING,
        value: TARGET_TABLES.TICKETING_USER,
        icon: 'ncUsers',
        label: 'User',
        description: 'Sync all users on tickets from the source',
        required: true,
    },
    [TARGET_TABLES.TICKETING_COMMENT]: {
        category: SyncCategory.TICKETING,
        value: TARGET_TABLES.TICKETING_COMMENT,
        icon: 'ncMessageCircle',
        label: 'Comment',
        description: 'Sync all comments on tickets',
        required: false,
    },
    [TARGET_TABLES.TICKETING_TEAM]: {
        category: SyncCategory.TICKETING,
        value: TARGET_TABLES.TICKETING_TEAM,
        icon: 'ncUsers',
        label: 'Team',
        description: 'Sync all teams from the source',
        required: false,
    },
};
