// nocodb-sdk/build/main/lib/payment/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUpgradeMessage = exports.PlanFeatureUpgradeMessages = exports.PlanLimitUpgradeMessages = exports.LOYALTY_SEAT_PRICE_CAP = exports.SEAT_PRICE_CAP = exports.LOYALTY_GRACE_PERIOD_END_DATE = exports.GRACE_PERIOD_DURATION = exports.HigherPlan = exports.PlanOrderToPlan = exports.PlanOrder = exports.PlanMeta = exports.LoyaltyPriceReverseLookupKeyMap = exports.LoyaltyPriceLookupKeyMap = exports.PlanPriceLookupKeys = exports.PlanTitles = exports.PlanFeatureTypes = exports.PlanLimitTypes = void 0;
var PlanLimitTypes;
(function (PlanLimitTypes) {
    PlanLimitTypes["LIMIT_FREE_WORKSPACE"] = "limit_free_workspace";
    PlanLimitTypes["LIMIT_EDITOR"] = "limit_editor";
    PlanLimitTypes["LIMIT_COMMENTER"] = "limit_commenter";
    PlanLimitTypes["LIMIT_API_PER_SECOND"] = "limit_api_per_second";
    PlanLimitTypes["LIMIT_AI_TOKEN"] = "limit_ai_token";
    PlanLimitTypes["LIMIT_API_CALL"] = "limit_api_call";
    PlanLimitTypes["LIMIT_AUDIT_RETENTION"] = "limit_audit_retention";
    PlanLimitTypes["LIMIT_AUTOMATION_RUN"] = "limit_automation_run";
    PlanLimitTypes["LIMIT_AUTOMATION_RETENTION"] = "limit_automation_retention";
    PlanLimitTypes["LIMIT_WEBHOOK_PER_WORKSPACE"] = "limit_webhook";
    PlanLimitTypes["LIMIT_EXTENSION_PER_WORKSPACE"] = "limit_extension";
    PlanLimitTypes["LIMIT_SNAPSHOT_PER_WORKSPACE"] = "limit_snapshot";
    PlanLimitTypes["LIMIT_STORAGE_PER_WORKSPACE"] = "limit_storage";
    PlanLimitTypes["LIMIT_RECORD_PER_WORKSPACE"] = "limit_record";
    PlanLimitTypes["LIMIT_BASE_PER_WORKSPACE"] = "limit_base";
    PlanLimitTypes["LIMIT_EXTERNAL_SOURCE_PER_WORKSPACE"] = "limit_external_source";
    PlanLimitTypes["LIMIT_TABLE_PER_BASE"] = "limit_table_per_base";
    PlanLimitTypes["LIMIT_COLUMN_PER_TABLE"] = "limit_column_per_table";
    PlanLimitTypes["LIMIT_WEBHOOK_PER_TABLE"] = "limit_webhook_per_table";
    PlanLimitTypes["LIMIT_VIEW_PER_TABLE"] = "limit_view_per_table";
    PlanLimitTypes["LIMIT_FILTER_PER_VIEW"] = "limit_filter_per_view";
    PlanLimitTypes["LIMIT_SORT_PER_VIEW"] = "limit_sort_per_view";
    PlanLimitTypes["LIMIT_ATTACHMENTS_IN_CELL"] = "limit_attachments_in_cell";
})(PlanLimitTypes || (exports.PlanLimitTypes = PlanLimitTypes = {}));
var PlanFeatureTypes;
(function (PlanFeatureTypes) {
    PlanFeatureTypes["FEATURE_AI"] = "feature_ai";
    PlanFeatureTypes["FEATURE_AI_INTEGRATIONS"] = "feature_ai_integrations";
    PlanFeatureTypes["FEATURE_AT_MENTION"] = "feature_at_mention";
    PlanFeatureTypes["FEATURE_AUDIT_WORKSPACE"] = "feature_audit_workspace";
    PlanFeatureTypes["FEATURE_COMMENT_RESOLVE"] = "feature_comment_resolve";
    PlanFeatureTypes["FEATURE_CUSTOM_URL"] = "feature_custom_url";
    PlanFeatureTypes["FEATURE_DISCUSSION_MODE"] = "feature_discussion_mode";
    PlanFeatureTypes["FEATURE_EXTENSIONS"] = "feature_extensions";
    PlanFeatureTypes["FEATURE_FILE_MODE"] = "feature_file_mode";
    PlanFeatureTypes["FEATURE_FORM_URL_REDIRECTION"] = "feature_form_url_redirection";
    PlanFeatureTypes["FEATURE_FORM_CUSTOM_LOGO"] = "feature_form_custom_logo";
    PlanFeatureTypes["FEATURE_FORM_FIELD_ON_CONDITION"] = "feature_form_field_on_condition";
    PlanFeatureTypes["FEATURE_FORM_FIELD_VALIDATION"] = "feature_form_field_validation";
    PlanFeatureTypes["FEATURE_GROUP_BY_AGGREGATIONS"] = "feature_group_by_aggregations";
    PlanFeatureTypes["FEATURE_HIDE_BRANDING"] = "feature_hide_branding";
    PlanFeatureTypes["FEATURE_LTAR_LIMIT_SELECTION_BY_FILTER"] = "feature_ltar_limit_selection_by_filter";
    PlanFeatureTypes["FEATURE_LOOKUP_LIMIT_RECORDS_BY_FILTER"] = "feature_lookup_limit_records_by_filter";
    PlanFeatureTypes["FEATURE_ROLLUP_LIMIT_RECORDS_BY_FILTER"] = "feature_rollup_limit_records_by_filter";
    PlanFeatureTypes["FEATURE_PERSONAL_VIEWS"] = "feature_personal_views";
    PlanFeatureTypes["FEATURE_SCRIPTS"] = "feature_scripts";
    PlanFeatureTypes["FEATURE_DASHBOARD"] = "feature_dashboard";
    PlanFeatureTypes["FEATURE_SSO"] = "feature_sso";
    PlanFeatureTypes["FEATURE_WEBHOOK_CUSTOM_PAYLOAD"] = "feature_webhook_custom_payload";
    PlanFeatureTypes["FEATURE_WORKSPACE_CUSTOM_LOGO"] = "feature_workspace_custom_logo";
    PlanFeatureTypes["FEATURE_CURRENT_USER_FILTER"] = "feature_current_user_filter";
    PlanFeatureTypes["FEATURE_ROW_COLOUR"] = "feature_row_colour";
    PlanFeatureTypes["FEATURE_TABLE_AND_FIELD_PERMISSIONS"] = "feature_table_and_field_permissions";
    PlanFeatureTypes["FEATURE_PRIVATE_BASES"] = "feature_private_bases";
    PlanFeatureTypes["FEATURE_API_MEMBER_MANAGEMENT"] = "feature_api_member_management";
})(PlanFeatureTypes || (exports.PlanFeatureTypes = PlanFeatureTypes = {}));
// todo: separate as a new enum
var PlanTitles;
(function (PlanTitles) {
    PlanTitles["FREE"] = "Free";
    PlanTitles["PLUS"] = "Plus";
    PlanTitles["BUSINESS"] = "Business";
    PlanTitles["ENTERPRISE"] = "Enterprise";
    // on-prem
    PlanTitles["ENTERPRISE_STARTER"] = "EnterpriseStarter";
})(PlanTitles || (exports.PlanTitles = PlanTitles = {}));
var PlanPriceLookupKeys;
(function (PlanPriceLookupKeys) {
    PlanPriceLookupKeys["PLUS_MONTHLY"] = "plus_monthly";
    PlanPriceLookupKeys["PLUS_YEARLY"] = "plus_yearly";
    PlanPriceLookupKeys["BUSINESS_MONTHLY"] = "business_monthly";
    PlanPriceLookupKeys["BUSINESS_YEARLY"] = "business_yearly";
})(PlanPriceLookupKeys || (exports.PlanPriceLookupKeys = PlanPriceLookupKeys = {}));
exports.LoyaltyPriceLookupKeyMap = {
    [PlanPriceLookupKeys.PLUS_MONTHLY]: 'loyalty_plus_monthly',
    [PlanPriceLookupKeys.PLUS_YEARLY]: 'loyalty_plus_yearly',
    [PlanPriceLookupKeys.BUSINESS_MONTHLY]: 'loyalty_business_monthly',
    [PlanPriceLookupKeys.BUSINESS_YEARLY]: 'loyalty_business_yearly',
};
exports.LoyaltyPriceReverseLookupKeyMap = {
    [exports.LoyaltyPriceLookupKeyMap[PlanPriceLookupKeys.PLUS_MONTHLY]]: PlanPriceLookupKeys.PLUS_MONTHLY,
    [exports.LoyaltyPriceLookupKeyMap[PlanPriceLookupKeys.PLUS_YEARLY]]: PlanPriceLookupKeys.PLUS_YEARLY,
    [exports.LoyaltyPriceLookupKeyMap[PlanPriceLookupKeys.BUSINESS_MONTHLY]]: PlanPriceLookupKeys.BUSINESS_MONTHLY,
    [exports.LoyaltyPriceLookupKeyMap[PlanPriceLookupKeys.BUSINESS_YEARLY]]: PlanPriceLookupKeys.BUSINESS_YEARLY,
};
exports.PlanMeta = {
    [PlanTitles.FREE]: {
        title: PlanTitles.FREE,
        color: '#F9F9FA',
        accent: '#E7E7E9',
        primary: '#1F293A',
        bgLight: '#F9F9FA',
        bgDark: '#F4F4F5',
        border: '#E7E7E9',
        chartFillColor: '#6A7184',
    },
    [PlanTitles.PLUS]: {
        title: PlanTitles.PLUS,
        color: '#EDF9FF',
        accent: '#AFE5FF',
        primary: '#207399',
        bgLight: '#EDF9FF',
        bgDark: '#D7F2FF',
        border: '#AFE5FF',
        chartFillColor: '#207399',
    },
    [PlanTitles.BUSINESS]: {
        title: PlanTitles.BUSINESS,
        color: '#FAF5FF',
        accent: '#FEB0E8',
        primary: '#972377',
        bgLight: '#FFEEFB',
        bgDark: '#FED8F4',
        border: '#FEB0E8',
        chartFillColor: '#972377',
    },
    [PlanTitles.ENTERPRISE]: {
        title: PlanTitles.ENTERPRISE,
        color: '#0D1117',
        accent: '#663B1F',
        primary: '#C86827',
        bgLight: '#FFF5EF',
        bgDark: '#FEE6D6',
        border: '#FDCDAD',
        chartFillColor: '#C86827',
    },
};
exports.PlanOrder = {
    [PlanTitles.FREE]: 0,
    [PlanTitles.PLUS]: 1,
    [PlanTitles.BUSINESS]: 2,
    [PlanTitles.ENTERPRISE]: 3,
};
exports.PlanOrderToPlan = Object.entries(exports.PlanOrder).reduce((acc, [plan, order]) => {
    acc[order] = plan;
    return acc;
}, {});
exports.HigherPlan = {
    [PlanTitles.FREE]: PlanTitles.PLUS,
    [PlanTitles.PLUS]: PlanTitles.BUSINESS,
    [PlanTitles.BUSINESS]: PlanTitles.ENTERPRISE,
};
exports.GRACE_PERIOD_DURATION = 14;
exports.LOYALTY_GRACE_PERIOD_END_DATE = '2025-07-31';
exports.SEAT_PRICE_CAP = 9;
exports.LOYALTY_SEAT_PRICE_CAP = 4;
exports.PlanLimitUpgradeMessages = {
    [PlanLimitTypes.LIMIT_FREE_WORKSPACE]: 'to add more workspaces.',
    [PlanLimitTypes.LIMIT_EDITOR]: 'to add more editors.',
    [PlanLimitTypes.LIMIT_COMMENTER]: 'to add more commenters.',
    [PlanLimitTypes.LIMIT_API_PER_SECOND]: 'due to reaching the API per second limit.',
    [PlanLimitTypes.LIMIT_AI_TOKEN]: 'due to reaching the AI token usage limit.',
    [PlanLimitTypes.LIMIT_API_CALL]: 'due to reaching the API call limit.',
    [PlanLimitTypes.LIMIT_AUDIT_RETENTION]: 'to increase audit retention.',
    [PlanLimitTypes.LIMIT_AUTOMATION_RUN]: 'to run more automations.',
    [PlanLimitTypes.LIMIT_AUTOMATION_RETENTION]: 'to increase automation retention.',
    [PlanLimitTypes.LIMIT_WEBHOOK_PER_WORKSPACE]: 'to add more webhooks.',
    [PlanLimitTypes.LIMIT_EXTENSION_PER_WORKSPACE]: 'to add more extensions.',
    [PlanLimitTypes.LIMIT_SNAPSHOT_PER_WORKSPACE]: 'due to reaching the snapshot limit.',
    [PlanLimitTypes.LIMIT_STORAGE_PER_WORKSPACE]: 'due to reaching the storage limit.',
    [PlanLimitTypes.LIMIT_RECORD_PER_WORKSPACE]: 'as the record limit has been reached.',
    [PlanLimitTypes.LIMIT_BASE_PER_WORKSPACE]: 'to add more bases.',
    [PlanLimitTypes.LIMIT_EXTERNAL_SOURCE_PER_WORKSPACE]: 'to connect more external sources.',
    [PlanLimitTypes.LIMIT_TABLE_PER_BASE]: 'to add more tables in a base.',
    [PlanLimitTypes.LIMIT_COLUMN_PER_TABLE]: 'to add more columns in a table.',
    [PlanLimitTypes.LIMIT_WEBHOOK_PER_TABLE]: 'to add more table-level webhooks.',
    [PlanLimitTypes.LIMIT_VIEW_PER_TABLE]: 'to add more views in a table.',
    [PlanLimitTypes.LIMIT_FILTER_PER_VIEW]: 'to add more filters in a view.',
    [PlanLimitTypes.LIMIT_SORT_PER_VIEW]: 'to add more sort rules in a view.',
    [PlanLimitTypes.LIMIT_ATTACHMENTS_IN_CELL]: 'to add more attachments in a cell.',
};
exports.PlanFeatureUpgradeMessages = {
    [PlanFeatureTypes.FEATURE_AI]: 'to enable AI features.',
    [PlanFeatureTypes.FEATURE_AI_INTEGRATIONS]: 'to use AI integrations.',
    [PlanFeatureTypes.FEATURE_AT_MENTION]: 'to use @mention in comments.',
    [PlanFeatureTypes.FEATURE_AUDIT_WORKSPACE]: 'to access workspace audit logs.',
    [PlanFeatureTypes.FEATURE_COMMENT_RESOLVE]: 'to enable comment resolution.',
    [PlanFeatureTypes.FEATURE_CUSTOM_URL]: 'to use a custom URL.',
    [PlanFeatureTypes.FEATURE_DISCUSSION_MODE]: 'to use discussion mode.',
    [PlanFeatureTypes.FEATURE_EXTENSIONS]: 'to enable extensions.',
    [PlanFeatureTypes.FEATURE_FILE_MODE]: 'to enable file mode.',
    [PlanFeatureTypes.FEATURE_FORM_URL_REDIRECTION]: 'to access redirect after form submission feature',
    [PlanFeatureTypes.FEATURE_FORM_CUSTOM_LOGO]: 'to add a custom logo to forms.',
    [PlanFeatureTypes.FEATURE_FORM_FIELD_ON_CONDITION]: 'to access conditional form fields feature',
    [PlanFeatureTypes.FEATURE_FORM_FIELD_VALIDATION]: 'to access form field validation feature',
    [PlanFeatureTypes.FEATURE_GROUP_BY_AGGREGATIONS]: 'to use group-by aggregations.',
    [PlanFeatureTypes.FEATURE_HIDE_BRANDING]: 'to remove branding.',
    [PlanFeatureTypes.FEATURE_LTAR_LIMIT_SELECTION_BY_FILTER]: 'to limit row selection by filters.',
    [PlanFeatureTypes.FEATURE_LOOKUP_LIMIT_RECORDS_BY_FILTER]: 'to limit lookup records by filters.',
    [PlanFeatureTypes.FEATURE_ROLLUP_LIMIT_RECORDS_BY_FILTER]: 'to limit rollup records by filters.',
    [PlanFeatureTypes.FEATURE_PERSONAL_VIEWS]: 'to use personal views.',
    [PlanFeatureTypes.FEATURE_SCRIPTS]: 'to enable scripts.',
    [PlanFeatureTypes.FEATURE_DASHBOARD]: 'to use dashboards.',
    [PlanFeatureTypes.FEATURE_SSO]: 'to enable SSO (Single Sign-On).',
    [PlanFeatureTypes.FEATURE_WEBHOOK_CUSTOM_PAYLOAD]: 'to send custom webhook payloads.',
    [PlanFeatureTypes.FEATURE_WORKSPACE_CUSTOM_LOGO]: 'to upload a custom image as workspace avatar',
    [PlanFeatureTypes.FEATURE_CURRENT_USER_FILTER]: 'to filter view by current user',
    [PlanFeatureTypes.FEATURE_ROW_COLOUR]: 'to use row colouring.',
    [PlanFeatureTypes.FEATURE_TABLE_AND_FIELD_PERMISSIONS]: 'to use table and field permissions.',
    [PlanFeatureTypes.FEATURE_PRIVATE_BASES]: 'to use private bases.',
    [PlanFeatureTypes.FEATURE_API_MEMBER_MANAGEMENT]: 'to use member management api.',
};
const getUpgradeMessage = (limitOrFeature) => {
    if (!limitOrFeature)
        return '';
    if (exports.PlanLimitUpgradeMessages[limitOrFeature]) {
        return exports.PlanLimitUpgradeMessages[limitOrFeature];
    }
    if (exports.PlanFeatureUpgradeMessages[limitOrFeature]) {
        return exports.PlanFeatureUpgradeMessages[limitOrFeature];
    }
    return limitOrFeature;
};
exports.getUpgradeMessage = getUpgradeMessage;
