// nocodb-sdk/build/module/lib/error/nc-base.error.js
export class NcBaseError extends Error {
    constructor(message) {
        super(message);
    }
}
export class BadRequestV2 extends NcBaseError {
}
export class NotAllowed extends NcBaseError {
}
export class Unauthorized extends NcBaseError {
}
export class Forbidden extends NcBaseError {
}
export class NotFound extends NcBaseError {
}
export class SsoError extends NcBaseError {
}
export class MetaError extends NcBaseError {
    constructor(param) {
        super(param.message);
        Object.assign(this, param);
    }
}
export class ExternalError extends NcBaseError {
    constructor(error) {
        super(error.message);
        Object.assign(this, error);
    }
}
export class ExternalTimeout extends ExternalError {
}
export class UnprocessableEntity extends NcBaseError {
}
export class OptionsNotExistsError extends BadRequestV2 {
    constructor({ columnTitle, options, validOptions, }) {
        super(`Invalid option(s) "${options.join(', ')}" provided for column "${columnTitle}". Valid options are "${validOptions.join(', ')}"`);
        this.columnTitle = columnTitle;
        this.options = options;
        this.validOptions = validOptions;
    }
}
export class TestConnectionError extends NcBaseError {
    constructor(message, sql_code) {
        super(message);
        this.sql_code = sql_code;
    }
}
export class NcBaseErrorv2 extends NcBaseError {
    constructor(message, code, error, args) {
        super(message);
        this.error = error;
        this.code = code;
        this.details = args === null || args === void 0 ? void 0 : args.details;
    }
}
